<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="generator" content="pandoc" />

        <meta name="author" content="James Dorey (Script)" />
        <meta name="author" content="Robert OReilly (RMD)" />
    
    
    <title>A globally-synthesized and flagged bee occurrence dataset and cleaning workflow</title>

        <script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
        <script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link href="site_libs/bootstrap-3.3.7/css/bootstrap.min.css" rel="stylesheet" />
        <script src="site_libs/bootstrap-3.3.7/js/bootstrap.min.js"></script>
        <script src="site_libs/navigation-1.1/tabsets.js"></script>
        <script src="site_libs/navigation-1.1/codefolding.js"></script>
        <link href="site_libs/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
        <script src="site_libs/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
        <link href="site_libs/readthedown-0.1/readthedown.css" rel="stylesheet" />
        <link href="site_libs/readthedown-0.1/readthedown_fonts_embed.css" rel="stylesheet" />
        <script src="site_libs/readthedown-0.1/readthedown.js"></script>
        <script src="site_libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
        <link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
        <script src="site_libs/datatables-binding-0.28/datatables.js"></script>
        <link href="site_libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
        <link href="site_libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
        <script src="site_libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
        <link href="site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
        <script src="site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
        <link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
        <link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
    
    
        <style type="text/css">code{white-space: pre;}</style>
    <style type="text/css">
      pre > code.sourceCode { white-space: pre; position: relative; }
      pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
      pre > code.sourceCode > span:empty { height: 1.2em; }
      .sourceCode { overflow: visible; }
      code.sourceCode > span { color: inherit; text-decoration: inherit; }
      div.sourceCode { margin: 1em 0; }
      pre.sourceCode { margin: 0; }
      @media screen {
      div.sourceCode { overflow: auto; }
      }
      @media print {
      pre > code.sourceCode { white-space: pre-wrap; }
      pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
      }
      pre.numberSource code
        { counter-reset: source-line 0; }
      pre.numberSource code > span
        { position: relative; left: -4em; counter-increment: source-line; }
      pre.numberSource code > span > a:first-child::before
        { content: counter(source-line);
          position: relative; left: -1em; text-align: right; vertical-align: baseline;
          border: none; display: inline-block;
          -webkit-touch-callout: none; -webkit-user-select: none;
          -khtml-user-select: none; -moz-user-select: none;
          -ms-user-select: none; user-select: none;
          padding: 0 4px; width: 4em;
          color: #aaaaaa;
        }
      pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
      div.sourceCode
        {  background-color: #f8f8f8; }
      @media screen {
      pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
      }
      code span.al { color: #ef2929; } /* Alert */
      code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
      code span.at { color: #204a87; } /* Attribute */
      code span.bn { color: #0000cf; } /* BaseN */
      code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
      code span.ch { color: #4e9a06; } /* Char */
      code span.cn { color: #8f5902; } /* Constant */
      code span.co { color: #8f5902; font-style: italic; } /* Comment */
      code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
      code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
      code span.dt { color: #204a87; } /* DataType */
      code span.dv { color: #0000cf; } /* DecVal */
      code span.er { color: #a40000; font-weight: bold; } /* Error */
      code span.ex { } /* Extension */
      code span.fl { color: #0000cf; } /* Float */
      code span.fu { color: #204a87; font-weight: bold; } /* Function */
      code span.im { } /* Import */
      code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
      code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
      code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
      code span.ot { color: #8f5902; } /* Other */
      code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
      code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
      code span.ss { color: #4e9a06; } /* SpecialString */
      code span.st { color: #4e9a06; } /* String */
      code span.va { color: #000000; } /* Variable */
      code span.vs { color: #4e9a06; } /* VerbatimString */
      code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
    </style>
    
    
    <!-- tabsets -->
    <script>
      $(document).ready(function () {
	  window.buildTabsets("toc");
      });
      $(document).ready(function () {
	  $('.tabset-dropdown > .nav-tabs > li').click(function () {
	      $(this).parent().toggleClass('nav-tabs-open')
	  });
      });
    </script>

    <!-- code folding -->
        <script>
      $(document).ready(function () {
	  	  	  window.initializeCodeFolding("show" === "show");
	        });
    </script>
    
    <!-- code download -->
    
    <!-- tabsets dropdown -->

    <style type="text/css">
      .tabset-dropdown > .nav-tabs {
	  display: inline-table;
	  max-height: 500px;
	  min-height: 44px;
	  overflow-y: auto;
	  background: white;
	  border: 1px solid #ddd;
	  border-radius: 4px;
      }
      
      .tabset-dropdown > .nav-tabs > li.active:before {
	  content: "";
	  font-family: 'Glyphicons Halflings';
	  display: inline-block;
	  padding: 10px;
	  border-right: 1px solid #ddd;
      }
      
      .tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
	  content: "&#xe258;";
	  border: none;
      }
      
      .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
	  content: "";
	  font-family: 'Glyphicons Halflings';
	  display: inline-block;
	  padding: 10px;
	  border-right: 1px solid #ddd;
      }
      
      .tabset-dropdown > .nav-tabs > li.active {
	  display: block;
      }

      .tabset-dropdown > .nav-tabs > li.active a {
  	  padding: 0 15px !important;
      }

      .tabset-dropdown > .nav-tabs > li > a,
      .tabset-dropdown > .nav-tabs > li > a:focus,
      .tabset-dropdown > .nav-tabs > li > a:hover {
	  border: none;
	  display: inline-block;
	  border-radius: 4px;
	  background-color: transparent;
      }
      
      .tabset-dropdown > .nav-tabs.nav-tabs-open > li {
	  display: block;
	  float: none;
      }
      
      .tabset-dropdown > .nav-tabs > li {
	  display: none;
	  margin-left: 0 !important;
      }
    </style>
    
</head>

<body class="preload">

   	
         <!-- readthedown start -->   
   <div id="content" data-toggle="wy-nav-shift">
     <nav id="nav-top" role="navigation" aria-label="top navigation">
       <a role="button" href="#" data-toggle="wy-nav-top"><span class="glyphicon glyphicon-menu-hamburger"></span></a>
     </nav>
         
      <div class="btn-group pull-right">
     <button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
     <ul class="dropdown-menu" style="min-width: 50px;">
	    	    <li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
	    <li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
	    	    	         </ul>
   </div>
   
      
      <div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
        <div class="container">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="index.html">BeeDC</a>
          </div>
          <div id="navbar" class="navbar-collapse collapse">
            <ul class="nav navbar-nav">
              <li>
        <a href="index.html">
          <span class="fa fa-home"></span>
           
          Home
        </a>
      </li>
      <li>
        <a href="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23.html">
          <span class="fa fa-beer"></span>
           
          Tutorial
        </a>
      </li>
      <li>
        <a href="about.html">
          <span class="fa fa-info"></span>
           
          About
        </a>
      </li>
            </ul>
            <ul class="nav navbar-nav navbar-right">
              <li>
        <a href="https://www.biorxiv.org/content/10.1101/2023.06.30.547152v1">
          <span class="fa fa-question fa-lg"></span>
           
        </a>
      </li>
            </ul>
          </div><!--/.nav-collapse -->
        </div><!--/.container -->
      </div><!--/.navbar -->
        
      <h1 class="title">A globally-synthesized and flagged bee
occurrence dataset and cleaning workflow</h1>
      
         <!-- readthedown authors -->
   <div id="sidebar">
    <h2><a href="#content">A globally-synthesized and flagged bee
occurrence dataset and cleaning workflow</a></h2>
    <div id="toc">
      <ul>
      <li><a href="#script-preparation" id="toc-script-preparation">0.0
      Script preparation</a>
      <ul>
      <li><a href="#working-directory" id="toc-working-directory">0.1
      Working directory</a></li>
      <li><a href="#install-packages-if-needed"
      id="toc-install-packages-if-needed">0.2 Install packages (if
      needed)</a></li>
      <li><a href="#load-packages" id="toc-load-packages">0.3 Load
      packages</a></li>
      </ul></li>
      <li><a href="#data-merge" id="toc-data-merge">1.0 Data merge</a>
      <ul>
      <li><a href="#download-ala-data" id="toc-download-ala-data">1.1
      Download ALA data</a></li>
      <li><a href="#import-and-merge-ala-scan-idigbio-and-gbif-data"
      id="toc-import-and-merge-ala-scan-idigbio-and-gbif-data">1.2
      Import and merge ALA, SCAN, iDigBio and GBIF data</a></li>
      <li><a href="#import-usgs-data" id="toc-import-usgs-data">1.3
      Import USGS Data</a></li>
      <li><a href="#formatted-source-importer"
      id="toc-formatted-source-importer">1.4 Formatted Source
      Importer</a></li>
      <li><a href="#save-data" id="toc-save-data">1.5 Save data</a></li>
      </ul></li>
      <li><a href="#data-preperation" id="toc-data-preperation">2.0 Data
      preperation</a>
      <ul>
      <li><a href="#standardise-datasets"
      id="toc-standardise-datasets">2.1 standardise datasets</a></li>
      <li><a href="#integrating-paige-dataset"
      id="toc-integrating-paige-dataset">2.2 Integrating Paige
      dataset</a></li>
      <li><a href="#additional-datasets"
      id="toc-additional-datasets">2.3 Additional datasets</a></li>
      </ul></li>
      <li><a href="#initial-flags" id="toc-initial-flags">3.0 Initial
      flags</a>
      <ul>
      <li><a href="#sciname" id="toc-sciname">3.1 SciName</a></li>
      <li><a href="#misscoords" id="toc-misscoords">3.2
      MissCoords</a></li>
      <li><a href="#outofrange" id="toc-outofrange">3.3
      OutOfRange</a></li>
      <li><a href="#source" id="toc-source">3.4 Source</a></li>
      <li><a href="#countryname" id="toc-countryname">3.5
      CountryName</a></li>
      <li><a href="#standardconames" id="toc-standardconames">3.6
      StandardCoNames</a></li>
      <li><a href="#transpcoords" id="toc-transpcoords">3.7
      TranspCoords</a></li>
      <li><a href="#coord-country" id="toc-coord-country">3.8
      Coord-country</a></li>
      <li><a href="#georefissue" id="toc-georefissue">3.9
      GeoRefIssue</a></li>
      <li><a href="#flag-absent" id="toc-flag-absent">3.10 Flag
      Absent</a></li>
      <li><a href="#flag-license" id="toc-flag-license">3.11 flag
      License</a></li>
      <li><a href="#gbif-issue" id="toc-gbif-issue">3.12 GBIF
      issue</a></li>
      <li><a href="#flag-reports" id="toc-flag-reports">3.13 Flag
      Reports</a></li>
      <li><a href="#save" id="toc-save">3.14 Save</a></li>
      </ul></li>
      <li><a href="#taxonomy" id="toc-taxonomy">4.0 Taxonomy</a>
      <ul>
      <li><a href="#prep-data-names" id="toc-prep-data-names">4.1 Prep
      data names</a></li>
      <li><a href="#taxo-harmonization" id="toc-taxo-harmonization">4.2
      Taxo harmonization</a></li>
      <li><a href="#save-flags" id="toc-save-flags">4.3 Save
      flags</a></li>
      </ul></li>
      <li><a href="#space-the-final-frontier"
      id="toc-space-the-final-frontier">5.0 Space: The final
      frontier</a>
      <ul>
      <li><a href="#coord-precision" id="toc-coord-precision">5.1 Coord
      precision</a></li>
      <li><a href="#common-spatial-issues"
      id="toc-common-spatial-issues">5.2 Common spatial issues</a></li>
      <li><a href="#diagonal-grid" id="toc-diagonal-grid">5.3 diagonal +
      grid</a></li>
      <li><a href="#uncertainty" id="toc-uncertainty">5.4
      Uncertainty</a></li>
      <li><a href="#country-checklist" id="toc-country-checklist">5.5
      Country Checklist</a></li>
      <li><a href="#map-spatial-errors" id="toc-map-spatial-errors">5.6
      Map spatial errors</a></li>
      <li><a href="#space-report" id="toc-space-report">5.7 Space
      report</a></li>
      <li><a href="#space-figures" id="toc-space-figures">5.8 Space
      figures</a></li>
      <li><a href="#save-flags-1" id="toc-save-flags-1">5.9 Save
      flags</a></li>
      <li><a href="#save-1" id="toc-save-1">5.10 Save</a></li>
      </ul></li>
      <li><a href="#time" id="toc-time">6.0 Time</a>
      <ul>
      <li><a href="#recover-dates" id="toc-recover-dates">6.1 Recover
      dates</a></li>
      <li><a href="#no-eventdate" id="toc-no-eventdate">6.2 No
      eventDate</a></li>
      <li><a href="#year-from-eventdate"
      id="toc-year-from-eventdate">6.3 Year from eventDate</a></li>
      <li><a href="#old-records" id="toc-old-records">6.4 Old
      records</a></li>
      <li><a href="#time-report" id="toc-time-report">6.6 Time
      report</a></li>
      <li><a href="#time-figures" id="toc-time-figures">6.7 Time
      figures</a></li>
      <li><a href="#save-flags-2" id="toc-save-flags-2">6.8 Save
      flags</a></li>
      </ul></li>
      <li><a href="#de-duplication" id="toc-de-duplication">7.0
      De-duplication</a>
      <ul>
      <li><a href="#deduplicate" id="toc-deduplicate">7.1
      deDuplicate</a></li>
      </ul></li>
      <li><a href="#data-filtering" id="toc-data-filtering">8.0 Data
      filtering</a>
      <ul>
      <li><a href="#rm-outliers" id="toc-rm-outliers">8.1 rm
      Outliers</a></li>
      <li><a href="#save-uncleaned" id="toc-save-uncleaned">8.2 Save
      uncleaned</a></li>
      <li><a href="#filter" id="toc-filter">8.3 Filter</a></li>
      </ul></li>
      <li><a href="#summary-figures" id="toc-summary-figures">9.0
      Summary figures</a>
      <ul>
      <li><a href="#duplicate-chorddiagrams"
      id="toc-duplicate-chorddiagrams">9.1 Duplicate
      chordDiagrams</a></li>
      <li><a href="#duplicate-histogram"
      id="toc-duplicate-histogram">9.2 Duplicate histogram</a></li>
      <li><a href="#flags-by-source" id="toc-flags-by-source">9.3 Flags
      by source</a></li>
      <li><a href="#maps" id="toc-maps">9.4 Maps</a></li>
      </ul></li>
      <li><a href="#fin" id="toc-fin">FIN</a></li>
      </ul>
    </div>
    <div id="postamble" data-toggle="wy-nav-shift" class="status">
                  <p class="author"><span class="glyphicon glyphicon-user"></span> James
Dorey (Script)</p>
                        <p class="author"><span class="glyphicon glyphicon-user"></span> Robert
OReilly (RMD)</p>
                      </div>
   </div>
     

   
      
   
<!-- Don't indent these lines or it will mess pre blocks indentation --> 
<div id="main">
<div class="alert alert-info">
<p><strong> Attention:</strong><br />
This R script was written by James Dorey, starting on the 2nd of May
2022. The script intends to clean occurrence data from several sources
using a combination of custom functions and functions from the “bdc”
package. / For queries, please feel free to contact James Dorey at <a
href="mailto:jbdorey@me.com" class="email">jbdorey@me.com</a><br />
</p>
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">error =</span> <span class="cn">TRUE</span>, <span class="at">eval =</span> <span class="cn">TRUE</span>, <span class="at">tidy =</span> <span class="cn">TRUE</span>, <span class="at">warning =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div id="script-preparation" class="section level1">
<h1>0.0 Script preparation</h1>
<div id="working-directory" class="section level2">
<h2>0.1 Working directory</h2>
<p>Choose the path to the root folder in which all other folders can be
found</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>RootPath <span class="ot">&lt;-</span> <span class="st">&quot;E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning&quot;</span></span></code></pre></div>
<p>Choose the location of your data</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>DataPath <span class="ot">&lt;-</span> <span class="fu">paste</span>(RootPath,</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                  <span class="st">&quot;/BeeData&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Choose the location of the R script</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>ScriptPath <span class="ot">&lt;-</span> <span class="fu">paste</span>(RootPath,<span class="st">&quot;/R_scripting/BeeCleaning_SciData/R_functions&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Choose the taxonomy path</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>DiscLifePath <span class="ot">&lt;-</span> <span class="fu">paste</span>(RootPath,</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>                      <span class="st">&quot;/BeeData/DiscoverLife_Data&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Choose the path for the intermediate data storage</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>intermediatePath <span class="ot">&lt;-</span> <span class="fu">paste</span>(RootPath,</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                          <span class="st">&quot;/BeeData/Output/Intermediate&quot;</span>, </span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                          <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Set the working directory</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">setwd</span>(DataPath)</span></code></pre></div>
</div>
<div id="install-packages-if-needed" class="section level2">
<h2>0.2 Install packages (if needed)</h2>
<p>Install only those packages that are not already present in your
system. You may need to install gdal on your computer. This can be done
on mac by using Homebrew in the terminal and the command “brew install
gdal”. Somethtimes renv does not capture all dependancies, and will be
unable to detect packages used in certain ways<br />
</p>
<p>E.g. for (package in c(“dplyr”, “ggplot2”)) {library(package,
character.only = TRUE)}<br />
</p>
<p>can use “renv::settings<span
class="math inline">\(snapshot.type(&quot;all&quot;)&quot; to capture
all packages installed in library or
&quot;renv::settings\)</span>snapshot.type(”explicit”)” and specify
packages.</p>
<pre><code>        install.packages(&quot;renv&quot;)
  
        renv::init()
  </code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>            list.of.packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;R.utils&quot;</span>,           <span class="co"># To use gunzip</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>                                  <span class="st">&quot;magrittr&quot;</span>,          <span class="co"># to use pipes</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>                                  <span class="st">&quot;ggplot2&quot;</span>,           <span class="co">#  Creates many easthetic plots</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>                                  <span class="st">&quot;ggVennDiagram&quot;</span>,     <span class="co"># Extends ggplot2 to make venn diagrams</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>                                  <span class="st">&quot;dplyr&quot;</span>,             <span class="co">#  Part of the tidyverse</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>                                  <span class="st">&quot;tibble&quot;</span>,            <span class="co"># To use tibbles</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>                                  <span class="st">&quot;forcats&quot;</span>,           <span class="co"># tidyverse for working with factors</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>                                  <span class="st">&quot;tidyr&quot;</span>,             <span class="co">#  Part of the tidyverse</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>                                  <span class="st">&quot;rlist&quot;</span>,             <span class="co"># Package to save lists</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>                                  <span class="st">&quot;galah&quot;</span>,             <span class="co">#  To download ALA data</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>                                  <span class="st">&quot;praise&quot;</span>,            <span class="co">#  To whispers sweet nothings </span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>                                  <span class="st">&quot;EML&quot;</span>,               <span class="co">#  To work with .eml files</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>                                  <span class="st">&quot;emld&quot;</span>,              <span class="co">#  To work with .eml files</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>                                  <span class="st">&quot;rlang&quot;</span>,             <span class="co">#  Part of the tidyverse — core functions</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>                                  <span class="st">&quot;xml2&quot;</span>,              <span class="co">#  Part of the tidyverse — reads .xml files</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>                                  <span class="st">&quot;readr&quot;</span>,             <span class="co">#  Part of the tidyverse — reads files (e.g., .csv)</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>                                  <span class="st">&quot;stringr&quot;</span>,           <span class="co">#  Part of the tidyverse — works with text strings</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>                                  <span class="st">&quot;lubridate&quot;</span>,         <span class="co">#  Part of the tidyverse — works with dates</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a>                                  <span class="st">&quot;tidyselect&quot;</span>,        <span class="co">#  Part of the tidyverse</span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>                                  <span class="st">&quot;mgsub&quot;</span>,             <span class="co">#  To perform multiple text substitutions</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>                                  <span class="st">&quot;bdc&quot;</span>,               <span class="co"># data cleaning package</span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>                                  <span class="st">&quot;rvest&quot;</span>,             <span class="co"># Package for interfacing with and downloading from the internet</span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>                                  <span class="st">&quot;rnaturalearth&quot;</span>,     <span class="co">#  Global vector map data </span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>                                  <span class="st">&quot;rnaturalearthdata&quot;</span>, <span class="co">#  To access the above global map data</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>                                  <span class="st">&quot;countrycode&quot;</span>,       <span class="co"># Package to deal with country codes</span></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a>                                  <span class="st">&quot;rangeBuilder&quot;</span>,</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a>                                  <span class="st">&quot;rworldmap&quot;</span>,</span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a>                                  <span class="st">&quot;rmarkdown&quot;</span>,</span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>                                  <span class="st">&quot;rmdformats&quot;</span>,</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>                                  <span class="st">&quot;prettydoc&quot;</span>,</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>                                  <span class="st">&quot;hexbin&quot;</span>,</span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>                                  <span class="st">&quot;readxl&quot;</span>,</span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a>                                  <span class="st">&quot;cowplot&quot;</span>,           <span class="co"># ggplot2 helper package</span></span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a>                                  <span class="st">&quot;ggspatial&quot;</span>,</span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a>                                  <span class="st">&quot;rmarkdown&quot;</span>,</span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a>                                  <span class="st">&quot;rmdformats&quot;</span>,</span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a>                                  <span class="st">&quot;htmltools&quot;</span>,</span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a>                                  <span class="st">&quot;prettydoc&quot;</span>)         <span class="co">#  Makes ggplot2 create north arrows or scale bars</span></span></code></pre></div>
<pre><code>  Install sf and terra seperately
  
  
        renv::install(c(&quot;sf&quot;,&quot;terra&quot;), type = &quot;binary&quot;)
        renv::install(c(&quot;sf&quot;,&quot;terra&quot;), type = &quot;binary&quot;)
  
  
  #List the new (not installed) packages and then if there are any, install them.
  
  
              renv::install(packages = c(list.of.packages, &quot;ropensci/rnaturalearthhires&quot;, &quot;mattflor/chorddiag&quot;), 
                      rebuild = FALSE) # try changing to TRUE if you&#39;re having package troubles
  </code></pre>
</div>
<div id="load-packages" class="section level2">
<h2>0.3 Load packages</h2>
<p>Load all packages from the list specified above, with the addition of
“rnaturalearthhires”</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>      <span class="fu">lapply</span>(<span class="fu">c</span>(list.of.packages, <span class="st">&quot;rnaturalearthhires&quot;</span>, <span class="st">&quot;chorddiag&quot;</span>, <span class="st">&quot;sf&quot;</span>,<span class="st">&quot;terra&quot;</span>), </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>       library, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Linking to GEOS 3.11.2, GDAL 3.6.2, PROJ 9.2.0; sf_use_s2() is TRUE</code></pre>
<pre><code>## terra 1.7.39</code></pre>
<pre><code>## 
## Attaching package: &#39;terra&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:tidyr&#39;:
## 
##     extract</code></pre>
<pre><code>## The following objects are masked from &#39;package:R.utils&#39;:
## 
##     draw, extract, resample, unwrap, wrap</code></pre>
<pre><code>## The following objects are masked from &#39;package:R.oo&#39;:
## 
##     extend, trim</code></pre>
<pre><code>## The following object is masked from &#39;package:knitr&#39;:
## 
##     spin</code></pre>
<pre><code>## The following objects are masked from &#39;package:magrittr&#39;:
## 
##     extract, inset</code></pre>
<pre><code>## [[1]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[2]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[3]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[4]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[5]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[6]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[7]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[8]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[9]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[10]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[11]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[12]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[13]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[14]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[15]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[16]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[17]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[18]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[19]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[20]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[21]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[22]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[23]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[24]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[25]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[26]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[27]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[28]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[29]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[30]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[31]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[32]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[33]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[34]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[35]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[36]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[37]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[38]]
##  [1] &quot;htmltools&quot;         &quot;ggspatial&quot;         &quot;cowplot&quot;          
##  [4] &quot;readxl&quot;            &quot;hexbin&quot;            &quot;prettydoc&quot;        
##  [7] &quot;rmdformats&quot;        &quot;rmarkdown&quot;         &quot;rworldmap&quot;        
## [10] &quot;sp&quot;                &quot;rangeBuilder&quot;      &quot;countrycode&quot;      
## [13] &quot;rnaturalearthdata&quot; &quot;rnaturalearth&quot;     &quot;rvest&quot;            
## [16] &quot;bdc&quot;               &quot;mgsub&quot;             &quot;tidyselect&quot;       
## [19] &quot;lubridate&quot;         &quot;stringr&quot;           &quot;readr&quot;            
## [22] &quot;xml2&quot;              &quot;rlang&quot;             &quot;emld&quot;             
## [25] &quot;EML&quot;               &quot;praise&quot;            &quot;galah&quot;            
## [28] &quot;rlist&quot;             &quot;tidyr&quot;             &quot;forcats&quot;          
## [31] &quot;tibble&quot;            &quot;ggVennDiagram&quot;     &quot;ggplot2&quot;          
## [34] &quot;R.utils&quot;           &quot;R.oo&quot;              &quot;R.methodsS3&quot;      
## [37] &quot;knitr&quot;             &quot;magrittr&quot;          &quot;dplyr&quot;            
## [40] &quot;rgnparser&quot;         &quot;stats&quot;             &quot;graphics&quot;         
## [43] &quot;grDevices&quot;         &quot;utils&quot;             &quot;datasets&quot;         
## [46] &quot;methods&quot;           &quot;base&quot;             
## 
## [[39]]
##  [1] &quot;rnaturalearthhires&quot; &quot;htmltools&quot;          &quot;ggspatial&quot;         
##  [4] &quot;cowplot&quot;            &quot;readxl&quot;             &quot;hexbin&quot;            
##  [7] &quot;prettydoc&quot;          &quot;rmdformats&quot;         &quot;rmarkdown&quot;         
## [10] &quot;rworldmap&quot;          &quot;sp&quot;                 &quot;rangeBuilder&quot;      
## [13] &quot;countrycode&quot;        &quot;rnaturalearthdata&quot;  &quot;rnaturalearth&quot;     
## [16] &quot;rvest&quot;              &quot;bdc&quot;                &quot;mgsub&quot;             
## [19] &quot;tidyselect&quot;         &quot;lubridate&quot;          &quot;stringr&quot;           
## [22] &quot;readr&quot;              &quot;xml2&quot;               &quot;rlang&quot;             
## [25] &quot;emld&quot;               &quot;EML&quot;                &quot;praise&quot;            
## [28] &quot;galah&quot;              &quot;rlist&quot;              &quot;tidyr&quot;             
## [31] &quot;forcats&quot;            &quot;tibble&quot;             &quot;ggVennDiagram&quot;     
## [34] &quot;ggplot2&quot;            &quot;R.utils&quot;            &quot;R.oo&quot;              
## [37] &quot;R.methodsS3&quot;        &quot;knitr&quot;              &quot;magrittr&quot;          
## [40] &quot;dplyr&quot;              &quot;rgnparser&quot;          &quot;stats&quot;             
## [43] &quot;graphics&quot;           &quot;grDevices&quot;          &quot;utils&quot;             
## [46] &quot;datasets&quot;           &quot;methods&quot;            &quot;base&quot;              
## 
## [[40]]
##  [1] &quot;chorddiag&quot;          &quot;rnaturalearthhires&quot; &quot;htmltools&quot;         
##  [4] &quot;ggspatial&quot;          &quot;cowplot&quot;            &quot;readxl&quot;            
##  [7] &quot;hexbin&quot;             &quot;prettydoc&quot;          &quot;rmdformats&quot;        
## [10] &quot;rmarkdown&quot;          &quot;rworldmap&quot;          &quot;sp&quot;                
## [13] &quot;rangeBuilder&quot;       &quot;countrycode&quot;        &quot;rnaturalearthdata&quot; 
## [16] &quot;rnaturalearth&quot;      &quot;rvest&quot;              &quot;bdc&quot;               
## [19] &quot;mgsub&quot;              &quot;tidyselect&quot;         &quot;lubridate&quot;         
## [22] &quot;stringr&quot;            &quot;readr&quot;              &quot;xml2&quot;              
## [25] &quot;rlang&quot;              &quot;emld&quot;               &quot;EML&quot;               
## [28] &quot;praise&quot;             &quot;galah&quot;              &quot;rlist&quot;             
## [31] &quot;tidyr&quot;              &quot;forcats&quot;            &quot;tibble&quot;            
## [34] &quot;ggVennDiagram&quot;      &quot;ggplot2&quot;            &quot;R.utils&quot;           
## [37] &quot;R.oo&quot;               &quot;R.methodsS3&quot;        &quot;knitr&quot;             
## [40] &quot;magrittr&quot;           &quot;dplyr&quot;              &quot;rgnparser&quot;         
## [43] &quot;stats&quot;              &quot;graphics&quot;           &quot;grDevices&quot;         
## [46] &quot;utils&quot;              &quot;datasets&quot;           &quot;methods&quot;           
## [49] &quot;base&quot;              
## 
## [[41]]
##  [1] &quot;sf&quot;                 &quot;chorddiag&quot;          &quot;rnaturalearthhires&quot;
##  [4] &quot;htmltools&quot;          &quot;ggspatial&quot;          &quot;cowplot&quot;           
##  [7] &quot;readxl&quot;             &quot;hexbin&quot;             &quot;prettydoc&quot;         
## [10] &quot;rmdformats&quot;         &quot;rmarkdown&quot;          &quot;rworldmap&quot;         
## [13] &quot;sp&quot;                 &quot;rangeBuilder&quot;       &quot;countrycode&quot;       
## [16] &quot;rnaturalearthdata&quot;  &quot;rnaturalearth&quot;      &quot;rvest&quot;             
## [19] &quot;bdc&quot;                &quot;mgsub&quot;              &quot;tidyselect&quot;        
## [22] &quot;lubridate&quot;          &quot;stringr&quot;            &quot;readr&quot;             
## [25] &quot;xml2&quot;               &quot;rlang&quot;              &quot;emld&quot;              
## [28] &quot;EML&quot;                &quot;praise&quot;             &quot;galah&quot;             
## [31] &quot;rlist&quot;              &quot;tidyr&quot;              &quot;forcats&quot;           
## [34] &quot;tibble&quot;             &quot;ggVennDiagram&quot;      &quot;ggplot2&quot;           
## [37] &quot;R.utils&quot;            &quot;R.oo&quot;               &quot;R.methodsS3&quot;       
## [40] &quot;knitr&quot;              &quot;magrittr&quot;           &quot;dplyr&quot;             
## [43] &quot;rgnparser&quot;          &quot;stats&quot;              &quot;graphics&quot;          
## [46] &quot;grDevices&quot;          &quot;utils&quot;              &quot;datasets&quot;          
## [49] &quot;methods&quot;            &quot;base&quot;              
## 
## [[42]]
##  [1] &quot;terra&quot;              &quot;sf&quot;                 &quot;chorddiag&quot;         
##  [4] &quot;rnaturalearthhires&quot; &quot;htmltools&quot;          &quot;ggspatial&quot;         
##  [7] &quot;cowplot&quot;            &quot;readxl&quot;             &quot;hexbin&quot;            
## [10] &quot;prettydoc&quot;          &quot;rmdformats&quot;         &quot;rmarkdown&quot;         
## [13] &quot;rworldmap&quot;          &quot;sp&quot;                 &quot;rangeBuilder&quot;      
## [16] &quot;countrycode&quot;        &quot;rnaturalearthdata&quot;  &quot;rnaturalearth&quot;     
## [19] &quot;rvest&quot;              &quot;bdc&quot;                &quot;mgsub&quot;             
## [22] &quot;tidyselect&quot;         &quot;lubridate&quot;          &quot;stringr&quot;           
## [25] &quot;readr&quot;              &quot;xml2&quot;               &quot;rlang&quot;             
## [28] &quot;emld&quot;               &quot;EML&quot;                &quot;praise&quot;            
## [31] &quot;galah&quot;              &quot;rlist&quot;              &quot;tidyr&quot;             
## [34] &quot;forcats&quot;            &quot;tibble&quot;             &quot;ggVennDiagram&quot;     
## [37] &quot;ggplot2&quot;            &quot;R.utils&quot;            &quot;R.oo&quot;              
## [40] &quot;R.methodsS3&quot;        &quot;knitr&quot;              &quot;magrittr&quot;          
## [43] &quot;dplyr&quot;              &quot;rgnparser&quot;          &quot;stats&quot;             
## [46] &quot;graphics&quot;           &quot;grDevices&quot;          &quot;utils&quot;             
## [49] &quot;datasets&quot;           &quot;methods&quot;            &quot;base&quot;</code></pre>
<p>Load in R scripts and character strings in our package</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>      <span class="fu">sapply</span>(<span class="fu">list.files</span>(<span class="fu">file.path</span>(ScriptPath), <span class="at">pattern =</span> <span class="st">&quot;.R$&quot;</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>), source)  <span class="co"># loads in all functions</span></span></code></pre></div>
<pre><code>##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/addFlagInfo.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/additionalData_readRs.R
## value   ?                                                                                                                         
## visible FALSE                                                                                                                     
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/addMetaAttributes.R
## value   ?                                                                                                                     
## visible FALSE                                                                                                                 
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/ALA_downloader.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/attr_builder.R
## value   ?                                                                                                                
## visible FALSE                                                                                                            
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/aux_functions.R
## value   ?                                                                                                                 
## visible FALSE                                                                                                             
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/bdc_coord_trans.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/bdc_correct_coordinates.R
## value   ?                                                                                                                           
## visible FALSE                                                                                                                       
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/bdc_create_dir.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/bdc_get_world_map.R
## value   ?                                                                                                                     
## visible FALSE                                                                                                                 
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/chordDiagramR.R
## value   ?                                                                                                                 
## visible FALSE                                                                                                             
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/ColTypeR.R
## value   ?                                                                                                            
## visible FALSE                                                                                                        
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/coordUncerFlagR.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/countryOutliers.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/data_finder.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/data_reader.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/dateFindR.R
## value   ?                                                                                                             
## visible FALSE                                                                                                         
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/dirMaker.R
## value   ?                                                                                                            
## visible FALSE                                                                                                        
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/dupePlotR.R
## value   ?                                                                                                             
## visible FALSE                                                                                                         
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/flag_converter.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/flagChartR.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/FlagManager.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/flagRecorder.R
## value   ?                                                                                                                
## visible FALSE                                                                                                            
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/formatted_combiner.R
## value   ?                                                                                                                      
## visible FALSE                                                                                                                  
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/HarmoniseR.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/Harmonize_chunkR.R
## value   ?                                                                                                                    
## visible FALSE                                                                                                                
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/HigherNamer.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/idMatchR.R
## value   ?                                                                                                            
## visible FALSE                                                                                                        
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/importOccurrences.R
## value   ?                                                                                                                     
## visible FALSE                                                                                                                 
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/install_gnparser.R
## value   ?                                                                                                                    
## visible FALSE                                                                                                                
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_CCInconsistent_chunker.R
## value   ?                                                                                                                              
## visible FALSE                                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_CfC_chunker.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_coordCountryInconsistent.R
## value   ?                                                                                                                                
## visible FALSE                                                                                                                            
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_coordinates_transposed.R
## value   ?                                                                                                                              
## visible FALSE                                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_country_from_coordinates.R
## value   ?                                                                                                                                
## visible FALSE                                                                                                                            
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_countryNameCleanR.R
## value   ?                                                                                                                         
## visible FALSE                                                                                                                     
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_create_figures.R
## value   ?                                                                                                                      
## visible FALSE                                                                                                                  
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_Ctrans_chunker.R
## value   ?                                                                                                                      
## visible FALSE                                                                                                                  
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_diagonAlley.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_dupeSummary.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_flagAbsent.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_flagLicense.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/jbd_GBIFissues.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/manualCorrectR.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/nameSplitR.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/occurrenceFlags.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/PaigeEdits.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/PaigeIntegrateR.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/PaigesSyns.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/plotFlagSummary.R
## value   ?                                                                                                                   
## visible FALSE                                                                                                               
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/repo_merge.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/StateOutliers.R
## value   ?                                                                                                                 
## visible FALSE                                                                                                             
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/summaryFun.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/summaryMaps.R
## value   ?                                                                                                               
## visible FALSE                                                                                                           
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/taxoDuplicator.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/taxoMergeR.R
## value   ?                                                                                                              
## visible FALSE                                                                                                          
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/USGS_formatter.R
## value   ?                                                                                                                  
## visible FALSE                                                                                                              
##         E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/R_scripting/BeeCleaning_SciData/R_functions/utils.R
## value   ?                                                                                                         
## visible FALSE</code></pre>
<p>Save a snapshot of the environment</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>    <span class="co">#  renv::snapshot() #or</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>      <span class="co">#renv::settings$snapshot.type(&quot;all&quot;)</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="data-merge" class="section level1">
<h1>1.0 Data merge</h1>
<div class="alert alert-info">
<p><strong> Attention:</strong><br />
Although each line of code has been validated, in order to save time
knitting the RMarkdown document the next section is display only. If you
are not data merging (section 1.0) or preparing the data (section 2.0),
feel free to skip to Section 3.0 Initial flags</p>
</div>
<div id="download-ala-data" class="section level2">
<h2>1.1 Download ALA data</h2>
<p>Downloads ALA data and creates a new file in the DataPath to put
those data into.</p>
<pre><code>  ALA_downloader(path = DataPath,
           userEmail = &quot;robert.oreilly@flinders.edu.au&quot;,
           ALA_taxon = &quot;Apiformes&quot;)</code></pre>
</div>
<div id="import-and-merge-ala-scan-idigbio-and-gbif-data"
class="section level2">
<h2>1.2 Import and merge ALA, SCAN, iDigBio and GBIF data</h2>
<p>Supply the path to where the data is, the save_type is either
“csv_files” or “R_file”.</p>
<pre><code>  DataImp &lt;- repo_merge(path = DataPath, save_type = &quot;R_file&quot;)</code></pre>
<p>If there is error finding a file, data_finder() will help
troubleshoot. For example:</p>
<pre><code>            #data_finder(path = DataPath)
            #OUTPUT:
            #$ALA_data
            #[1] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/ALA_galah_path/galah_download_2022-09-15/data.csv&quot;
  
            #$GBIF_data
            #[1] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0000165-220831081235567/occurrence.txt&quot;
            #[2] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436695-210914110416597/occurrence.txt&quot;
            #[3] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436697-210914110416597/occurrence.txt&quot;
            #[4] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436704-210914110416597/occurrence.txt&quot;
            #[5] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436732-210914110416597/occurrence.txt&quot;
            #[6] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436733-210914110416597/occurrence.txt&quot;
            #[7] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/GBIF_webDL_30Aug2022/0436734-210914110416597/occurrence.txt&quot;
                    
            #$iDigBio_data
            #[1] &quot;F:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/iDigBio_webDL_30Aug2022/5aa5abe1-62e0-4d8c-bebf-4ac13bd9e56f/occurrence_raw.csv&quot;
  
            #$SCAN_data
            #character(0)
            #Failing because SCAN_data seems to be missing.Downloaded separetly from the one drive</code></pre>
<p>Load in the most-recent version of these data if needed.This will
return a list with:</p>
<ol style="list-style-type: decimal">
<li>the occurrence dataset with attributes and</li>
<li>the appended eml file</li>
</ol>
<p>if meta_process = FALSE, then it will re-import as-is and be quicker.
If TRUE, it will re-import but also with extra metadata files in the
attributes and a uuid column.</p>
<pre><code>    DataImp &lt;- importOccurrences(path = DataPath,
                           meta_process = FALSE)</code></pre>
<p>Format columns</p>
<pre><code>  DataImp$Data_WebDL &lt;- DataImp$Data_WebDL %&gt;%
    readr::type_convert(col_types = ColTypeR()) </code></pre>
</div>
<div id="import-usgs-data" class="section level2">
<h2>1.3 Import USGS Data</h2>
<p>The USGS_formatter will find, import, format and create metadata for
the USGS dataset. The pubDate must be in day-month-year format.</p>
<pre><code>  USGS_data &lt;- USGS_formatter(path = DataPath, pubDate = &quot;27-04-2022&quot;)</code></pre>
</div>
<div id="formatted-source-importer" class="section level2">
<h2>1.4 Formatted Source Importer</h2>
<p>Formatted source importer. Use this importer to find files that have
been formatted and need to be added to the larger data file. Input is
the:<br />
</p>
<ul>
<li>strings to find and import<br />
- USGS_ &gt; finds the start of the file name, from the
USGS_formatter<br />
- [a-zA-Z_]+ &gt; Finds any words and/or underscores<br />
- [0-9]{4}-[0-9]{2}-[0-9]{2} &gt; Finds a date in the XXXX-XX-XX
format</li>
</ul>
<p><br />
</p>
<p>The attributes file must contain “attribute” in its name, and the
occurrence file must not.</p>
<pre><code>  strings = c(&quot;USGS_[a-zA-Z_]+[0-9]{4}-[0-9]{2}-[0-9]{2}&quot;)</code></pre>
<p>Combine the USGS data and the existing big dataset</p>
<pre><code>  Complete_data &lt;- formatted_combiner(path = DataPath, 
                                      strings = strings, 
                                      # This should be the list-format with eml attached
                                      existingOccurrences = DataImp$Data_WebDL,
                                      existingEMLs = DataImp$eml_files) </code></pre>
<p>In the catalogNumber, remove “.*specimennumber:” as what comes after
should be the USGS number to match for duplicates</p>
<pre><code>  Complete_data$Data_WebDL &lt;- Complete_data$Data_WebDL %&gt;%
    dplyr::mutate(catalogNumber = stringr::str_replace(catalogNumber,
                                                       pattern = &quot;.*\\| specimennumber:&quot;,
                                                       replacement = &quot;&quot;))
  </code></pre>
</div>
<div id="save-data" class="section level2">
<h2>1.5 Save data</h2>
<p>Choose the type of data format you want.</p>
<pre><code>  data_saver(path = DataPath,# The main path to look for data in
             save_type = &quot;CSV_file&quot;, # &quot;R_file&quot; OR &quot;CSV_file&quot;
             occurrences = Complete_data$Data_WebDL, # The existing datasheet
             eml_files = Complete_data$eml_files, # The existing EML files
             file_prefix = &quot;Fin_&quot;) # The prefix for the filenames
  rm(Complete_data, DataImp)</code></pre>
</div>
</div>
<div id="data-preperation" class="section level1">
<h1>2.0 Data preperation</h1>
<div id="standardise-datasets" class="section level2">
<h2>2.1 standardise datasets</h2>
<p>You may either use:</p>
<pre><code>- (a) the bdc import method (works well with general datasets) ***or*** 
- (b) the jbd import method (works well with above data merge)</code></pre>
<div id="a.-bdc-import" class="section level3">
<h3>a. bdc import</h3>
<p>Read in the bdc metadata and standardise the dataset to bdc</p>
<pre><code>        bdc_metadata &lt;- readr::read_csv(paste(DataPath, &quot;out_file&quot;, &quot;bdc_integration.csv&quot;, sep = &quot;/&quot;))
        # ?issue — datasetName is a darwinCore field already!
        # Standardise the dataset to bdc
        db_standardized &lt;- bdc::bdc_standardize_datasets(
          metadata = bdc_metadata,
          format = &quot;csv&quot;,
          overwrite = TRUE,
          save_database = TRUE)
        # read in configuration description file of the column header info
        config_description &lt;- readr::read_csv(paste(DataPath, &quot;Output&quot;, &quot;bdc_configDesc.csv&quot;,
                                                    sep = &quot;/&quot;), 
                                              show_col_types = FALSE, trim_ws = TRUE)
  </code></pre>
</div>
<div id="b.-jbd-import" class="section level3">
<h3>b. jbd import</h3>
<p>You can also just read the data in using the below script. This will
likely be quicker and more-reliable.<br />
</p>
<p>Find the path, read in the file, and add the database_id columnq</p>
<pre><code>  occPath &lt;- file_finder(path = DataPath, fileName = &quot;Fin_BeeData_combined_&quot;)


  db_standardized &lt;- readr::read_csv(occPath, 
                                     # Use the basic ColTypeR function to determine types
                                     col_types = ColTypeR(), trim_ws = TRUE)




   db_standardized &lt;- db_standardized %&gt;%
            dplyr::mutate(
              database_id = paste(&quot;Dorey_data_&quot;, 1:nrow(db_standardized), sep = &quot;&quot;),
              .before = family)
  
   
    </code></pre>
</div>
<div id="c.-match-database_id" class="section level3">
<h3>c. match database_id</h3>
<p>IF you have prior runs from which you’d like to match database_ids
with from the current run. You may use the below script to try to match
database IDs with prior runs.<br />
</p>
<p>read in a prior run of choice</p>
<pre><code>  priorRun &lt;- file_finder(path = DataPath,
                          file = &quot;01_prefilter_database_9Aug22.csv&quot;) %&gt;%
    readr::read_csv(file = ., col_types = ColTypeR())</code></pre>
<p>Which datasets are static and should be excluded from matching?</p>
<pre><code>  excludeDataset &lt;- c(&quot;ASP&quot;, &quot;BMin&quot;, &quot;BMont&quot;, &quot;CAES&quot;, &quot;EaCO&quot;, &quot;Ecd&quot;, &quot;EcoS&quot;,
                      &quot;Gai&quot;, &quot;KP&quot;, &quot;SFU&quot;, &quot;USGS&quot;)</code></pre>
<p>This function will attempt to find the database_ids from prior
runs</p>
<pre><code>  source(paste(ScriptPath, &quot;idMatchR.R&quot;, sep = &quot;/&quot;))
  db_standardized &lt;- idMatchR(
    currentData = db_standardized,
    priorData = priorRun,
    # First matches will be given preference over later ones
    matchBy = tibble::lst(c(&quot;gbifID&quot;),
                          c(&quot;catalogNumber&quot;, &quot;institutionCode&quot;, &quot;dataSource&quot;),
                          c(&quot;occurrenceID&quot;, &quot;dataSource&quot;),
                          c(&quot;recordId&quot;, &quot;dataSource&quot;),
                          c(&quot;id&quot;),
                          # Because INHS was entered as it&#39;s own dataset but is now included in the GBIF download...
                          c(&quot;catalogNumber&quot;, &quot;institutionCode&quot;)),
    # You can exclude datasets from prior by matching their prefixs — before first underscore:
    excludeDataset = excludeDataset)
  </code></pre>
<p>Remove redundant files</p>
<pre><code>  rm(priorRun, excludeDataset)</code></pre>
</div>
<div id="d.-optional-thin" class="section level3">
<h3>d. optional thin</h3>
<p>You can thin the dataset for <strong><em>TESTING
ONLY!</em></strong><br />
</p>
<pre><code>         check_pf &lt;- check_pf %&gt;%
           # take every 100th record
           filter(row_number() %% 100 == 1)</code></pre>
<hr />
</div>
</div>
<div id="integrating-paige-dataset" class="section level2">
<h2>2.2 Integrating Paige dataset</h2>
<p>Paige Chesshire’s cleaned American dataset.</p>
<div id="import-data" class="section level3">
<h3>Import data</h3>
<p>IF you haven’t figured out by now, don’t worry about the column name
warning — not all columns occur here.</p>
<pre><code>  PaigeNAm &lt;- readr::read_csv(paste(DataPath, &quot;Paige_data&quot;, &quot;NorAmer_highQual_only_ALLfamilies.csv&quot;,
                                    sep = &quot;/&quot;), col_types = ColTypeR()) %&gt;%
    # Change the column name from Source to dataSource to match the rest of the data.
    dplyr::rename(dataSource = Source) %&gt;%
    # EXTRACT WAS HERE
    # add a NEW database_id column
    dplyr::mutate(
      database_id = paste0(&quot;Paige_data_&quot;, 1:nrow(.)),
      .before = scientificName)</code></pre>
<p>This is a list of columns by which to match Paige’s data to the
most-recent download with. Each vector will be matched individually.</p>
<pre><code>  columnList &lt;- list(
    c(&quot;decimalLatitude&quot;, &quot;decimalLongitude&quot;, 
      &quot;recordNumber&quot;, &quot;recordedBy&quot;, &quot;individualCount&quot;, &quot;samplingProtocol&quot;,
      &quot;associatedTaxa&quot;, &quot;sex&quot;, &quot;catalogNumber&quot;, &quot;institutionCode&quot;, &quot;otherCatalogNumbers&quot;,
      &quot;recordId&quot;, &quot;occurrenceID&quot;, &quot;collectionID&quot;),         # Iteration 1
    c(&quot;catalogNumber&quot;, &quot;institutionCode&quot;, &quot;otherCatalogNumbers&quot;,
      &quot;recordId&quot;, &quot;occurrenceID&quot;, &quot;collectionID&quot;), # Iteration 2
    c(&quot;decimalLatitude&quot;, &quot;decimalLongitude&quot;, 
      &quot;recordedBy&quot;, &quot;genus&quot;, &quot;specificEpithet&quot;),# Iteration 3
    c(&quot;id&quot;, &quot;decimalLatitude&quot;, &quot;decimalLongitude&quot;),# Iteration 4
    c(&quot;recordedBy&quot;, &quot;genus&quot;, &quot;specificEpithet&quot;, &quot;locality&quot;), # Iteration 5
    c(&quot;recordedBy&quot;, &quot;institutionCode&quot;, &quot;genus&quot;, 
      &quot;specificEpithet&quot;,&quot;locality&quot;),# Iteration 6
    c(&quot;occurrenceID&quot;,&quot;decimalLatitude&quot;, &quot;decimalLongitude&quot;),# Iteration 7
    c(&quot;catalogNumber&quot;,&quot;decimalLatitude&quot;, &quot;decimalLongitude&quot;),# Iteration 8
    c(&quot;catalogNumber&quot;, &quot;locality&quot;) # Iteration 9
  ) </code></pre>
<div class="alert alert-info">
<p><strong> Attention:</strong><br />
It is recommended to run the below code with more than 16GB RAM. Robert
ran this on a laptop with 16GB RAM and an Intel(R) Core(TM) i7-8550U
processor (4 cores and 8 threads) — it struggled.</p>
</div>
</div>
<div id="merge-paiges-data-with-downloaded-data" class="section level3">
<h3>Merge Paige’s data with downloaded data</h3>
<pre><code>  source(paste(ScriptPath, &quot;PaigeIntegrater.R&quot;, sep = &quot;/&quot;))
  db_standardized &lt;- PaigeIntegrater(
    db_standardized = db_standardized,
    PaigeNAm = PaigeNAm,
    columnStrings = columnList)</code></pre>
<p>Remove spent data</p>
<pre><code>  rm(PaigeNAm)</code></pre>
</div>
</div>
<div id="additional-datasets" class="section level2">
<h2>2.3 Additional datasets</h2>
<div id="import-additional-and-potentially-private-datasets."
class="section level3">
<h3>Import additional and potentially private datasets.</h3>
<p><strong>Note:</strong> Private dataset functions are provided but the
data itself is not integrated here until those datasets become freely
available.</p>
<pre><code>  source(paste(ScriptPath, &quot;additionalData_readRs.R&quot;, sep = &quot;/&quot;))</code></pre>
<div id="a.-canada" class="section level4">
<h4>a. Canada</h4>
<pre><code>  SFU_Data &lt;- SFU_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                        inFile = &quot;/InputDatasets/bee_data_canada.csv&quot;,
                        outFile = &quot;jbd_SFU_data.csv&quot;,
                        dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="b.-allan-smith-pardo" class="section level4">
<h4>b. Allan Smith-Pardo</h4>
<pre><code>  ASP_Data &lt;- ASP_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                        inFile = &quot;/InputDatasets/Allan_Smith-Pardo_Dorey_ready2.csv&quot;,
                        outFile = &quot;jbd_ASP_data.csv&quot;,
                        dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="c.-minckley" class="section level4">
<h4>c. Minckley</h4>
<pre><code>  BMin_Data &lt;- BMin_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                          inFile = &quot;/InputDatasets/Bob_Minckley_6_1_22_ScanRecent-mod_Dorey.csv&quot;,
                          outFile = &quot;jbd_BMin_data.csv&quot;,
                          dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="d.-bmont" class="section level4">
<h4>d. BMont</h4>
<pre><code>  BMont_Data &lt;- BMont_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                            inFile = &quot;/InputDatasets/Bombus_Montana_dorey.csv&quot;,
                            outFile = &quot;jbd_BMont_data.csv&quot;,
                            dataLicense = &quot;https://creativecommons.org/licenses/by-sa/4.0/&quot;)</code></pre>
</div>
<div id="e.-ecd" class="section level4">
<h4>e. Ecd</h4>
<pre><code>  Ecd_Data &lt;- Ecd_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                        inFile = &quot;/InputDatasets/Ecdysis_occs.csv&quot;,
                        outFile = &quot;jbd_Ecd_data.csv&quot;,
                        dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="f.-gai" class="section level4">
<h4>f. Gai</h4>
<pre><code>  Gai_Data &lt;- Gai_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                        inFile = &quot;/InputDatasets/upload_to_scan_Gaiarsa et al_Dorey.csv&quot;,
                        outFile = &quot;jbd_Gai_data.csv&quot;,
                        dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="g.-caes" class="section level4">
<h4>g. CAES</h4>
<p>This is a little slower and will have a few date warnings — 215
failed to parse.</p>
<pre><code>  CAES_Data &lt;- CAES_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                          inFile = &quot;/InputDatasets/CT_Bee_Data.csv&quot;,
                          outFile = &quot;jbd_CT_Data.csv&quot;,
                          dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="h.-kp" class="section level4">
<h4>h. KP</h4>
<p>This is a little slower and will have a few date warnings — 215
failed to parse.</p>
<pre><code>   KP_Data &lt;- KP_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                       inFile = &quot;/InputDatasets/KP_output.xlsx&quot;,
                       outFile = &quot;jbd_KP_Data.csv&quot;,
                       dataLicense = &quot;All rights reserved&quot;)</code></pre>
</div>
<div id="i.-inhs" class="section level4">
<h4>i. INHS</h4>
<pre><code>   INHS_Data &lt;- INHS_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                          inFile = &quot;/InputDatasets/INHS_Bees_webreq_DwC-A/occurrences_INHS.csv&quot;,
                           outFile = &quot;occurrences_INHS.csv&quot;,
                           dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="j.-ecos" class="section level4">
<h4>j. EcoS</h4>
<pre><code>  EcoS_Data &lt;- EcoS_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                           inFile = &quot;/InputDatasets/database_ecosur_cecon_shared.csv&quot;,
                           outFile = &quot;jbd_EcoS_Data.csv&quot;,
                           dataLicense = &quot;All rights reserved&quot;)</code></pre>
</div>
<div id="k.-geol" class="section level4">
<h4>k. GeoL</h4>
<pre><code>  GeoL_Data &lt;- GeoL_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                          inFile = &quot;/InputDatasets/Geolocate and BELS_certain and accurate.xlsx&quot;,
                          outFile = &quot;jbd_GeoL_Data.csv&quot;)</code></pre>
</div>
<div id="l.-eaco" class="section level4">
<h4>l. EaCO</h4>
<pre><code>  EaCO_Data &lt;- EaCO_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                          inFile = &quot;/InputDatasets/Eastern Colorado bee 2017 sampling.xlsx&quot;,
                          outFile = &quot;jbd_EaCo_Data.csv&quot;,
                          dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
<div id="m.-mabc" class="section level4">
<h4>m. MABC</h4>
<p>Data from Diego E. de Pedro</p>
<pre><code>  MABC_Data &lt;- MABC_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                           inFile = &quot;/InputDatasets/MABC_dataset.xlsx&quot;,
                           outFile = &quot;jbd_MABC_Data.csv&quot;,
                           dataLicense = &quot;All rights reserved&quot;)</code></pre>
</div>
<div id="n.-arm" class="section level4">
<h4>n. Arm</h4>
<pre><code>  Arm_Data &lt;- Arm_readr(path = paste0(DataPath, &quot;/Additional_Datasets&quot;),
                        inFile = &quot;/InputDatasets/Bee database Armando_Final.xlsx&quot;,
                        outFile = &quot;jbd_Arm_Data.csv&quot;,
                        sheet = &quot;Sheet1&quot;,
                        dataLicense = &quot;https://creativecommons.org/licenses/by-nc-sa/4.0/&quot;)</code></pre>
</div>
</div>
<div id="merge-all" class="section level3">
<h3>Merge all</h3>
<p>Remove these spent datasets</p>
<pre><code>  rm(SFU_Data, ASP_Data, BMin_Data, BMont_Data, Ecd_Data, Gai_Data, CAES_Data, 
      INHS_Data, GeoL_Data, EaCO_Data, Arm_Data)</code></pre>
<p>Read in and merge all. Hashed out datasets mean they were not
included as they are currently private.</p>
<pre><code>  db_standardized &lt;- db_standardized %&gt;%
    dplyr::bind_rows(
      
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_ASP_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_SFU_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_BMin_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_BMont_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_Ecd_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_Gai_data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_CT_Data.csv&quot;), col_types = ColTypeR()),
       #readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                              #&quot;/jbd_KP_Data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                              &quot;/jbd_INHS_Data.csv&quot;), col_types = ColTypeR()),
       #readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                              #&quot;/jbd_EcoS_Data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_GeoL_Data.csv&quot;), col_types = ColTypeR()),
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_EaCo_Data.csv&quot;), col_types = ColTypeR()), 
       #readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                              #&quot;/jbd_MABC_Data.csv&quot;), col_types = ColTypeR())
      readr::read_csv(paste0(DataPath, &quot;/Additional_Datasets&quot;, 
                             &quot;/jbd_Arm_Data.csv&quot;), col_types = ColTypeR())) %&gt;% 
     # END bind_rows
    suppressWarnings(classes = &quot;warning&quot;) # End suppressWarnings — due to col_types</code></pre>
<p>Save the dataset</p>
<pre><code>  db_standardized %&gt;%
    readr::write_csv(.,
                     paste(DataPath, &quot;Output&quot;, &quot;Intermediate&quot;, &quot;00_prefilter_database.csv&quot;,
                           sep = &quot;/&quot;))</code></pre>
</div>
</div>
</div>
<div id="initial-flags" class="section level1">
<h1>3.0 Initial flags</h1>
<p>Read data back in if needed</p>
<div class="alert alert-info">
<p><strong> Attention:</strong><br />
As the analysis gets computationally heavy further on, I will filter the
Check_pf object to 1% by grabbing every 100th row.<br />
</p>
<p>With the below code, you can thin the dataset for <strong><em>TESTING
ONLY!</em></strong></p>
</div>
<pre><code>    db_standardized &lt;- readr::read_csv(paste(DataPath, &quot;Output&quot;, &quot;Intermediate&quot;, &quot;00_prefilter_database.csv&quot;, sep = &quot;/&quot;), col_types = ColTypeR())


  check_pf &lt;- bdc::bdc_scientificName_empty(
    data = db_standardized,
    sci_name = &quot;scientificName&quot;)</code></pre>
<p>I did not read the data back in. Instead, I filtered data to 1%.</p>
<pre><code>  check_pf &lt;- db_standardized %&gt;%
               # take every 100th record
               filter(row_number() %% 100 == 1)



  check_pf %&gt;%
          readr::write_csv(.,
                           paste(DataPath, &quot;Output&quot;, &quot;Intermediate&quot;, &quot;check_pf.csv&quot;,
                                 sep = &quot;/&quot;))</code></pre>
<p>See <a
href="https://brunobrr.github.io/bdc/articles/prefilter.html">here</a>
for bdc tutorial.</p>
<div id="sciname" class="section level2">
<h2>3.1 SciName</h2>
<p>Flag occurrences without scientificName provided</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;check_pf.csv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>), <span class="at">col_types =</span> <span class="fu">ColTypeR</span>())</span></code></pre></div>
<p>now that this is saved, remove it to save space in memory</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="co">#rm(db_standardized) </span></span></code></pre></div>
</div>
<div id="misscoords" class="section level2">
<h2>3.2 MissCoords</h2>
<p>Flag occurrences with missing lat and lon</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_coordinates_empty</span>(</span>
<span id="cb68-2"><a href="#cb68-2" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb68-3"><a href="#cb68-3" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb68-4"><a href="#cb68-4" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>)</span></code></pre></div>
<pre><code>## 
## bdc_coordinates_empty:
## Flagged 19054 records.
## One column was added to the database.</code></pre>
</div>
<div id="outofrange" class="section level2">
<h2>3.3 OutOfRange</h2>
<p>Flag occurrences that are not on earth (outside of -180 to 180 or -90
to 90 degrees)</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_coordinates_outOfRange</span>(</span>
<span id="cb70-2"><a href="#cb70-2" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb70-3"><a href="#cb70-3" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb70-4"><a href="#cb70-4" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>)</span></code></pre></div>
<pre><code>## 
## bdc_coordinates_outOfRange:
## Flagged 0 records.
## One column was added to the database.</code></pre>
</div>
<div id="source" class="section level2">
<h2>3.4 Source</h2>
<p>Flag occurrences that don’t match the basisOfRecord types below</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_basisOfRecords_notStandard</span>(</span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a>  <span class="at">basisOfRecord =</span> <span class="st">&quot;basisOfRecord&quot;</span>,</span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a>  <span class="at">names_to_keep =</span> <span class="fu">c</span>(</span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a>    <span class="co"># Keep all plus some at the bottom.</span></span>
<span id="cb72-6"><a href="#cb72-6" tabindex="-1"></a>    <span class="st">&quot;Event&quot;</span>,</span>
<span id="cb72-7"><a href="#cb72-7" tabindex="-1"></a>    <span class="st">&quot;HUMAN_OBSERVATION&quot;</span>,</span>
<span id="cb72-8"><a href="#cb72-8" tabindex="-1"></a>    <span class="st">&quot;HumanObservation&quot;</span>,</span>
<span id="cb72-9"><a href="#cb72-9" tabindex="-1"></a>    <span class="st">&quot;LIVING_SPECIMEN&quot;</span>,</span>
<span id="cb72-10"><a href="#cb72-10" tabindex="-1"></a>    <span class="st">&quot;LivingSpecimen&quot;</span>,</span>
<span id="cb72-11"><a href="#cb72-11" tabindex="-1"></a>    <span class="st">&quot;MACHINE_OBSERVATION&quot;</span>,</span>
<span id="cb72-12"><a href="#cb72-12" tabindex="-1"></a>    <span class="st">&quot;MachineObservation&quot;</span>,</span>
<span id="cb72-13"><a href="#cb72-13" tabindex="-1"></a>    <span class="st">&quot;MATERIAL_SAMPLE&quot;</span>,</span>
<span id="cb72-14"><a href="#cb72-14" tabindex="-1"></a>    <span class="st">&quot;O&quot;</span>,</span>
<span id="cb72-15"><a href="#cb72-15" tabindex="-1"></a>    <span class="st">&quot;Occurrence&quot;</span>,</span>
<span id="cb72-16"><a href="#cb72-16" tabindex="-1"></a>    <span class="st">&quot;MaterialSample&quot;</span>,</span>
<span id="cb72-17"><a href="#cb72-17" tabindex="-1"></a>    <span class="st">&quot;OBSERVATION&quot;</span>,</span>
<span id="cb72-18"><a href="#cb72-18" tabindex="-1"></a>    <span class="st">&quot;Preserved Specimen&quot;</span>,</span>
<span id="cb72-19"><a href="#cb72-19" tabindex="-1"></a>    <span class="st">&quot;PRESERVED_SPECIMEN&quot;</span>,</span>
<span id="cb72-20"><a href="#cb72-20" tabindex="-1"></a>    <span class="st">&quot;preservedspecimen Specimen&quot;</span>,</span>
<span id="cb72-21"><a href="#cb72-21" tabindex="-1"></a>    <span class="st">&quot;Preservedspecimen&quot;</span>,</span>
<span id="cb72-22"><a href="#cb72-22" tabindex="-1"></a>    <span class="st">&quot;PreservedSpecimen&quot;</span>,</span>
<span id="cb72-23"><a href="#cb72-23" tabindex="-1"></a>    <span class="st">&quot;preservedspecimen&quot;</span>,</span>
<span id="cb72-24"><a href="#cb72-24" tabindex="-1"></a>    <span class="st">&quot;S&quot;</span>,</span>
<span id="cb72-25"><a href="#cb72-25" tabindex="-1"></a>    <span class="st">&quot;Specimen&quot;</span>,</span>
<span id="cb72-26"><a href="#cb72-26" tabindex="-1"></a>    <span class="st">&quot;Taxon&quot;</span>,</span>
<span id="cb72-27"><a href="#cb72-27" tabindex="-1"></a>    <span class="st">&quot;UNKNOWN&quot;</span>,</span>
<span id="cb72-28"><a href="#cb72-28" tabindex="-1"></a>    <span class="st">&quot;&quot;</span>,</span>
<span id="cb72-29"><a href="#cb72-29" tabindex="-1"></a>    <span class="cn">NA</span>,</span>
<span id="cb72-30"><a href="#cb72-30" tabindex="-1"></a>    <span class="st">&quot;NA&quot;</span>,</span>
<span id="cb72-31"><a href="#cb72-31" tabindex="-1"></a>    <span class="st">&quot;LITERATURE&quot;</span>, </span>
<span id="cb72-32"><a href="#cb72-32" tabindex="-1"></a>    <span class="st">&quot;None&quot;</span>, <span class="st">&quot;Pinned Specimen&quot;</span>, <span class="st">&quot;Voucher reared&quot;</span>, <span class="st">&quot;Emerged specimen&quot;</span></span>
<span id="cb72-33"><a href="#cb72-33" tabindex="-1"></a>  ))</span></code></pre></div>
<pre><code>## 
## bdc_basisOfRecords_notStandard:
## Flagged 298 of the following specific nature:
##  c(7, 8, 18) 
## One column was added to the database.</code></pre>
</div>
<div id="countryname" class="section level2">
<h2>3.5 CountryName</h2>
<p>Try to harmonise country names.<br />
</p>
<p>Create a Tibble of common issues in country names and their
replacements</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a>commonProblems <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">problem =</span> <span class="fu">c</span>(<span class="st">&#39;U.S.A.&#39;</span>, <span class="st">&#39;US&#39;</span>,<span class="st">&#39;USA&#39;</span>,<span class="st">&#39;usa&#39;</span>,<span class="st">&#39;UNITED STATES&#39;</span>,<span class="st">&#39;United States&#39;</span>,<span class="st">&#39;U.S.A&#39;</span>,<span class="st">&#39;MX&#39;</span>,<span class="st">&#39;CA&#39;</span>,<span class="st">&#39;Bras.&#39;</span>,<span class="st">&#39;Braz.&#39;</span>,<span class="st">&#39;Brasil&#39;</span>,<span class="st">&#39;CNMI&#39;</span>,<span class="st">&#39;USA TERRITORY: PUERTO RICO&#39;</span>),</span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a>                                 <span class="at">fix =</span> <span class="fu">c</span>(<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;United States of America&#39;</span>,<span class="st">&#39;Mexico&#39;</span>,<span class="st">&#39;Canada&#39;</span>,<span class="st">&#39;Brazil&#39;</span>,<span class="st">&#39;Brazil&#39;</span>,<span class="st">&#39;Brazil&#39;</span>,<span class="st">&#39;Northern Mariana Islands&#39;</span>,<span class="st">&#39;PUERTO.RICO&#39;</span>))</span></code></pre></div>
<div id="a.-prepare-dataset" class="section level3">
<h3>a. prepare dataset</h3>
<p>Fix up country names based on common problems above and extract ISO2
codes for occurrences</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_countryNameCleanR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb75-2"><a href="#cb75-2" tabindex="-1"></a>check_pf_noNa <span class="ot">&lt;-</span> <span class="fu">jbd_countryNameCleanR</span>(</span>
<span id="cb75-3"><a href="#cb75-3" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb75-4"><a href="#cb75-4" tabindex="-1"></a>  <span class="at">commonProblems =</span> commonProblems)</span></code></pre></div>
<pre><code>##  — Using default country names and codes from https:en.wikipedia.org/wiki/ISO_3166-1_alpha-2 — static version from July 2022.</code></pre>
</div>
<div id="b.-run-function" class="section level3">
<h3>b. run function</h3>
<p>Get country name from coordinates using a wrapper around the
bdc_country_from_coordinates function. Because our dataset is much
larger than those used to design bdc, we have made it so that you can
analyse data in smaller pieces.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_CfC_chunker.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a><span class="fu">suppressWarnings</span>(</span>
<span id="cb77-3"><a href="#cb77-3" tabindex="-1"></a>  countryOutput <span class="ot">&lt;-</span> <span class="fu">jbd_CfC_chunker</span>(<span class="at">data =</span> check_pf_noNa,</span>
<span id="cb77-4"><a href="#cb77-4" tabindex="-1"></a>                                   <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb77-5"><a href="#cb77-5" tabindex="-1"></a>                                   <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb77-6"><a href="#cb77-6" tabindex="-1"></a>                                   <span class="at">country =</span> <span class="st">&quot;country&quot;</span>,</span>
<span id="cb77-7"><a href="#cb77-7" tabindex="-1"></a>                                   <span class="co"># How many rows to process at a time</span></span>
<span id="cb77-8"><a href="#cb77-8" tabindex="-1"></a>                                   <span class="at">stepSize =</span> <span class="dv">1000000</span>,</span>
<span id="cb77-9"><a href="#cb77-9" tabindex="-1"></a>                                   <span class="co"># Start row</span></span>
<span id="cb77-10"><a href="#cb77-10" tabindex="-1"></a>                                   <span class="at">chunkStart =</span> <span class="dv">1</span>,</span>
<span id="cb77-11"><a href="#cb77-11" tabindex="-1"></a>                                   <span class="at">append =</span> <span class="cn">FALSE</span>),</span>
<span id="cb77-12"><a href="#cb77-12" tabindex="-1"></a>  <span class="at">classes =</span> <span class="st">&quot;warning&quot;</span>)</span></code></pre></div>
<pre><code>##  — Starting chunk 1...
## From 1 to 1,000,000
##  — Cleaning RAM.
##  — Finished chunk 1 of 1 chunks. Records examined: 1,000,000</code></pre>
</div>
<div id="c.-re-merge" class="section level3">
<h3>c. re-merge</h3>
<p>Left join these datasets</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">left_join</span>(check_pf, countryOutput, <span class="at">by =</span> <span class="st">&quot;database_id&quot;</span>)  <span class="sc">%&gt;%</span> </span>
<span id="cb79-2"><a href="#cb79-2" tabindex="-1"></a>  <span class="co"># merge the two country name columns into the &quot;country&quot; column</span></span>
<span id="cb79-3"><a href="#cb79-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">country =</span> dplyr<span class="sc">::</span><span class="fu">coalesce</span>(country.x, country.y)) <span class="sc">%&gt;%</span></span>
<span id="cb79-4"><a href="#cb79-4" tabindex="-1"></a>  <span class="co"># remove the now redundant country columns </span></span>
<span id="cb79-5"><a href="#cb79-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(country.x, country.y)) <span class="sc">%&gt;%</span></span>
<span id="cb79-6"><a href="#cb79-6" tabindex="-1"></a>  <span class="co"># put the column back in its original place</span></span>
<span id="cb79-7"><a href="#cb79-7" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">relocate</span>(country, <span class="at">.after =</span> basisOfRecord) <span class="sc">%&gt;%</span> </span>
<span id="cb79-8"><a href="#cb79-8" tabindex="-1"></a>  <span class="co"># Remove duplicates if they arose!</span></span>
<span id="cb79-9"><a href="#cb79-9" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span></code></pre></div>
<p>Remove illegal characters</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a>check_pf<span class="sc">$</span>country <span class="ot">&lt;-</span> check_pf<span class="sc">$</span>country <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a href="#cb80-2" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_replace</span>(., <span class="at">pattern =</span> <span class="fu">paste</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">[&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">]&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">?&quot;</span>,</span>
<span id="cb80-3"><a href="#cb80-3" tabindex="-1"></a>                                          <span class="at">sep=</span>  <span class="st">&quot;|&quot;</span>), <span class="at">replacement =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb80-4"><a href="#cb80-4" tabindex="-1"></a><span class="co"># Replace the problems as they occur</span></span>
<span id="cb80-5"><a href="#cb80-5" tabindex="-1"></a></span>
<span id="cb80-6"><a href="#cb80-6" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> check_pf <span class="sc">%&gt;%</span></span>
<span id="cb80-7"><a href="#cb80-7" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(commonProblems, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;country&quot;</span> <span class="ot">=</span> <span class="st">&quot;problem&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb80-8"><a href="#cb80-8" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">country =</span> </span>
<span id="cb80-9"><a href="#cb80-9" tabindex="-1"></a>                  dplyr<span class="sc">::</span><span class="fu">if_else</span>(country <span class="sc">%in%</span> commonProblems<span class="sc">$</span>problem,</span>
<span id="cb80-10"><a href="#cb80-10" tabindex="-1"></a>                                 fix, country)) <span class="sc">%&gt;%</span></span>
<span id="cb80-11"><a href="#cb80-11" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>fix)</span></code></pre></div>
<p>Save the dataset</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a>check_pf <span class="sc">%&gt;%</span></span>
<span id="cb81-2"><a href="#cb81-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb81-3"><a href="#cb81-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;01_prefilter_database.csv&quot;</span>,</span>
<span id="cb81-4"><a href="#cb81-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Read in <em>if</em> needed<br />
</p>
<pre><code>   check_pf &lt;- readr::read_csv(paste(DataPath, 
                 &quot;Output&quot;, &quot;Intermediate&quot;, &quot;01_prefilter_database.csv&quot;, sep = &quot;/&quot;),
                 col_types = ColTypeR())</code></pre>
<p>remove the interim datasets</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="fu">rm</span>(check_pf_noNa, countryOutput)</span></code></pre></div>
</div>
</div>
<div id="standardconames" class="section level2">
<h2>3.6 StandardCoNames</h2>
<p>Remove the countryCode and country_suggested columns to avoid an
error with bdc_country_standardized where two “countryCode” columns
exist.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> check_pf <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>tidyselect<span class="sc">::</span><span class="fu">any_of</span>(<span class="fu">c</span>(<span class="st">&quot;countryCode&quot;</span>, <span class="st">&quot;country_suggested&quot;</span>))) </span></code></pre></div>
<p>Run the function (Standardise country names and add ISO2 codes if
needed).</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_country_standardized</span>(</span>
<span id="cb85-2"><a href="#cb85-2" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb85-3"><a href="#cb85-3" tabindex="-1"></a>  <span class="at">country =</span> <span class="st">&quot;country&quot;</span></span>
<span id="cb85-4"><a href="#cb85-4" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb85-5"><a href="#cb85-5" tabindex="-1"></a>  <span class="co"># Put country_suggested into the country column (otherwise this column is a useless mess)</span></span>
<span id="cb85-6"><a href="#cb85-6" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">country =</span> country_suggested)</span></code></pre></div>
<pre><code>## Loading auxiliary data: country names</code></pre>
<pre><code>## Standardizing country names</code></pre>
<pre><code>## country found: Afghanistan</code></pre>
<pre><code>## country found: Åland Islands</code></pre>
<pre><code>## country found: Albania</code></pre>
<pre><code>## country found: Algeria</code></pre>
<pre><code>## country found: American Samoa</code></pre>
<pre><code>## country found: Andorra</code></pre>
<pre><code>## country found: Angola</code></pre>
<pre><code>## country found: Antarctica</code></pre>
<pre><code>## country found: Argentina</code></pre>
<pre><code>## country found: Armenia</code></pre>
<pre><code>## country found: Australia</code></pre>
<pre><code>## country found: Austria</code></pre>
<pre><code>## country found: Bahamas</code></pre>
<pre><code>## country found: Bangladesh</code></pre>
<pre><code>## country found: Belarus</code></pre>
<pre><code>## country found: Belgium</code></pre>
<pre><code>## country found: Belize</code></pre>
<pre><code>## country found: Benin</code></pre>
<pre><code>## country found: Bhutan</code></pre>
<pre><code>## country found: Bolivia</code></pre>
<pre><code>## country found: Botswana</code></pre>
<pre><code>## country found: Brazil</code></pre>
<pre><code>## country found: Bulgaria</code></pre>
<pre><code>## country found: Burkina Faso</code></pre>
<pre><code>## country found: Burundi</code></pre>
<pre><code>## country found: Cameroon</code></pre>
<pre><code>## country found: Canada</code></pre>
<pre><code>## country found: Central African Republic</code></pre>
<pre><code>## country found: Chile</code></pre>
<pre><code>## country found: China</code></pre>
<pre><code>## country found: Colombia</code></pre>
<pre><code>## country found: Republic of the Congo</code></pre>
<pre><code>## country found: Cook Islands</code></pre>
<pre><code>## country found: Costa Rica</code></pre>
<pre><code>## country found: Croatia</code></pre>
<pre><code>## country found: Cuba</code></pre>
<pre><code>## country found: Cyprus</code></pre>
<pre><code>## country found: Czech Republic</code></pre>
<pre><code>## country found: Denmark</code></pre>
<pre><code>## country found: Dominican Republic</code></pre>
<pre><code>## country found: Ecuador</code></pre>
<pre><code>## country found: Egypt</code></pre>
<pre><code>## country found: El Salvador</code></pre>
<pre><code>## country found: Estonia</code></pre>
<pre><code>## country found: Ethiopia</code></pre>
<pre><code>## country found: Fiji</code></pre>
<pre><code>## country found: Finland</code></pre>
<pre><code>## country found: France</code></pre>
<pre><code>## country found: French Guiana</code></pre>
<pre><code>## country found: Gabon</code></pre>
<pre><code>## country found: Gambia</code></pre>
<pre><code>## country found: Georgia</code></pre>
<pre><code>## country found: Germany</code></pre>
<pre><code>## country found: Ghana</code></pre>
<pre><code>## country found: Greece</code></pre>
<pre><code>## country found: Greenland</code></pre>
<pre><code>## country found: Guadeloupe</code></pre>
<pre><code>## country found: Guam</code></pre>
<pre><code>## country found: Guatemala</code></pre>
<pre><code>## country found: Guyana</code></pre>
<pre><code>## country found: Haiti</code></pre>
<pre><code>## country found: Honduras</code></pre>
<pre><code>## country found: Hong Kong</code></pre>
<pre><code>## country found: Hungary</code></pre>
<pre><code>## country found: Iceland</code></pre>
<pre><code>## country found: India</code></pre>
<pre><code>## country found: Indonesia</code></pre>
<pre><code>## country found: Iran</code></pre>
<pre><code>## country found: Iraq</code></pre>
<pre><code>## country found: Ireland</code></pre>
<pre><code>## country found: Israel</code></pre>
<pre><code>## country found: Italy</code></pre>
<pre><code>## country found: Jamaica</code></pre>
<pre><code>## country found: Japan</code></pre>
<pre><code>## country found: Jersey</code></pre>
<pre><code>## country found: Jordan</code></pre>
<pre><code>## country found: Kazakhstan</code></pre>
<pre><code>## country found: Kenya</code></pre>
<pre><code>## country found: Kuwait</code></pre>
<pre><code>## country found: Kyrgyzstan</code></pre>
<pre><code>## country found: Latvia</code></pre>
<pre><code>## country found: Lebanon</code></pre>
<pre><code>## country found: Lesotho</code></pre>
<pre><code>## country found: Liberia</code></pre>
<pre><code>## country found: Libya</code></pre>
<pre><code>## country found: Liechtenstein</code></pre>
<pre><code>## country found: Lithuania</code></pre>
<pre><code>## country found: Luxembourg</code></pre>
<pre><code>## country found: Macao</code></pre>
<pre><code>## country found: Madagascar</code></pre>
<pre><code>## country found: Malawi</code></pre>
<pre><code>## country found: Malaysia</code></pre>
<pre><code>## country found: Maldives</code></pre>
<pre><code>## country found: Malta</code></pre>
<pre><code>## country found: Martinique</code></pre>
<pre><code>## country found: Mexico</code></pre>
<pre><code>## country found: Mongolia</code></pre>
<pre><code>## country found: Montenegro</code></pre>
<pre><code>## country found: Montserrat</code></pre>
<pre><code>## country found: Morocco</code></pre>
<pre><code>## country found: Mozambique</code></pre>
<pre><code>## country found: Myanmar</code></pre>
<pre><code>## country found: Namibia</code></pre>
<pre><code>## country found: Nepal</code></pre>
<pre><code>## country found: Netherlands</code></pre>
<pre><code>## country found: New Caledonia</code></pre>
<pre><code>## country found: New Zealand</code></pre>
<pre><code>## country found: Nicaragua</code></pre>
<pre><code>## country found: Nigeria</code></pre>
<pre><code>## country found: Northern Mariana Islands</code></pre>
<pre><code>## country found: Norway</code></pre>
<pre><code>## country found: Oman</code></pre>
<pre><code>## country found: Pakistan</code></pre>
<pre><code>## country found: Palestine</code></pre>
<pre><code>## country found: Panama</code></pre>
<pre><code>## country found: Papua New Guinea</code></pre>
<pre><code>## country found: Paraguay</code></pre>
<pre><code>## country found: Peru</code></pre>
<pre><code>## country found: Philippines</code></pre>
<pre><code>## country found: Poland</code></pre>
<pre><code>## country found: Portugal</code></pre>
<pre><code>## country found: Puerto Rico</code></pre>
<pre><code>## country found: Qatar</code></pre>
<pre><code>## country found: Romania</code></pre>
<pre><code>## country found: Russian Federation</code></pre>
<pre><code>## country found: Saint Pierre and Miquelon</code></pre>
<pre><code>## country found: Saint Vincent and the Grenadines</code></pre>
<pre><code>## country found: Samoa</code></pre>
<pre><code>## country found: Saudi Arabia</code></pre>
<pre><code>## country found: Serbia</code></pre>
<pre><code>## country found: Singapore</code></pre>
<pre><code>## country found: Slovakia</code></pre>
<pre><code>## country found: Slovenia</code></pre>
<pre><code>## country found: Solomon Islands</code></pre>
<pre><code>## country found: Somalia</code></pre>
<pre><code>## country found: South Africa</code></pre>
<pre><code>## country found: South Georgia and the South Sandwich Islands</code></pre>
<pre><code>## country found: Spain</code></pre>
<pre><code>## country found: Sri Lanka</code></pre>
<pre><code>## country found: Sudan</code></pre>
<pre><code>## country found: Suriname</code></pre>
<pre><code>## country found: Svalbard and Jan Mayen</code></pre>
<pre><code>## country found: Sweden</code></pre>
<pre><code>## country found: Switzerland</code></pre>
<pre><code>## country found: Syria</code></pre>
<pre><code>## country found: Taiwan</code></pre>
<pre><code>## country found: Tajikistan</code></pre>
<pre><code>## country found: Tanzania</code></pre>
<pre><code>## country found: Thailand</code></pre>
<pre><code>## country found: Trinidad and Tobago</code></pre>
<pre><code>## country found: Tunisia</code></pre>
<pre><code>## country found: Turkmenistan</code></pre>
<pre><code>## country found: Uganda</code></pre>
<pre><code>## country found: Ukraine</code></pre>
<pre><code>## country found: United Arab Emirates</code></pre>
<pre><code>## country found: United Kingdom</code></pre>
<pre><code>## country found: United States of America</code></pre>
<pre><code>## country found: Uruguay</code></pre>
<pre><code>## country found: Uzbekistan</code></pre>
<pre><code>## country found: Vanuatu</code></pre>
<pre><code>## country found: Venezuela</code></pre>
<pre><code>## country found: Vietnam</code></pre>
<pre><code>## country found: Yemen</code></pre>
<pre><code>## country found: Zambia</code></pre>
<pre><code>## country found: Zimbabwe</code></pre>
<pre><code>## 
## bdc_country_standardized:
## The country names of 9969 records were standardized.
## Two columns (&#39;country_suggested&#39; and &#39;countryCode&#39;) were added to the database.</code></pre>
</div>
<div id="transpcoords" class="section level2">
<h2>3.7 TranspCoords</h2>
<p>Flag and correct records when lat and long appear to be transposed.
We have chunked this because it is too RAM-heavy to run on our large
dataset.</p>
<div class="sourceCode" id="cb254"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="#cb254-1" tabindex="-1"></a><span class="fu">suppressWarnings</span>(</span>
<span id="cb254-2"><a href="#cb254-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">jbd_Ctrans_chunker</span>(</span>
<span id="cb254-3"><a href="#cb254-3" tabindex="-1"></a>  <span class="co"># bdc_coordinates_transposed inputs</span></span>
<span id="cb254-4"><a href="#cb254-4" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb254-5"><a href="#cb254-5" tabindex="-1"></a>  <span class="at">id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb254-6"><a href="#cb254-6" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb254-7"><a href="#cb254-7" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb254-8"><a href="#cb254-8" tabindex="-1"></a>  <span class="at">country =</span> <span class="st">&quot;country&quot;</span>,</span>
<span id="cb254-9"><a href="#cb254-9" tabindex="-1"></a>  <span class="at">countryCode =</span> <span class="st">&quot;countryCode&quot;</span>,</span>
<span id="cb254-10"><a href="#cb254-10" tabindex="-1"></a>  <span class="at">border_buffer =</span> <span class="fl">0.2</span>, <span class="co"># in decimal degrees (~22 km at the equator)</span></span>
<span id="cb254-11"><a href="#cb254-11" tabindex="-1"></a>  <span class="at">save_outputs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb254-12"><a href="#cb254-12" tabindex="-1"></a>  <span class="at">sci_names =</span> <span class="st">&quot;scientificName&quot;</span>,</span>
<span id="cb254-13"><a href="#cb254-13" tabindex="-1"></a>  <span class="co"># chunker inputs</span></span>
<span id="cb254-14"><a href="#cb254-14" tabindex="-1"></a>  <span class="at">stepSize =</span> <span class="dv">1000000</span>,  <span class="co"># How many rows to process at a time</span></span>
<span id="cb254-15"><a href="#cb254-15" tabindex="-1"></a>  <span class="at">chunkStart =</span> <span class="dv">1</span>,  <span class="co"># Start row</span></span>
<span id="cb254-16"><a href="#cb254-16" tabindex="-1"></a>  <span class="at">append =</span> <span class="cn">FALSE</span>  <span class="co"># If FALSE it may overwrite existing dataset</span></span>
<span id="cb254-17"><a href="#cb254-17" tabindex="-1"></a>) , <span class="at">classes =</span> <span class="st">&quot;warning&quot;</span>)</span></code></pre></div>
<pre><code>##  — Running chunker with:
## stepSize = 1,000,000
## chunkStart = 1
## chunkEnd = 1,000,000
## append = FALSE
##  — Starting chunk 1...
## From 1 to 1,000,000
##  — Cleaning RAM.
##  — Finished chunk 1 with 1 remaining. Records examined: 1,000,000</code></pre>
<p>Get a summary of the number of transposed records.</p>
<div class="sourceCode" id="cb256"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="#cb256-1" tabindex="-1"></a><span class="fu">table</span>(check_pf<span class="sc">$</span>coordinates_transposed, <span class="at">useNA =</span> <span class="st">&quot;always&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  FALSE   TRUE   &lt;NA&gt; 
##     32 153104      0</code></pre>
<p>Save the dataset</p>
<div class="sourceCode" id="cb258"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="#cb258-1" tabindex="-1"></a>check_pf <span class="sc">%&gt;%</span></span>
<span id="cb258-2"><a href="#cb258-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb258-3"><a href="#cb258-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;01_prefilter_database.csv&quot;</span>,</span>
<span id="cb258-4"><a href="#cb258-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb258-5"><a href="#cb258-5" tabindex="-1"></a><span class="fu">gc</span>()</span></code></pre></div>
<pre><code>##            used  (Mb) gc trigger   (Mb)  max used   (Mb)
## Ncells  4953113 264.6   15180695  810.8  23719834 1266.8
## Vcells 26115760 199.3  174785796 1333.6 218482236 1666.9</code></pre>
<p>Read data in again if needed<br />
</p>
<pre><code>  check_pf &lt;- readr::read_csv(paste(DataPath,&quot;Output&quot;, &quot;Intermediate&quot;, &quot;01_prefilter_database.csv&quot;,
                                   sep = &quot;/&quot;))</code></pre>
</div>
<div id="coord-country" class="section level2">
<h2>3.8 Coord-country</h2>
<p>Collect all country names in the country_suggested column. We rebuilt
a bdc function to flag occurrences where the coordinates are
inconsistent with the provided country name.</p>
<div class="sourceCode" id="cb261"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="#cb261-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_coordCountryInconsistent.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb261-2"><a href="#cb261-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">jbd_coordCountryInconsistent</span>(</span>
<span id="cb261-3"><a href="#cb261-3" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb261-4"><a href="#cb261-4" tabindex="-1"></a>  <span class="at">path =</span> DataPath,</span>
<span id="cb261-5"><a href="#cb261-5" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb261-6"><a href="#cb261-6" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb261-7"><a href="#cb261-7" tabindex="-1"></a>  <span class="at">mapResolution =</span> <span class="dv">50</span>,</span>
<span id="cb261-8"><a href="#cb261-8" tabindex="-1"></a>  <span class="at">pointBuffer =</span> <span class="fl">0.01</span>)</span></code></pre></div>
<pre><code>##  — Downloading naturalearth map...</code></pre>
<pre><code>## Spherical geometry (s2) switched off</code></pre>
<pre><code>##  — Extracting initial country names without buffer...</code></pre>
<pre><code>## although coordinates are longitude/latitude, st_intersection assumes that they
## are planar</code></pre>
<pre><code>##  — Buffering natualearth map by pointBuffer...</code></pre>
<pre><code>## dist is assumed to be in decimal degrees (arc_degrees).</code></pre>
<pre><code>##  — Extracting FAILED country names WITH buffer...</code></pre>
<pre><code>## although coordinates are longitude/latitude, st_intersection assumes that they
## are planar</code></pre>
<pre><code>##  — Saving the unique failed country matches as  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/BeeDC-website/BeeDCtes2/Output/Check/failed_unique_CCinconsistent.csv</code></pre>
<pre><code>## 
## bdc_coordinates_country_inconsistent:
## Flagged 97 records.
## The column, &#39;.coordinates_country_inconsistent&#39;, was added to the database.</code></pre>
<pre><code>##  — Completed in 4.38 minutes.</code></pre>
<p>Save the dataset</p>
<div class="sourceCode" id="cb273"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="#cb273-1" tabindex="-1"></a>check_pf <span class="sc">%&gt;%</span></span>
<span id="cb273-2"><a href="#cb273-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb273-3"><a href="#cb273-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;01_prefilter_database.csv&quot;</span>,</span>
<span id="cb273-4"><a href="#cb273-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
</div>
<div id="georefissue" class="section level2">
<h2>3.9 GeoRefIssue</h2>
<p>This function Identifies records whose coordinates can potentially be
extracted from locality information (must be manually checked
later).</p>
<div class="sourceCode" id="cb274"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="#cb274-1" tabindex="-1"></a>xyFromLocality <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_coordinates_from_locality</span>(</span>
<span id="cb274-2"><a href="#cb274-2" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb274-3"><a href="#cb274-3" tabindex="-1"></a>  <span class="at">locality =</span> <span class="st">&quot;locality&quot;</span>,</span>
<span id="cb274-4"><a href="#cb274-4" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb274-5"><a href="#cb274-5" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb274-6"><a href="#cb274-6" tabindex="-1"></a>  <span class="at">save_outputs =</span> <span class="cn">TRUE</span></span>
<span id="cb274-7"><a href="#cb274-7" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## 
## bdc_coordinates_from_locality 
## Found 15238 records missing or with invalid coordinates but with potentially useful information on locality.
##  
## Check database in: Output/Check/01_coordinates_from_locality.csv</code></pre>
<pre><code>## 
## bdc_coordinates_from_locality 
## Found 15238 records missing or with invalid coordinates but with potentially useful information on locality.</code></pre>
<p>Save data if needed.</p>
<div class="sourceCode" id="cb277"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="#cb277-1" tabindex="-1"></a>xyFromLocality <span class="sc">%&gt;%</span></span>
<span id="cb277-2"><a href="#cb277-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;Output/Check/01_coordinates_from_locality.csv&quot;</span>,</span>
<span id="cb277-3"><a href="#cb277-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Remove spent data</p>
<div class="sourceCode" id="cb278"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="#cb278-1" tabindex="-1"></a><span class="fu">rm</span>(xyFromLocality)</span></code></pre></div>
</div>
<div id="flag-absent" class="section level2">
<h2>3.10 Flag Absent</h2>
<p>Flag the records marked as “absent”</p>
<div class="sourceCode" id="cb279"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="#cb279-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_flagAbsent.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb279-2"><a href="#cb279-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">jbd_flagAbsent</span>(<span class="at">data =</span> check_pf,</span>
<span id="cb279-3"><a href="#cb279-3" tabindex="-1"></a>                           <span class="at">PresAbs =</span> <span class="st">&quot;occurrenceStatus&quot;</span>)</span></code></pre></div>
<pre><code>## \.occurrenceAbsent:
##  Flagged 10,904 absent records:
##  One column was added to the database.</code></pre>
</div>
<div id="flag-license" class="section level2">
<h2>3.11 flag License</h2>
<p>Flag the records that may not be used according to their license
information</p>
<div class="sourceCode" id="cb281"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="#cb281-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_flagLicense.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb281-2"><a href="#cb281-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">jbd_flagLicense</span>(<span class="at">data =</span> check_pf,</span>
<span id="cb281-3"><a href="#cb281-3" tabindex="-1"></a>                            <span class="at">strings_to_restrict =</span> <span class="st">&quot;all&quot;</span>,</span>
<span id="cb281-4"><a href="#cb281-4" tabindex="-1"></a>                            <span class="co"># DON&#39;T flag if in the following dataSource(s)</span></span>
<span id="cb281-5"><a href="#cb281-5" tabindex="-1"></a>                            <span class="at">excludeDataSource =</span> <span class="fu">c</span>(<span class="st">&quot;KP_Anthophila&quot;</span>))</span></code></pre></div>
<pre><code>## \.unLicensed:
##  Flagged 7 records that may NOT be used.
##  One column was added to the database.</code></pre>
</div>
<div id="gbif-issue" class="section level2">
<h2>3.12 GBIF issue</h2>
<p>Flag select issues that are flagged by GBIF</p>
<div class="sourceCode" id="cb283"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="#cb283-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_GBIFissues.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb283-2"><a href="#cb283-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">jbd_GBIFissues</span>(<span class="at">data =</span> check_pf, </span>
<span id="cb283-3"><a href="#cb283-3" tabindex="-1"></a>                           <span class="at">issueColumn =</span> <span class="st">&quot;issue&quot;</span>, </span>
<span id="cb283-4"><a href="#cb283-4" tabindex="-1"></a>                           <span class="at">GBIFflags =</span> <span class="fu">c</span>(<span class="st">&quot;COORDINATE_INVALID&quot;</span>, <span class="st">&quot;PRESUMED_NEGATED_LONGITUDE&quot;</span>, </span>
<span id="cb283-5"><a href="#cb283-5" tabindex="-1"></a>                                         <span class="st">&quot;PRESUMED_NEGATED_LATITUDE&quot;</span>, </span>
<span id="cb283-6"><a href="#cb283-6" tabindex="-1"></a>                                         <span class="st">&quot;COUNTRY_COORDINATE_MISMATCH&quot;</span>, <span class="st">&quot;ZERO_COORDINATE&quot;</span>)) </span></code></pre></div>
<pre><code>##  — jbd_GBIFissues:
## Flagged 2 
##   The .GBIFflags column was added to the database.</code></pre>
</div>
<div id="flag-reports" class="section level2">
<h2>3.13 Flag Reports</h2>
<div id="a.-save-flags" class="section level3">
<h3>a. Save flags</h3>
<p>Save the flags so far. This function will make sure that you keep a
copy of everything that has been flagged up until now. This will be
updated throughout the script and accessed at the end, so be wary of
moving files around manually.</p>
<div class="sourceCode" id="cb285"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="#cb285-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;flagRecorder.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb285-2"><a href="#cb285-2" tabindex="-1"></a>flagFile <span class="ot">&lt;-</span> <span class="fu">flagRecorder</span>(</span>
<span id="cb285-3"><a href="#cb285-3" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb285-4"><a href="#cb285-4" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Report&quot;</span>, <span class="at">sep =</span><span class="st">&quot;&quot;</span>),</span>
<span id="cb285-5"><a href="#cb285-5" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">&quot;flagsRecorded_&quot;</span>, <span class="fu">Sys.Date</span>(),  <span class="st">&quot;.csv&quot;</span>),</span>
<span id="cb285-6"><a href="#cb285-6" tabindex="-1"></a>    <span class="co"># These are the columns that will be kept along with the flags</span></span>
<span id="cb285-7"><a href="#cb285-7" tabindex="-1"></a>  <span class="at">idColumns =</span> <span class="fu">c</span>(<span class="st">&quot;database_id&quot;</span>, <span class="st">&quot;id&quot;</span>, <span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>, <span class="st">&quot;dataSource&quot;</span>),</span>
<span id="cb285-8"><a href="#cb285-8" tabindex="-1"></a>    <span class="co"># TRUE if you want to find a file from a previous part of the script to append to</span></span>
<span id="cb285-9"><a href="#cb285-9" tabindex="-1"></a>  <span class="at">findFile =</span> <span class="cn">FALSE</span>,</span>
<span id="cb285-10"><a href="#cb285-10" tabindex="-1"></a>  <span class="at">findFilePath =</span> DataPath)</span></code></pre></div>
<pre><code>##  — NO .summary column detected. This will added to the data.</code></pre>
<pre><code>##  — Data saved to E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>##  — Selected 14 columns. These include:
## database_id, id, catalogNumber, occurrenceID, dataSource, .scientificName_empty, .coordinates_empty, .coordinates_outOfRange, .basisOfRecords_notStandard, .coordinates_country_inconsistent, .occurrenceAbsent, .unLicensed, .GBIFflags, and .summary</code></pre>
</div>
<div id="b.-produce-summary-column-in-main-dataset"
class="section level3">
<h3>b. Produce summary column in main dataset</h3>
<div class="sourceCode" id="cb289"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="#cb289-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;summaryFun.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb289-2"><a href="#cb289-2" tabindex="-1"></a>check_pf <span class="ot">&lt;-</span> <span class="fu">summaryFun</span>(</span>
<span id="cb289-3"><a href="#cb289-3" tabindex="-1"></a>  <span class="at">data =</span> check_pf,</span>
<span id="cb289-4"><a href="#cb289-4" tabindex="-1"></a>    <span class="co"># Don&#39;t filter these columns (or NULL)</span></span>
<span id="cb289-5"><a href="#cb289-5" tabindex="-1"></a>  <span class="at">dontFilterThese =</span> <span class="cn">NULL</span>,</span>
<span id="cb289-6"><a href="#cb289-6" tabindex="-1"></a>    <span class="co"># Remove the filtering columns?</span></span>
<span id="cb289-7"><a href="#cb289-7" tabindex="-1"></a>  <span class="at">removeFilterColumns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb289-8"><a href="#cb289-8" tabindex="-1"></a>    <span class="co"># Filter to ONLY cleaned data?</span></span>
<span id="cb289-9"><a href="#cb289-9" tabindex="-1"></a>  <span class="at">filterClean =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>##  — We will flag all columns starting with &#39;.&#39;</code></pre>
<pre><code>##  — summaryColumnR:
## Flagged 31,588 
##   The .summary column was added to the database.</code></pre>
</div>
<div id="c.-reporting" class="section level3">
<h3>c. Reporting</h3>
<div class="sourceCode" id="cb292"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="#cb292-1" tabindex="-1"></a>(report <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_create_report</span>(<span class="at">data =</span> check_pf,</span>
<span id="cb292-2"><a href="#cb292-2" tabindex="-1"></a>                                  <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb292-3"><a href="#cb292-3" tabindex="-1"></a>                                  <span class="at">workflow_step =</span> <span class="st">&quot;prefilter&quot;</span>,</span>
<span id="cb292-4"><a href="#cb292-4" tabindex="-1"></a>                                  <span class="at">save_report =</span> <span class="cn">TRUE</span>)</span>
<span id="cb292-5"><a href="#cb292-5" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Loading required package: DT</code></pre>
<pre><code>## 
## bdc_create_report:
## Check the report summarizing the results of the prefilter in:
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/BeeDC-website/BeeDCtes2/Output/Report</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-b806ef77aa2e5b48f7f1" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-b806ef77aa2e5b48f7f1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8"],["Records with empty scientific name","Records with empty coordinates","Records with coordinates out-of-range","Records from a doubtful source","Records outside one or multiple reference countries","Summary of all tests","","(*) calculated in relation to total number of records, i.e. 153136 records"],[".scientificName_empty",".coordinates_empty",".coordinates_outOfRange",".basisOfRecords_notStandard",".coordinates_country_inconsistent",".summary","",""],["1394","19054","0","298","97","20676","",""],["0.91","12.44","0","0.19","0.06","13.5","",""]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Description<\/th>\n      <th>Test_name<\/th>\n      <th>Records_flagged<\/th>\n      <th>perc_number_records(*)<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":10,"dom":"Brtip","scrollX":true,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="d.-figures" class="section level3">
<h3>d. Figures</h3>
<div class="sourceCode" id="cb295"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="#cb295-1" tabindex="-1"></a>figures <span class="ot">&lt;-</span></span>
<span id="cb295-2"><a href="#cb295-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_create_figures</span>(<span class="at">data =</span> check_pf,</span>
<span id="cb295-3"><a href="#cb295-3" tabindex="-1"></a>                          <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb295-4"><a href="#cb295-4" tabindex="-1"></a>                          <span class="at">workflow_step =</span> <span class="st">&quot;prefilter&quot;</span>,</span>
<span id="cb295-5"><a href="#cb295-5" tabindex="-1"></a>                          <span class="at">save_figures =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Check figures in E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/BeeDC-website/BeeDCtes2/Output/Figures</code></pre>
<p>You can check figures using</p>
<div class="sourceCode" id="cb297"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="#cb297-1" tabindex="-1"></a>figures<span class="sc">$</span>.coordinates_country_inconsistent</span></code></pre></div>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-41-1.png" width="768" /></p>
</div>
</div>
<div id="save" class="section level2">
<h2>3.14 Save</h2>
<p>Save the intermediate dataset</p>
<div class="sourceCode" id="cb298"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="#cb298-1" tabindex="-1"></a>check_pf <span class="sc">%&gt;%</span></span>
<span id="cb298-2"><a href="#cb298-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(., <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/01_prefilter_output.csv&quot;</span>,</span>
<span id="cb298-3"><a href="#cb298-3" tabindex="-1"></a>                            <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
<p>Remove spent dataset</p>
<div class="sourceCode" id="cb299"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb299-1"><a href="#cb299-1" tabindex="-1"></a>database <span class="ot">&lt;-</span> check_pf</span></code></pre></div>
<div class="sourceCode" id="cb300"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="#cb300-1" tabindex="-1"></a><span class="fu">rm</span>(check_pf)</span></code></pre></div>
</div>
</div>
<div id="taxonomy" class="section level1">
<h1>4.0 Taxonomy</h1>
<p>See bdc tutorial <a
href="https://brunobrr.github.io/bdc/articles/taxonomy.html">here</a>.<br />
</p>
<p>Read in the filtered dataset</p>
<div class="sourceCode" id="cb301"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="#cb301-1" tabindex="-1"></a>database <span class="ot">&lt;-</span></span>
<span id="cb301-2"><a href="#cb301-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_csv</span>( <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/01_prefilter_output.csv&quot;</span>,</span>
<span id="cb301-3"><a href="#cb301-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>), <span class="at">col_types =</span> <span class="fu">ColTypeR</span>())</span></code></pre></div>
<p>Remove names_clean if it already exists (i.e. you have run this
before on this dataset)</p>
<pre><code>        database &lt;- database %&gt;%
          dplyr::select(!tidyselect::any_of(&quot;names_clean&quot;))</code></pre>
<div id="prep-data-names" class="section level2">
<h2>4.1 Prep data names</h2>
<p>This next step cleans the database’s scientificName column. The
cleaning tool will: - Flag and remove family names pre-pended to species
names - Flag and remove qualifiers denoting uncertain or provisional
status of taxonomic identification (e.g., confer, species, affinis, and
among others) - Flag and remove infraspecific terms (e.g., variety
[var.], subspecies [subsp], forma [f.], and their spelling variations) -
Standardize names, i.e., capitalize only the first letter of the genus
name and remove extra whitespaces) - Parse names, i.e., separate author,
date, annotations from taxon name.<br />
</p>
<p>***! MAC: You need to install gnparser through terminal — brew brew
tap gnames/gn brew install gnparser</p>
<div class="alert alert-info">
<p><strong> Attention:</strong><br />
This was difficult for a Windows install. Ensure you have the most
recent version of R, Rstudio, and R packages. Also check package
‘rgnparser’ is installed correctly. If you still can not get the below
code to work, you may have to download the latest version of ‘gnparser’
from <a
href="https://github.com/gnames/gnparser/releases/tag/v1.6.9">here</a>.
You may then need to manually install it and edit your systems
environmental variable PATH to locate ‘gnparser.exe’. See <a
href="https://github.com/gnames/gnparser#installation">here</a>.</p>
</div>
<div class="sourceCode" id="cb303"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="#cb303-1" tabindex="-1"></a>parse_names <span class="ot">&lt;-</span></span>
<span id="cb303-2"><a href="#cb303-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_clean_names</span>(<span class="at">sci_names =</span> database<span class="sc">$</span>scientificName, <span class="at">save_outputs =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 
## &gt;&gt; Family names prepended to scientific names were flagged and removed from 6 records.</code></pre>
<pre><code>## &gt;&gt; Terms denoting taxonomic uncertainty were flagged and removed from 190 records.</code></pre>
<pre><code>## &gt;&gt; Other issues, capitalizing the first letter of the generic name, replacing empty names by NA, and removing extra spaces, were flagged and corrected or removed from 288 records.</code></pre>
<pre><code>## &gt;&gt; Infraspecific terms were flagged and removed from 105 records.</code></pre>
<pre><code>## Error: gnparser not found. You can install it via rgnparser::install_gnparser()</code></pre>
<p>Keep only the .uncer_terms and names_clean columns</p>
<div class="sourceCode" id="cb309"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="#cb309-1" tabindex="-1"></a>parse_names <span class="ot">&lt;-</span></span>
<span id="cb309-2"><a href="#cb309-2" tabindex="-1"></a>  parse_names <span class="sc">%&gt;%</span></span>
<span id="cb309-3"><a href="#cb309-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(.uncer_terms, names_clean)</span></code></pre></div>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;parse_names&#39; not found</code></pre>
<p>Merge names with the complete dataset</p>
<div class="sourceCode" id="cb311"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="#cb311-1" tabindex="-1"></a>database <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">bind_cols</span>(database, parse_names)</span></code></pre></div>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;parse_names&#39; not found</code></pre>
<div class="sourceCode" id="cb313"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="#cb313-1" tabindex="-1"></a><span class="co">#rm(parse_names)</span></span></code></pre></div>
<div class="sourceCode" id="cb314"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="#cb314-1" tabindex="-1"></a><span class="co">#database &lt;-database %&gt;% select(!starts_with(&quot;names_clean&quot;))</span></span></code></pre></div>
<div class="sourceCode" id="cb315"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="#cb315-1" tabindex="-1"></a><span class="co">#database &lt;-database %&gt;% select(!starts_with(&quot;.uncer_terms&quot;))</span></span></code></pre></div>
</div>
<div id="taxo-harmonization" class="section level2">
<h2>4.2 Taxo harmonization</h2>
<p>Read in the custom taxonomy file</p>
<div class="sourceCode" id="cb316"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb316-1"><a href="#cb316-1" tabindex="-1"></a>BeeTaxonomy <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="fu">paste</span>(DiscLifePath, <span class="st">&quot;/Paige_DLdf_2022-07-29.csv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb316-2"><a href="#cb316-2" tabindex="-1"></a>                               <span class="at">col_types =</span> <span class="fu">c</span>(<span class="at">flags =</span> <span class="st">&quot;c&quot;</span>))</span></code></pre></div>
<p>Harmonise the names in the occurrence tibble</p>
<div class="sourceCode" id="cb317"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="#cb317-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;HarmoniseR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb317-2"><a href="#cb317-2" tabindex="-1"></a><span class="co">#   # This flags the occurrences without a matched name and matches names to their correct name </span></span>
<span id="cb317-3"><a href="#cb317-3" tabindex="-1"></a>  <span class="co"># according to DiscoverLife</span></span>
<span id="cb317-4"><a href="#cb317-4" tabindex="-1"></a>database <span class="ot">&lt;-</span> <span class="fu">HarmoniseR</span>(<span class="at">path =</span> DataPath, <span class="co">#The path to a folder that the output can be saved</span></span>
<span id="cb317-5"><a href="#cb317-5" tabindex="-1"></a>                       <span class="at">SynList =</span> BeeTaxonomy, <span class="co"># The formatted taxonomy file</span></span>
<span id="cb317-6"><a href="#cb317-6" tabindex="-1"></a>                       <span class="at">occurrences =</span> database)</span></code></pre></div>
<pre><code>##  — Formatting SynList for matching...
## 
##  — Harmonise the occurrence data with unambiguous names...
## 
##  — Attempting to harmonise the occurrence data with ambiguous names...
##  — Formatting merged datasets...
##  — We matched valid names to 129,593 of 153,136 occurrence records. This leaves a total of 23,543 unmatched occurrence records. Of the unmatched records, approximately 21,040 are only identified to genus rank or higher.
## 
## HarmoniseR:</code></pre>
<pre><code>## 26,157</code></pre>
<pre><code>## records were flagged.
## The column, &#39;.invalidName&#39; was added to the database.</code></pre>
<pre><code>##  — We updated the following columns: scientificName, species, family, subfamily, genus, subgenus, specificEpithet, infraspecificEpithet, and scientificNameAuthorship. The previous scientificName column was converted to verbatimScientificName</code></pre>
<pre><code>##  — Completed in 2.79 minutes.</code></pre>
<p>You don’t need this file any more…</p>
<div class="sourceCode" id="cb323"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="#cb323-1" tabindex="-1"></a><span class="fu">rm</span>(BeeTaxonomy)</span></code></pre></div>
<p>Save the harmonised file.</p>
<div class="sourceCode" id="cb324"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="#cb324-1" tabindex="-1"></a>database <span class="sc">%&gt;%</span></span>
<span id="cb324-2"><a href="#cb324-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb324-3"><a href="#cb324-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;02_taxonomy_database.csv&quot;</span>,</span>
<span id="cb324-4"><a href="#cb324-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
</div>
<div id="save-flags" class="section level2">
<h2>4.3 Save flags</h2>
<p>SAVE the flags so far.This will find the most-recent flag file and
append your new data to it. You can double-check the data and number of
columns if you’d like to be thorough and sure that all data are intact.
&lt;3</p>
<div class="sourceCode" id="cb325"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="#cb325-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;flagRecorder.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb325-2"><a href="#cb325-2" tabindex="-1"></a>flagFile <span class="ot">&lt;-</span> <span class="fu">flagRecorder</span>(</span>
<span id="cb325-3"><a href="#cb325-3" tabindex="-1"></a>  <span class="at">data =</span> database,</span>
<span id="cb325-4"><a href="#cb325-4" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Report&quot;</span>, <span class="at">sep =</span><span class="st">&quot;&quot;</span>),</span>
<span id="cb325-5"><a href="#cb325-5" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">&quot;flagsRecorded_&quot;</span>, <span class="fu">Sys.Date</span>(),  <span class="st">&quot;.csv&quot;</span>),</span>
<span id="cb325-6"><a href="#cb325-6" tabindex="-1"></a>  <span class="at">idColumns =</span> <span class="fu">c</span>(<span class="st">&quot;database_id&quot;</span>, <span class="st">&quot;id&quot;</span>, <span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>, <span class="st">&quot;dataSource&quot;</span>),</span>
<span id="cb325-7"><a href="#cb325-7" tabindex="-1"></a>  <span class="at">findFile =</span> <span class="cn">TRUE</span>,</span>
<span id="cb325-8"><a href="#cb325-8" tabindex="-1"></a>  <span class="at">printSummary =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##  — Dates found in file name(s). Finding most-recent file from file name...
##  — Found the following file(s): 
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>## Rows: 153136 Columns: 14
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr (5): database_id, id, catalogNumber, occurrenceID, dataSource
## lgl (9): .scientificName_empty, .coordinates_empty, .coordinates_outOfRange,...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
##  — .summary column detected. This will be over-written.
## 
##  — Data saved to E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>##  — Selected 15 columns. These include:
## database_id, id, catalogNumber, occurrenceID, dataSource, .scientificName_empty, .coordinates_empty, .coordinates_outOfRange, .basisOfRecords_notStandard, .coordinates_country_inconsistent, .occurrenceAbsent, .unLicensed, .GBIFflags, .summary, and .invalidName</code></pre>
</div>
</div>
<div id="space-the-final-frontier" class="section level1">
<h1>5.0 Space: The final frontier</h1>
<p>Read in the last database</p>
<div class="sourceCode" id="cb329"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="#cb329-1" tabindex="-1"></a><span class="co">#database &lt;-</span></span>
<span id="cb329-2"><a href="#cb329-2" tabindex="-1"></a>  <span class="co">#readr::read_csv(paste(DataPath, &quot;/Output/Intermediate/02_taxonomy_database.csv&quot;, sep = &quot;&quot;),</span></span>
<span id="cb329-3"><a href="#cb329-3" tabindex="-1"></a>                  <span class="co"># It is likely that not all columns are present in this list and hence might </span></span>
<span id="cb329-4"><a href="#cb329-4" tabindex="-1"></a>                  <span class="co"># throw a warning. Don&#39;t stress it.</span></span>
<span id="cb329-5"><a href="#cb329-5" tabindex="-1"></a>      <span class="co">#            col_types = ColTypeR())</span></span></code></pre></div>
<div id="coord-precision" class="section level2">
<h2>5.1 Coord precision</h2>
<p>This function identifies records with a coordinate precision below a
specified number of decimal places. For example, the precision of a
coordinate with 1 decimal place is 11.132 km at the equator, i.e., the
scale of a large city. “Coordinates with one, two, or three decimal
places present a precision of ~11.1 km, ~1.1 km, and ~111 m at the
equator, respectively.”</p>
<div class="sourceCode" id="cb330"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="#cb330-1" tabindex="-1"></a>check_space <span class="ot">&lt;-</span></span>
<span id="cb330-2"><a href="#cb330-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_coordinates_precision</span>(</span>
<span id="cb330-3"><a href="#cb330-3" tabindex="-1"></a>    <span class="at">data =</span> database,</span>
<span id="cb330-4"><a href="#cb330-4" tabindex="-1"></a>    <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb330-5"><a href="#cb330-5" tabindex="-1"></a>    <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb330-6"><a href="#cb330-6" tabindex="-1"></a>    <span class="at">ndec =</span> <span class="dv">2</span> <span class="co"># number of decimals to be tested</span></span>
<span id="cb330-7"><a href="#cb330-7" tabindex="-1"></a>  )</span></code></pre></div>
<pre><code>## bdc_coordinates_precision:
## Flagged 29200 records
## One column was added to the database.</code></pre>
<p>Remove the spent dataset</p>
<div class="sourceCode" id="cb332"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb332-1"><a href="#cb332-1" tabindex="-1"></a><span class="co">#rm(database)</span></span></code></pre></div>
<p>Save the harmonised file.</p>
<div class="sourceCode" id="cb333"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="#cb333-1" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb333-2"><a href="#cb333-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb333-3"><a href="#cb333-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;03_space_inter_database.csv&quot;</span>,</span>
<span id="cb333-4"><a href="#cb333-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
</div>
<div id="common-spatial-issues" class="section level2">
<h2>5.2 Common spatial issues</h2>
<p>Only run for occurrences through clean_coordinates that are spatially
‘valid’.</p>
<div class="sourceCode" id="cb334"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb334-1"><a href="#cb334-1" tabindex="-1"></a>tempSpace <span class="ot">&lt;-</span> check_space <span class="sc">%&gt;%</span> </span>
<span id="cb334-2"><a href="#cb334-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span>.coordinates_empty <span class="sc">==</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb334-3"><a href="#cb334-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span>.coordinates_outOfRange <span class="sc">==</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Next, we will flag common spatial issues using functions of the
package CoordinateCleaner. It addresses some common issues in
biodiversity datasets</p>
<div class="sourceCode" id="cb335"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="#cb335-1" tabindex="-1"></a>tempSpace <span class="ot">&lt;-</span></span>
<span id="cb335-2"><a href="#cb335-2" tabindex="-1"></a>  CoordinateCleaner<span class="sc">::</span><span class="fu">clean_coordinates</span>(</span>
<span id="cb335-3"><a href="#cb335-3" tabindex="-1"></a>    <span class="at">x =</span>  tempSpace,</span>
<span id="cb335-4"><a href="#cb335-4" tabindex="-1"></a>    <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb335-5"><a href="#cb335-5" tabindex="-1"></a>    <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb335-6"><a href="#cb335-6" tabindex="-1"></a>    <span class="at">species =</span> <span class="st">&quot;scientificName&quot;</span>,</span>
<span id="cb335-7"><a href="#cb335-7" tabindex="-1"></a>    <span class="at">countries =</span> <span class="cn">NULL</span>, <span class="co"># Tests if coords are from x country. This is not needed.</span></span>
<span id="cb335-8"><a href="#cb335-8" tabindex="-1"></a>    <span class="at">tests =</span> <span class="fu">c</span>(</span>
<span id="cb335-9"><a href="#cb335-9" tabindex="-1"></a>      <span class="st">&quot;capitals&quot;</span>,     <span class="co"># records within 0.5 km of capitals centroids</span></span>
<span id="cb335-10"><a href="#cb335-10" tabindex="-1"></a>      <span class="st">&quot;centroids&quot;</span>,    <span class="co"># records within 1 km around country and province centroids</span></span>
<span id="cb335-11"><a href="#cb335-11" tabindex="-1"></a>      <span class="co"># &quot;duplicates&quot;,   # duplicated records</span></span>
<span id="cb335-12"><a href="#cb335-12" tabindex="-1"></a>      <span class="st">&quot;equal&quot;</span>,      <span class="co"># records with equal coordinates</span></span>
<span id="cb335-13"><a href="#cb335-13" tabindex="-1"></a>      <span class="st">&quot;gbif&quot;</span>,         <span class="co"># records within 1 km of GBIF headquarters. (says 1 degree in package, but code says 1000 m)</span></span>
<span id="cb335-14"><a href="#cb335-14" tabindex="-1"></a>      <span class="st">&quot;institutions&quot;</span>, <span class="co"># records within 100m of zoo and herbaria</span></span>
<span id="cb335-15"><a href="#cb335-15" tabindex="-1"></a>      <span class="co"># &quot;outliers&quot;, # outliers</span></span>
<span id="cb335-16"><a href="#cb335-16" tabindex="-1"></a>      <span class="st">&quot;zeros&quot;</span>       <span class="co"># records with coordinates 0,0</span></span>
<span id="cb335-17"><a href="#cb335-17" tabindex="-1"></a>      <span class="co"># &quot;seas&quot;        # Not flagged as this should be flagged by coordinate country inconsistent</span></span>
<span id="cb335-18"><a href="#cb335-18" tabindex="-1"></a>    ),</span>
<span id="cb335-19"><a href="#cb335-19" tabindex="-1"></a>    <span class="at">capitals_rad =</span> <span class="dv">1000</span>,</span>
<span id="cb335-20"><a href="#cb335-20" tabindex="-1"></a>    <span class="at">centroids_rad =</span> <span class="dv">500</span>,</span>
<span id="cb335-21"><a href="#cb335-21" tabindex="-1"></a>    <span class="at">centroids_detail =</span> <span class="st">&quot;both&quot;</span>, <span class="co"># test both country and province centroids</span></span>
<span id="cb335-22"><a href="#cb335-22" tabindex="-1"></a>    <span class="at">inst_rad =</span> <span class="dv">100</span>, <span class="co"># remove zoo and herbaria within 100m</span></span>
<span id="cb335-23"><a href="#cb335-23" tabindex="-1"></a>    <span class="co"># outliers_method = &quot;quantile&quot;,</span></span>
<span id="cb335-24"><a href="#cb335-24" tabindex="-1"></a>    <span class="co"># outliers_mtp = 5,</span></span>
<span id="cb335-25"><a href="#cb335-25" tabindex="-1"></a>    <span class="co"># outliers_td = 1000,</span></span>
<span id="cb335-26"><a href="#cb335-26" tabindex="-1"></a>    <span class="co"># outliers_size = 10,</span></span>
<span id="cb335-27"><a href="#cb335-27" tabindex="-1"></a>    <span class="at">range_rad =</span> <span class="dv">0</span>,</span>
<span id="cb335-28"><a href="#cb335-28" tabindex="-1"></a>    <span class="at">zeros_rad =</span> <span class="fl">0.5</span>,</span>
<span id="cb335-29"><a href="#cb335-29" tabindex="-1"></a>    <span class="at">capitals_ref =</span> <span class="cn">NULL</span>,</span>
<span id="cb335-30"><a href="#cb335-30" tabindex="-1"></a>    <span class="at">centroids_ref =</span> <span class="cn">NULL</span>,</span>
<span id="cb335-31"><a href="#cb335-31" tabindex="-1"></a>    <span class="at">country_ref =</span> <span class="cn">NULL</span>,</span>
<span id="cb335-32"><a href="#cb335-32" tabindex="-1"></a>    <span class="at">country_refcol =</span> <span class="st">&quot;countryCode&quot;</span>,</span>
<span id="cb335-33"><a href="#cb335-33" tabindex="-1"></a>    <span class="at">inst_ref =</span> <span class="cn">NULL</span>,</span>
<span id="cb335-34"><a href="#cb335-34" tabindex="-1"></a>    <span class="at">range_ref =</span> <span class="cn">NULL</span>,</span>
<span id="cb335-35"><a href="#cb335-35" tabindex="-1"></a>    <span class="co"># seas_scale = 50,</span></span>
<span id="cb335-36"><a href="#cb335-36" tabindex="-1"></a>    <span class="at">value =</span> <span class="st">&quot;spatialvalid&quot;</span> <span class="co"># result of tests are appended in separate columns</span></span>
<span id="cb335-37"><a href="#cb335-37" tabindex="-1"></a>  )</span></code></pre></div>
<p>Flagged 503 records. Flagged 737 of 134082 records, EQ = 0.01.</p>
<p>re-merge the datasets</p>
<div class="sourceCode" id="cb336"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="#cb336-1" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> tempSpace <span class="sc">%&gt;%</span></span>
<span id="cb336-2"><a href="#cb336-2" tabindex="-1"></a>  <span class="co"># Re-bind with the records that were removed earlier</span></span>
<span id="cb336-3"><a href="#cb336-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(check_space <span class="sc">%&gt;%</span> </span>
<span id="cb336-4"><a href="#cb336-4" tabindex="-1"></a>                     dplyr<span class="sc">::</span><span class="fu">filter</span>(.coordinates_empty <span class="sc">==</span> <span class="cn">FALSE</span> <span class="sc">|</span> </span>
<span id="cb336-5"><a href="#cb336-5" tabindex="-1"></a>                                     .coordinates_outOfRange <span class="sc">==</span> <span class="cn">FALSE</span>) )</span></code></pre></div>
<pre><code>## New names:
## • `.summary` -&gt; `.summary...102`
## • `.summary` -&gt; `.summary...113`</code></pre>
<p>Remove the temporary dataset</p>
<div class="sourceCode" id="cb338"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="#cb338-1" tabindex="-1"></a><span class="fu">rm</span>(tempSpace)</span></code></pre></div>
<p>Remove the flag summary columns</p>
<div class="sourceCode" id="cb339"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="#cb339-1" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> check_space <span class="sc">%&gt;%</span></span>
<span id="cb339-2"><a href="#cb339-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">contains</span>(<span class="st">&quot;summary&quot;</span>))</span></code></pre></div>
<p>Save the intermediate dataset</p>
<div class="sourceCode" id="cb340"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb340-1"><a href="#cb340-1" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb340-2"><a href="#cb340-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/03_space_inter_database.csv&quot;</span>,</span>
<span id="cb340-3"><a href="#cb340-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
</div>
<div id="diagonal-grid" class="section level2">
<h2>5.3 diagonal + grid</h2>
<p>Finds sequential numbers that could be fill-down errors in lat and
long, and groups by eventDate, recordedBy. This will only find those
where <em>ALL</em> records in a group are filled-down. Missing
occurrences will break the chain</p>
<div class="sourceCode" id="cb341"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb341-1"><a href="#cb341-1" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> <span class="fu">jbd_diagonAlley</span>(</span>
<span id="cb341-2"><a href="#cb341-2" tabindex="-1"></a>  <span class="at">data =</span> check_space,</span>
<span id="cb341-3"><a href="#cb341-3" tabindex="-1"></a>  <span class="co"># The minimum number of repeats needed to find a sequence in for flagging</span></span>
<span id="cb341-4"><a href="#cb341-4" tabindex="-1"></a>  <span class="at">minRepeats =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>## jbd_diagonAlley:
## Flagged 0 records
## The .sequential column was added to the database.</code></pre>
<div class="sourceCode" id="cb343"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="#cb343-1" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb343-2"><a href="#cb343-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/03_space_inter_database.csv&quot;</span>,</span>
<span id="cb343-3"><a href="#cb343-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
<p>SPATIAL gridding from rasterisation: Select only the records with
more than X occurrences.</p>
<div class="sourceCode" id="cb344"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb344-1"><a href="#cb344-1" tabindex="-1"></a>griddingDF <span class="ot">&lt;-</span> check_space <span class="sc">%&gt;%</span></span>
<span id="cb344-2"><a href="#cb344-2" tabindex="-1"></a>  <span class="co"># Exclude NA lat and lon values</span></span>
<span id="cb344-3"><a href="#cb344-3" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">drop_na</span>(<span class="fu">c</span>(<span class="st">&quot;decimalLatitude&quot;</span>, <span class="st">&quot;decimalLongitude&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb344-4"><a href="#cb344-4" tabindex="-1"></a>  <span class="co"># Group by the dataset name</span></span>
<span id="cb344-5"><a href="#cb344-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(datasetName) <span class="sc">%&gt;%</span></span>
<span id="cb344-6"><a href="#cb344-6" tabindex="-1"></a>  <span class="co"># Remove rows that aren&#39;t unique for lat and long</span></span>
<span id="cb344-7"><a href="#cb344-7" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(decimalLongitude, decimalLatitude,</span>
<span id="cb344-8"><a href="#cb344-8" tabindex="-1"></a>                  <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb344-9"><a href="#cb344-9" tabindex="-1"></a>  <span class="co"># Find the groups with 4 or more occurrence records </span></span>
<span id="cb344-10"><a href="#cb344-10" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(dplyr<span class="sc">::</span><span class="fu">n</span>() <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb344-11"><a href="#cb344-11" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>()</span></code></pre></div>
<p>Run the gridding analysis to find datasets that might be gridded.</p>
<div class="sourceCode" id="cb345"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb345-1"><a href="#cb345-1" tabindex="-1"></a>gridded_datasets <span class="ot">&lt;-</span> CoordinateCleaner<span class="sc">::</span><span class="fu">cd_round</span>(</span>
<span id="cb345-2"><a href="#cb345-2" tabindex="-1"></a>  <span class="at">x =</span> griddingDF,</span>
<span id="cb345-3"><a href="#cb345-3" tabindex="-1"></a>  <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb345-4"><a href="#cb345-4" tabindex="-1"></a>  <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb345-5"><a href="#cb345-5" tabindex="-1"></a>  <span class="at">ds =</span> <span class="st">&quot;datasetName&quot;</span>,</span>
<span id="cb345-6"><a href="#cb345-6" tabindex="-1"></a>  <span class="at">T1 =</span> <span class="dv">7</span>,</span>
<span id="cb345-7"><a href="#cb345-7" tabindex="-1"></a>  <span class="at">min_unique_ds_size =</span> <span class="dv">4</span>,</span>
<span id="cb345-8"><a href="#cb345-8" tabindex="-1"></a>  <span class="at">test =</span> <span class="st">&quot;both&quot;</span>,</span>
<span id="cb345-9"><a href="#cb345-9" tabindex="-1"></a>  <span class="at">value =</span> <span class="st">&quot;dataset&quot;</span>,</span>
<span id="cb345-10"><a href="#cb345-10" tabindex="-1"></a>  <span class="at">graphs =</span> <span class="cn">FALSE</span>,</span>
<span id="cb345-11"><a href="#cb345-11" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span>,</span>
<span id="cb345-12"><a href="#cb345-12" tabindex="-1"></a>  <span class="at">reg_out_thresh =</span> <span class="dv">2</span>,</span>
<span id="cb345-13"><a href="#cb345-13" tabindex="-1"></a>  <span class="at">reg_dist_min =</span> <span class="fl">0.1</span>,</span>
<span id="cb345-14"><a href="#cb345-14" tabindex="-1"></a>  <span class="at">reg_dist_max =</span> <span class="dv">2</span></span>
<span id="cb345-15"><a href="#cb345-15" tabindex="-1"></a>) <span class="sc">%&gt;%</span> </span>
<span id="cb345-16"><a href="#cb345-16" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">tibble</span>()</span></code></pre></div>
<pre><code>## Testing for rasterized collection</code></pre>
<div class="sourceCode" id="cb347"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb347-1"><a href="#cb347-1" tabindex="-1"></a><span class="co"># The griddingDF is no longer needed. remove it.</span></span>
<span id="cb347-2"><a href="#cb347-2" tabindex="-1"></a><span class="co">#rm(griddingDF)</span></span></code></pre></div>
<p>Integrate these results with the main dataset</p>
<div class="sourceCode" id="cb348"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="#cb348-1" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> check_space <span class="sc">%&gt;%</span></span>
<span id="cb348-2"><a href="#cb348-2" tabindex="-1"></a>  <span class="co"># Join the datasets</span></span>
<span id="cb348-3"><a href="#cb348-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb348-4"><a href="#cb348-4" tabindex="-1"></a>    <span class="co"># Select the columns of interest</span></span>
<span id="cb348-5"><a href="#cb348-5" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(gridded_datasets, dataset, lon.flag, lat.flag, summary),</span>
<span id="cb348-6"><a href="#cb348-6" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;datasetName&quot;</span> <span class="ot">=</span> <span class="st">&quot;dataset&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb348-7"><a href="#cb348-7" tabindex="-1"></a>  <span class="co"># Make new columns with more-consistent naming and change the NA vlaues to = TRUE (not flagged)</span></span>
<span id="cb348-8"><a href="#cb348-8" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">.lonFlag =</span> tidyr<span class="sc">::</span><span class="fu">replace_na</span>(lon.flag, <span class="cn">TRUE</span>),</span>
<span id="cb348-9"><a href="#cb348-9" tabindex="-1"></a>                <span class="at">.latFlag =</span> tidyr<span class="sc">::</span><span class="fu">replace_na</span>(lat.flag, <span class="cn">TRUE</span>),</span>
<span id="cb348-10"><a href="#cb348-10" tabindex="-1"></a>                <span class="at">.gridSummary =</span> tidyr<span class="sc">::</span><span class="fu">replace_na</span>(summary, <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb348-11"><a href="#cb348-11" tabindex="-1"></a>  <span class="co"># Remove old columns</span></span>
<span id="cb348-12"><a href="#cb348-12" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(lon.flag, lat.flag, summary))</span></code></pre></div>
<p>Save the gridded_datasets file for later examination</p>
<div class="sourceCode" id="cb349"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb349-1"><a href="#cb349-1" tabindex="-1"></a>gridded_datasets <span class="sc">%&gt;%</span></span>
<span id="cb349-2"><a href="#cb349-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/03_space_griddedDatasets.csv&quot;</span>,</span>
<span id="cb349-3"><a href="#cb349-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
<p>Now remove this file</p>
<div class="sourceCode" id="cb350"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="#cb350-1" tabindex="-1"></a><span class="fu">rm</span>(gridded_datasets)</span></code></pre></div>
</div>
<div id="uncertainty" class="section level2">
<h2>5.4 Uncertainty</h2>
<p>Flag records that exceed a coordinateUncertaintyInMeters
threshold.</p>
<div class="sourceCode" id="cb351"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb351-1"><a href="#cb351-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;coordUncerFlagR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb351-2"><a href="#cb351-2" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> <span class="fu">coordUncerFlagR</span>(<span class="at">data =</span> check_space,</span>
<span id="cb351-3"><a href="#cb351-3" tabindex="-1"></a>                               <span class="at">uncerColumn =</span> <span class="st">&quot;coordinateUncertaintyInMeters&quot;</span>,</span>
<span id="cb351-4"><a href="#cb351-4" tabindex="-1"></a>                               <span class="at">threshold =</span> <span class="dv">1000</span>)</span></code></pre></div>
<pre><code>## \coordUncerFlagR:
##  Flagged 24,005 absent records:
##  The column &#39;.uncertaintyThreshold&#39; was added to the database.</code></pre>
</div>
<div id="country-checklist" class="section level2">
<h2>5.5 Country Checklist</h2>
<p>Read in the country-level checklist</p>
<div class="sourceCode" id="cb353"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="#cb353-1" tabindex="-1"></a>checklist <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="fu">paste0</span>(DiscLifePath, <span class="st">&quot;/runningFN_CheckL_2022-12-07.csv&quot;</span>),</span>
<span id="cb353-2"><a href="#cb353-2" tabindex="-1"></a>                              <span class="at">guess_max =</span> <span class="dv">65000</span>, <span class="at">lazy =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Rows: 63554 Columns: 29
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr (24): validName, ISO, Long_country_name, Country_name, Alpha-2, Alpha-3,...
## dbl  (4): ID, Freq, accid, id
## lgl  (1): subgenus
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb355"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="#cb355-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;countryOutlieRs.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb355-2"><a href="#cb355-2" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> <span class="fu">countryOutlieRs</span>(<span class="at">checklist =</span> checklist,</span>
<span id="cb355-3"><a href="#cb355-3" tabindex="-1"></a>                        <span class="at">occData =</span> check_space,</span>
<span id="cb355-4"><a href="#cb355-4" tabindex="-1"></a>                        <span class="at">keepAdjacentCountry =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##  — Extracting country data from points...</code></pre>
<pre><code>## although coordinates are longitude/latitude, st_intersection assumes that they
## are planar</code></pre>
<pre><code>## although coordinates are longitude/latitude, st_intersects assumes that they
## are planar</code></pre>
<pre><code>##  — Combining data...
##  — Finished. 
## We have matched 80,706 records to their exact country and 19,071 to an adjacent country
## We failed to match 26,301 occurrences to any &#39;exact&#39; or &#39;neighbouring&#39; country.
## There are 0 &#39;NA&#39; occurrences for this column.</code></pre>
<pre><code>## countryOutlieRs:
## Flagged 26,301 records.
## Two columns were added to the database:
##  1.  The &#39;.countryOutlier&#39; column was added which is the filtering column. 
##  2.  The &#39;countryMatch&#39; columns indicates exact, neighbour, or noMatch.</code></pre>
<div class="sourceCode" id="cb361"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="#cb361-1" tabindex="-1"></a>  <span class="co"># A list of failed species-country combinations and their numbers can be output here</span></span>
<span id="cb361-2"><a href="#cb361-2" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb361-3"><a href="#cb361-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(.countryOutlier <span class="sc">==</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb361-4"><a href="#cb361-4" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(database_id, scientificName, country) <span class="sc">%&gt;%</span></span>
<span id="cb361-5"><a href="#cb361-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(scientificName) <span class="sc">%&gt;%</span> </span>
<span id="cb361-6"><a href="#cb361-6" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">count_scientificName =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb361-7"><a href="#cb361-7" tabindex="-1"></a>  <span class="fu">distinct</span>(scientificName, country, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb361-8"><a href="#cb361-8" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/03_space_failedCountryChecklist.csv&quot;</span>,</span>
<span id="cb361-9"><a href="#cb361-9" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
</div>
<div id="map-spatial-errors" class="section level2">
<h2>5.6 Map spatial errors</h2>
<p>rebuild the .summary column</p>
<div class="sourceCode" id="cb362"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="#cb362-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;summaryFun.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb362-2"><a href="#cb362-2" tabindex="-1"></a>check_space <span class="ot">&lt;-</span> <span class="fu">summaryFun</span>(</span>
<span id="cb362-3"><a href="#cb362-3" tabindex="-1"></a>  <span class="at">data =</span> check_space,</span>
<span id="cb362-4"><a href="#cb362-4" tabindex="-1"></a>  <span class="co"># Don&#39;t filter these columns (or NULL)</span></span>
<span id="cb362-5"><a href="#cb362-5" tabindex="-1"></a>  <span class="at">dontFilterThese =</span> <span class="cn">NULL</span>,</span>
<span id="cb362-6"><a href="#cb362-6" tabindex="-1"></a>  <span class="co"># Remove the other filtering columns? — Do NOT remove the other filtering columns</span></span>
<span id="cb362-7"><a href="#cb362-7" tabindex="-1"></a>  <span class="at">removeFilterColumns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb362-8"><a href="#cb362-8" tabindex="-1"></a>  <span class="co"># Filter to ONLY cleaned data? — Do NOT filter for only clean data</span></span>
<span id="cb362-9"><a href="#cb362-9" tabindex="-1"></a>  <span class="at">filterClean =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>##  — We will flag all columns starting with &#39;.&#39;</code></pre>
<pre><code>##  — summaryColumnR:
## Flagged 90,110 
##   The .summary column was added to the database.</code></pre>
<p>Map ONE spatial flag at a time or map the .SUMMARY of all. Make this
selection in the col_to_map = section</p>
<div class="sourceCode" id="cb365"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="#cb365-1" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb365-2"><a href="#cb365-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(.summary <span class="sc">==</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> <span class="co"># map only records flagged as FALSE</span></span>
<span id="cb365-3"><a href="#cb365-3" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_quickmap</span>(</span>
<span id="cb365-4"><a href="#cb365-4" tabindex="-1"></a>    <span class="at">data =</span> .,</span>
<span id="cb365-5"><a href="#cb365-5" tabindex="-1"></a>    <span class="at">lon =</span> <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb365-6"><a href="#cb365-6" tabindex="-1"></a>    <span class="at">lat =</span> <span class="st">&quot;decimalLatitude&quot;</span>,</span>
<span id="cb365-7"><a href="#cb365-7" tabindex="-1"></a>    <span class="at">col_to_map =</span> <span class="st">&quot;.summary&quot;</span>,</span>
<span id="cb365-8"><a href="#cb365-8" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.9</span></span>
<span id="cb365-9"><a href="#cb365-9" tabindex="-1"></a>  )</span></code></pre></div>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-77-1.png" width="768" /></p>
</div>
<div id="space-report" class="section level2">
<h2>5.7 Space report</h2>
<p>Create the report</p>
<div class="sourceCode" id="cb366"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="#cb366-1" tabindex="-1"></a>(report <span class="ot">&lt;-</span></span>
<span id="cb366-2"><a href="#cb366-2" tabindex="-1"></a>   bdc<span class="sc">::</span><span class="fu">bdc_create_report</span>(</span>
<span id="cb366-3"><a href="#cb366-3" tabindex="-1"></a>     <span class="at">data =</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(check_space <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>.uncer_terms)),</span>
<span id="cb366-4"><a href="#cb366-4" tabindex="-1"></a>     <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb366-5"><a href="#cb366-5" tabindex="-1"></a>     <span class="at">workflow_step =</span> <span class="st">&quot;space&quot;</span>,</span>
<span id="cb366-6"><a href="#cb366-6" tabindex="-1"></a>     <span class="at">save_report =</span> <span class="cn">TRUE</span>)</span>
<span id="cb366-7"><a href="#cb366-7" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Error in `dplyr::select()`:
## ! Can&#39;t subset columns that don&#39;t exist.
## ✖ Column `.uncer_terms` doesn&#39;t exist.</code></pre>
</div>
<div id="space-figures" class="section level2">
<h2>5.8 Space figures</h2>
<p>Create figures of spacial data filtering.</p>
<div class="sourceCode" id="cb368"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="#cb368-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_create_figures.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb368-2"><a href="#cb368-2" tabindex="-1"></a>(figures <span class="ot">&lt;-</span></span>
<span id="cb368-3"><a href="#cb368-3" tabindex="-1"></a>    <span class="fu">jbd_create_figures</span>(</span>
<span id="cb368-4"><a href="#cb368-4" tabindex="-1"></a>      <span class="at">data =</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(check_space <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>.uncer_terms)),</span>
<span id="cb368-5"><a href="#cb368-5" tabindex="-1"></a>      <span class="at">path =</span> DataPath,</span>
<span id="cb368-6"><a href="#cb368-6" tabindex="-1"></a>      <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb368-7"><a href="#cb368-7" tabindex="-1"></a>      <span class="at">workflow_step =</span> <span class="st">&quot;space&quot;</span>,</span>
<span id="cb368-8"><a href="#cb368-8" tabindex="-1"></a>      <span class="at">save_figures =</span> <span class="cn">TRUE</span>)</span>
<span id="cb368-9"><a href="#cb368-9" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Error in `dplyr::select()`:
## ! Can&#39;t subset columns that don&#39;t exist.
## ✖ Column `.uncer_terms` doesn&#39;t exist.</code></pre>
<p>Check figure, the options are:<br />
- .cap == Records around country capital centroid<br />
- .cen == Records around country or province centroids - .dbl ==
Duplicated coordinates per species - .equ == Identical coordinates -
.otl == Geographical outliers - .gbf == Records around the GBIF
headquarters - .inst == Records around biodiversity institutions - .rou
== Rounded (probably imprecise) coordinates - .urb == Records within
urban areas — (Likely not relevant for bees.)<br />
Example:</p>
<pre><code>        figures$.rou</code></pre>
<p>Save interim dataset</p>
<div class="sourceCode" id="cb371"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="#cb371-1" tabindex="-1"></a>check_space <span class="sc">%&gt;%</span></span>
<span id="cb371-2"><a href="#cb371-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Intermediate&quot;</span>, <span class="st">&quot;/03_space_inter_database.csv&quot;</span>,</span>
<span id="cb371-3"><a href="#cb371-3" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
</div>
<div id="save-flags-1" class="section level2">
<h2>5.9 Save flags</h2>
<p>SAVE the flags so far.</p>
<div class="sourceCode" id="cb372"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="#cb372-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;flagRecorder.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb372-2"><a href="#cb372-2" tabindex="-1"></a><span class="fu">flagRecorder</span>(</span>
<span id="cb372-3"><a href="#cb372-3" tabindex="-1"></a>  <span class="at">data =</span> check_space,</span>
<span id="cb372-4"><a href="#cb372-4" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Report&quot;</span>, <span class="at">sep =</span><span class="st">&quot;&quot;</span>),</span>
<span id="cb372-5"><a href="#cb372-5" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">&quot;flagsRecorded_&quot;</span>, <span class="fu">Sys.Date</span>(),  <span class="st">&quot;.csv&quot;</span>),</span>
<span id="cb372-6"><a href="#cb372-6" tabindex="-1"></a>  <span class="at">idColumns =</span> <span class="fu">c</span>(<span class="st">&quot;database_id&quot;</span>, <span class="st">&quot;id&quot;</span>, <span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>, <span class="st">&quot;dataSource&quot;</span>),</span>
<span id="cb372-7"><a href="#cb372-7" tabindex="-1"></a>  <span class="at">findFile =</span> <span class="cn">TRUE</span>,</span>
<span id="cb372-8"><a href="#cb372-8" tabindex="-1"></a>  <span class="at">printSummary =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##  — Dates found in file name(s). Finding most-recent file from file name...
##  — Found the following file(s): 
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>## Rows: 153136 Columns: 15
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr  (5): database_id, id, catalogNumber, occurrenceID, dataSource
## lgl (10): .scientificName_empty, .coordinates_empty, .coordinates_outOfRange...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
##  — .summary column detected. This will be over-written.
## 
##  — Data saved to E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>##  — Selected 29 columns. These include:
## database_id, id, catalogNumber, occurrenceID, dataSource, .scientificName_empty, .coordinates_empty, .coordinates_outOfRange, .basisOfRecords_notStandard, .coordinates_country_inconsistent, .occurrenceAbsent, .unLicensed, .GBIFflags, .summary, .invalidName, .rou, .val, .equ, .zer, .cap, .cen, .gbf, .inst, .sequential, .lonFlag, .latFlag, .gridSummary, .uncertaintyThreshold, and .countryOutlier</code></pre>
<pre><code>## # A tibble: 153,136 × 29
##    database_id id    catalogNumber occurrenceID dataSource .scientificName_empty
##    &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;        &lt;chr&gt;      &lt;lgl&gt;                
##  1 Dorey_data… fd65… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  2 Dorey_data… b85a… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  3 Dorey_data… d95b… HYM36182      urn:lsid:oz… ALA_Apifo… TRUE                 
##  4 Dorey_data… 85bc… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  5 Dorey_data… b905… HYM36180      urn:lsid:oz… ALA_Apifo… TRUE                 
##  6 Dorey_data… f6fd… 132550620     https://www… ALA_Apifo… TRUE                 
##  7 Dorey_data… bd38… HYM42541      urn:lsid:oz… ALA_Apifo… TRUE                 
##  8 Dorey_data… f046… 31469707      https://www… ALA_Apifo… TRUE                 
##  9 Dorey_data… f3b7… 4244291       /sightings/… ALA_Apifo… TRUE                 
## 10 Dorey_data… 907a… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
## # ℹ 153,126 more rows
## # ℹ 23 more variables: .coordinates_empty &lt;lgl&gt;, .coordinates_outOfRange &lt;lgl&gt;,
## #   .basisOfRecords_notStandard &lt;lgl&gt;, .coordinates_country_inconsistent &lt;lgl&gt;,
## #   .occurrenceAbsent &lt;lgl&gt;, .unLicensed &lt;lgl&gt;, .GBIFflags &lt;lgl&gt;,
## #   .summary &lt;lgl&gt;, .invalidName &lt;lgl&gt;, .rou &lt;lgl&gt;, .val &lt;lgl&gt;, .equ &lt;lgl&gt;,
## #   .zer &lt;lgl&gt;, .cap &lt;lgl&gt;, .cen &lt;lgl&gt;, .gbf &lt;lgl&gt;, .inst &lt;lgl&gt;,
## #   .sequential &lt;lgl&gt;, .lonFlag &lt;lgl&gt;, .latFlag &lt;lgl&gt;, .gridSummary &lt;lgl&gt;, …</code></pre>
<p>BDC note:</p>
<p>Filtering the database: “It is possible to remove flagged records
(potentially problematic ones) to get a ‘clean’ database (i.e., without
test columns starting with “.”). However, to ensure that all records
will be evaluated in all the data quality tests (i.e., tests of the
taxonomic, spatial, and temporal module of the package), potentially
erroneous or suspect records will be removed in the final module of the
package”. However, I have decided to run all occurrences through all
flags so that data can be filtered and examined accordingly.</p>
<pre><code>        E.g. code:
        output &lt;- check_space
        # You can filter here if you want.
        #?     output &lt;-
        #?       check_space %&gt;%
        #?        # Dont filter out the uncertain term column.
        #?       dplyr::select(!.uncer_terms) %&gt;%
        #?        # Filter ALL where .summary = TRUE
        #?       dplyr::filter(.summary == TRUE) %&gt;%
        #?       bdc_filter_out_flags(data = ., col_to_remove = &quot;all&quot;)</code></pre>
<p>check_space is no longer needed and then remove it</p>
<div class="sourceCode" id="cb378"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="#cb378-1" tabindex="-1"></a><span class="co">#rm(check_space)</span></span></code></pre></div>
<p>For now, lets remove these columns to not mess up the Time step.</p>
<div class="sourceCode" id="cb379"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="#cb379-1" tabindex="-1"></a>   output <span class="ot">&lt;-</span></span>
<span id="cb379-2"><a href="#cb379-2" tabindex="-1"></a>     check_space <span class="sc">%&gt;%</span></span>
<span id="cb379-3"><a href="#cb379-3" tabindex="-1"></a>     <span class="fu">bdc_filter_out_flags</span>(<span class="at">data =</span> ., <span class="at">col_to_remove =</span> <span class="st">&quot;all&quot;</span>)</span></code></pre></div>
<pre><code>## 
## bdc_fiter_out_flags:
## The following columns were removed from the database:
## .scientificName_empty, .coordinates_empty, .coordinates_outOfRange, .basisOfRecords_notStandard, .coordinates_country_inconsistent, .occurrenceAbsent, .unLicensed, .invalidName, .rou, .val, .equ, .zer, .cap, .cen, .gbf, .inst, .sequential, .lonFlag, .latFlag, .gridSummary, .uncertaintyThreshold, .countryOutlier, .summary</code></pre>
</div>
<div id="save-1" class="section level2">
<h2>5.10 Save</h2>
<p>Save the intermediate dataset</p>
<div class="sourceCode" id="cb381"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb381-1"><a href="#cb381-1" tabindex="-1"></a>output <span class="sc">%&gt;%</span></span>
<span id="cb381-2"><a href="#cb381-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb381-3"><a href="#cb381-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;03_space_database.csv&quot;</span>,</span>
<span id="cb381-4"><a href="#cb381-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Remove the spent file</p>
<div class="sourceCode" id="cb382"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="#cb382-1" tabindex="-1"></a><span class="co">#rm(output)</span></span></code></pre></div>
</div>
</div>
<div id="time" class="section level1">
<h1>6.0 Time</h1>
<p>Read in the database again.</p>
<div class="sourceCode" id="cb383"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="#cb383-1" tabindex="-1"></a><span class="co">#database &lt;- readr::read_csv(paste0(DataPath, &quot;/Output/Intermediate/03_space_database.csv&quot;), </span></span>
<span id="cb383-2"><a href="#cb383-2" tabindex="-1"></a>                            <span class="co"># It is likely that not all columns are present in this list and hence might</span></span>
<span id="cb383-3"><a href="#cb383-3" tabindex="-1"></a>                            <span class="co"># throw a warning. Don&#39;t stress it.</span></span>
<span id="cb383-4"><a href="#cb383-4" tabindex="-1"></a>                  <span class="co">#          col_types = ColTypeR())</span></span></code></pre></div>
<p>Or, if you have not removed ‘output’ database you can change the
‘output’ object to ‘database’</p>
<div class="sourceCode" id="cb384"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="#cb384-1" tabindex="-1"></a>database <span class="ot">&lt;-</span> output</span></code></pre></div>
<p>You can plot a histogram of dates here, pre-cleaning to examine
potential issues.</p>
<div class="sourceCode" id="cb385"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="#cb385-1" tabindex="-1"></a><span class="fu">hist</span>(lubridate<span class="sc">::</span><span class="fu">ymd_hms</span>(database<span class="sc">$</span>eventDate, <span class="at">truncated =</span> <span class="dv">5</span>), <span class="at">breaks =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-88-1.png" width="768" />
Filter some silly dates that don’t make sense.</p>
<div class="sourceCode" id="cb386"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="#cb386-1" tabindex="-1"></a>database<span class="sc">$</span>year <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(database<span class="sc">$</span>year <span class="sc">&gt;</span> lubridate<span class="sc">::</span><span class="fu">year</span>(<span class="fu">Sys.Date</span>()) <span class="sc">|</span> database<span class="sc">$</span>year <span class="sc">&lt;</span> <span class="dv">1600</span>,</span>
<span id="cb386-2"><a href="#cb386-2" tabindex="-1"></a>                        <span class="cn">NA</span>, database<span class="sc">$</span>year)</span>
<span id="cb386-3"><a href="#cb386-3" tabindex="-1"></a>database<span class="sc">$</span>month <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(database<span class="sc">$</span>month <span class="sc">&gt;</span> <span class="dv">12</span> <span class="sc">|</span> database<span class="sc">$</span>month <span class="sc">&lt;</span> <span class="dv">1</span>,</span>
<span id="cb386-4"><a href="#cb386-4" tabindex="-1"></a>                         <span class="cn">NA</span>, database<span class="sc">$</span>month)</span>
<span id="cb386-5"><a href="#cb386-5" tabindex="-1"></a>database<span class="sc">$</span>day <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(database<span class="sc">$</span>day <span class="sc">&gt;</span> <span class="dv">31</span> <span class="sc">|</span> database<span class="sc">$</span>day <span class="sc">&lt;</span> <span class="dv">1</span>,</span>
<span id="cb386-6"><a href="#cb386-6" tabindex="-1"></a>                       <span class="cn">NA</span>, database<span class="sc">$</span>day)</span></code></pre></div>
<div id="recover-dates" class="section level2">
<h2>6.1 Recover dates</h2>
<p>RESCUE some records with poor date data if possible (e.g., from other
columns).</p>
<div class="sourceCode" id="cb387"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="#cb387-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;dateFindR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb387-2"><a href="#cb387-2" tabindex="-1"></a>check_time <span class="ot">&lt;-</span> <span class="fu">dateFindR</span>(<span class="at">data =</span> database,</span>
<span id="cb387-3"><a href="#cb387-3" tabindex="-1"></a>                        <span class="co"># Years above this are removed (from the recovered dates only)</span></span>
<span id="cb387-4"><a href="#cb387-4" tabindex="-1"></a>                        <span class="at">maxYear =</span> lubridate<span class="sc">::</span><span class="fu">year</span>(<span class="fu">Sys.Date</span>()),</span>
<span id="cb387-5"><a href="#cb387-5" tabindex="-1"></a>                        <span class="co"># Years below this are removed (from the recovered dates only)</span></span>
<span id="cb387-6"><a href="#cb387-6" tabindex="-1"></a>                        <span class="at">minYear =</span> <span class="dv">1700</span>)</span></code></pre></div>
<pre><code>##  — Preparing data...
##  — Extracting dates from year, month, day columns...
##  — Extracting dates from fieldNotes, locationRemarks, and verbatimEventDate columns in unambiguous ymd, dmy, mdy, and my formats...
##  — Extracting year from fieldNotes, locationRemarks, and verbatimEventDate columns in ambiguous formats...
##  — Formating and combining the new data..
##  — Merging all data, nearly there...</code></pre>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-90-1.png" width="768" /></p>
<pre><code>##  — Finished. 
## We rescued: 
## 3,869 occurrences with missing eventDate.
##  — As it stands, there are 137,378 complete eventDates and 15,758 missing dates.
##  — There are also 138,464 complete year occurrences to filter from. This is up from an initial count of 127,188 At this rate, you will stand to lose 14,672 occurrences on the basis of missing year — Operation time: 38.2999749183655</code></pre>
<p>Remove the spent database</p>
<div class="sourceCode" id="cb390"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="#cb390-1" tabindex="-1"></a><span class="co">#rm(database)</span></span></code></pre></div>
</div>
<div id="no-eventdate" class="section level2">
<h2>6.2 No eventDate</h2>
<p>Flag records that simply lack collection date :*(</p>
<div class="sourceCode" id="cb391"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="#cb391-1" tabindex="-1"></a>check_time <span class="ot">&lt;-</span></span>
<span id="cb391-2"><a href="#cb391-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_eventDate_empty</span>(<span class="at">data =</span> check_time, <span class="at">eventDate =</span> <span class="st">&quot;eventDate&quot;</span>)</span></code></pre></div>
<pre><code>## 
## bdc_eventDate_empty:
## Flagged 15758 records.
## One column was added to the database.</code></pre>
</div>
<div id="year-from-eventdate" class="section level2">
<h2>6.3 Year from eventDate</h2>
<p>See if bdc can extract the year from the eventDate column (this
should already be done by dateFindR).</p>
<div class="sourceCode" id="cb393"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="#cb393-1" tabindex="-1"></a>check_time_year <span class="ot">&lt;-</span></span>
<span id="cb393-2"><a href="#cb393-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_year_from_eventDate</span>(<span class="at">data =</span> check_time, <span class="at">eventDate =</span> <span class="st">&quot;eventDate&quot;</span>)</span></code></pre></div>
<pre><code>## 
## bdc_year_from_eventDate:
## Four-digit year were extracted from 137378 records.</code></pre>
<div class="sourceCode" id="cb395"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb395-1"><a href="#cb395-1" tabindex="-1"></a><span class="fu">writeLines</span>(<span class="fu">paste</span>(<span class="st">&quot; — If there is a difference in the number of years in each column this value will &quot;</span>,</span>
<span id="cb395-2"><a href="#cb395-2" tabindex="-1"></a>                 <span class="st">&quot;not equal ZERO —— &quot;</span>,</span>
<span id="cb395-3"><a href="#cb395-3" tabindex="-1"></a>                 <span class="fu">sum</span>(<span class="fu">complete.cases</span>(check_time<span class="sc">$</span>year)) <span class="sc">-</span> <span class="fu">sum</span>(<span class="fu">complete.cases</span>(check_time_year[,<span class="st">&quot;year&quot;</span>])),</span>
<span id="cb395-4"><a href="#cb395-4" tabindex="-1"></a>                 <span class="st">&quot; —— IF it is negative, then that value is the number of new year values.&quot;</span>,</span>
<span id="cb395-5"><a href="#cb395-5" tabindex="-1"></a>                 <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)) </span></code></pre></div>
<pre><code>##  — If there is a difference in the number of years in each column this value will not equal ZERO —— 0 —— IF it is negative, then that value is the number of new year values.</code></pre>
<div class="sourceCode" id="cb397"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="#cb397-1" tabindex="-1"></a><span class="co"># Because this function makes TWO year columns, take only those rows that are NA in the original</span></span>
<span id="cb397-2"><a href="#cb397-2" tabindex="-1"></a>check_time<span class="sc">$</span>year <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(check_time<span class="sc">$</span>year), </span>
<span id="cb397-3"><a href="#cb397-3" tabindex="-1"></a>                          <span class="co"># IF the year column in the dataset is empty, take the newly-created </span></span>
<span id="cb397-4"><a href="#cb397-4" tabindex="-1"></a>                          <span class="co"># &quot;year&quot; columns from the bdc_year_from_eventDate dataset</span></span>
<span id="cb397-5"><a href="#cb397-5" tabindex="-1"></a>                          check_time_year[,<span class="fu">ncol</span>(check_time_year)], </span>
<span id="cb397-6"><a href="#cb397-6" tabindex="-1"></a>                          <span class="co"># If it is NOT empty, keep the original year.</span></span>
<span id="cb397-7"><a href="#cb397-7" tabindex="-1"></a>                          check_time<span class="sc">$</span>year)</span></code></pre></div>
<div class="sourceCode" id="cb398"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="#cb398-1" tabindex="-1"></a><span class="fu">rm</span>(check_time_year)</span></code></pre></div>
</div>
<div id="old-records" class="section level2">
<h2>6.4 Old records</h2>
<p>This will flag records prior to the date selected. 1970 is frequently
chosen for SDM work. You may not need to filter old records at all, so
think critically about your use.</p>
<div class="sourceCode" id="cb399"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="#cb399-1" tabindex="-1"></a>check_time <span class="ot">&lt;-</span></span>
<span id="cb399-2"><a href="#cb399-2" tabindex="-1"></a>  bdc<span class="sc">::</span><span class="fu">bdc_year_outOfRange</span>(<span class="at">data =</span> check_time,</span>
<span id="cb399-3"><a href="#cb399-3" tabindex="-1"></a>                           <span class="at">eventDate =</span> <span class="st">&quot;year&quot;</span>,</span>
<span id="cb399-4"><a href="#cb399-4" tabindex="-1"></a>                           <span class="at">year_threshold =</span> <span class="dv">1950</span>)</span></code></pre></div>
<pre><code>## 
## bdc_year_outOfRange:
## Flagged 12342 records.
## One column was added to the database.</code></pre>
</div>
<div id="time-report" class="section level2">
<h2>6.6 Time report</h2>
<p>Not all of it, just the time pertaining to our precise occurrence
records. Create a .summary column with all of the time flags where FALSE
== records that passed all filtering.</p>
<div class="sourceCode" id="cb401"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="#cb401-1" tabindex="-1"></a>check_time <span class="ot">&lt;-</span> bdc<span class="sc">::</span><span class="fu">bdc_summary_col</span>(<span class="at">data =</span> check_time)</span></code></pre></div>
<pre><code>## 
## bdc_summary_col:
## Flagged 27862 records.
## One column was added to the database.</code></pre>
<div class="sourceCode" id="cb403"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="#cb403-1" tabindex="-1"></a>( report <span class="ot">&lt;-</span></span>
<span id="cb403-2"><a href="#cb403-2" tabindex="-1"></a>    bdc<span class="sc">::</span><span class="fu">bdc_create_report</span>(<span class="at">data =</span> check_time,</span>
<span id="cb403-3"><a href="#cb403-3" tabindex="-1"></a>                           <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb403-4"><a href="#cb403-4" tabindex="-1"></a>                           <span class="at">workflow_step =</span> <span class="st">&quot;time&quot;</span>,</span>
<span id="cb403-5"><a href="#cb403-5" tabindex="-1"></a>                           <span class="at">save_report =</span> <span class="cn">FALSE</span>)</span>
<span id="cb403-6"><a href="#cb403-6" tabindex="-1"></a>)   </span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-e1012985ffc42a267f99" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e1012985ffc42a267f99">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5"],["Records with empty eventDate","Records with year out-of-range (i.e., year in the future or\n                collected before a threshold defined by the user).","Summary of all tests","","(*) calculated in relation to total number of records, i.e. 153136 records"],[".eventDate_empty",".year_outOfRange",".summary","",""],["15758","12342","27862","",""],["10.29","8.06","18.19","",""]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Description<\/th>\n      <th>Test_name<\/th>\n      <th>Records_flagged<\/th>\n      <th>perc_number_records(*)<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":10,"dom":"Brtip","scrollX":true,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="time-figures" class="section level2">
<h2>6.7 Time figures</h2>
<p>Create figures</p>
<div class="sourceCode" id="cb404"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb404-1"><a href="#cb404-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_create_figures.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb404-2"><a href="#cb404-2" tabindex="-1"></a>figures <span class="ot">&lt;-</span></span>
<span id="cb404-3"><a href="#cb404-3" tabindex="-1"></a>  <span class="fu">jbd_create_figures</span>(<span class="at">data =</span> check_time,</span>
<span id="cb404-4"><a href="#cb404-4" tabindex="-1"></a>                     <span class="at">path =</span> DataPath,</span>
<span id="cb404-5"><a href="#cb404-5" tabindex="-1"></a>                     <span class="at">database_id =</span> <span class="st">&quot;database_id&quot;</span>,</span>
<span id="cb404-6"><a href="#cb404-6" tabindex="-1"></a>                     <span class="at">workflow_step =</span> <span class="st">&quot;time&quot;</span>,</span>
<span id="cb404-7"><a href="#cb404-7" tabindex="-1"></a>                     <span class="at">save_figures =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Check figures in E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Figures</code></pre>
<p>Check figures using</p>
<div class="sourceCode" id="cb406"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="#cb406-1" tabindex="-1"></a>figures<span class="sc">$</span>year</span></code></pre></div>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-98-1.png" width="768" /></p>
<p>Save the ~raw time dataset into the intermediate folder.</p>
<div class="sourceCode" id="cb407"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="#cb407-1" tabindex="-1"></a>check_time <span class="sc">%&gt;%</span></span>
<span id="cb407-2"><a href="#cb407-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb407-3"><a href="#cb407-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;04_time_database.csv&quot;</span>,</span>
<span id="cb407-4"><a href="#cb407-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Raw dataset can be re-read here</p>
<pre><code>        # check_time &lt;-
        #   readr::read_csv(paste(DataPath, &quot;Output&quot;, &quot;Intermediate&quot;, &quot;04_time_database.csv&quot;,
        #                          sep = &quot;/&quot;))</code></pre>
</div>
<div id="save-flags-2" class="section level2">
<h2>6.8 Save flags</h2>
<p>SAVE the flags so far.</p>
<div class="sourceCode" id="cb409"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="#cb409-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;flagRecorder.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb409-2"><a href="#cb409-2" tabindex="-1"></a><span class="fu">flagRecorder</span>(</span>
<span id="cb409-3"><a href="#cb409-3" tabindex="-1"></a>  <span class="at">data =</span> check_time,</span>
<span id="cb409-4"><a href="#cb409-4" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Report&quot;</span>, <span class="at">sep =</span><span class="st">&quot;&quot;</span>),</span>
<span id="cb409-5"><a href="#cb409-5" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">&quot;flagsRecorded_&quot;</span>, <span class="fu">Sys.Date</span>(),  <span class="st">&quot;.csv&quot;</span>),</span>
<span id="cb409-6"><a href="#cb409-6" tabindex="-1"></a>  <span class="at">idColumns =</span> <span class="fu">c</span>(<span class="st">&quot;database_id&quot;</span>, <span class="st">&quot;id&quot;</span>, <span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>, <span class="st">&quot;dataSource&quot;</span>),</span>
<span id="cb409-7"><a href="#cb409-7" tabindex="-1"></a>  <span class="at">findFile =</span> <span class="cn">TRUE</span>,</span>
<span id="cb409-8"><a href="#cb409-8" tabindex="-1"></a>  <span class="at">printSummary =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##  — Dates found in file name(s). Finding most-recent file from file name...
##  — Found the following file(s): 
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>## Rows: 153136 Columns: 29
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr  (5): database_id, id, catalogNumber, occurrenceID, dataSource
## lgl (24): .scientificName_empty, .coordinates_empty, .coordinates_outOfRange...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
##  — .summary column detected. This will be over-written.
## 
##  — Data saved to E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>##  — Selected 31 columns. These include:
## database_id, id, catalogNumber, occurrenceID, dataSource, .scientificName_empty, .coordinates_empty, .coordinates_outOfRange, .basisOfRecords_notStandard, .coordinates_country_inconsistent, .occurrenceAbsent, .unLicensed, .GBIFflags, .summary, .invalidName, .rou, .val, .equ, .zer, .cap, .cen, .gbf, .inst, .sequential, .lonFlag, .latFlag, .gridSummary, .uncertaintyThreshold, .countryOutlier, .eventDate_empty, and .year_outOfRange</code></pre>
<pre><code>## # A tibble: 153,136 × 31
##    database_id id    catalogNumber occurrenceID dataSource .scientificName_empty
##    &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;        &lt;chr&gt;      &lt;lgl&gt;                
##  1 Dorey_data… fd65… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  2 Dorey_data… b85a… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  3 Dorey_data… d95b… HYM36182      urn:lsid:oz… ALA_Apifo… TRUE                 
##  4 Dorey_data… 85bc… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
##  5 Dorey_data… b905… HYM36180      urn:lsid:oz… ALA_Apifo… TRUE                 
##  6 Dorey_data… f6fd… 132550620     https://www… ALA_Apifo… TRUE                 
##  7 Dorey_data… bd38… HYM42541      urn:lsid:oz… ALA_Apifo… TRUE                 
##  8 Dorey_data… f046… 31469707      https://www… ALA_Apifo… TRUE                 
##  9 Dorey_data… f3b7… 4244291       /sightings/… ALA_Apifo… TRUE                 
## 10 Dorey_data… 907a… &lt;NA&gt;          &lt;NA&gt;         ALA_Apifo… TRUE                 
## # ℹ 153,126 more rows
## # ℹ 25 more variables: .coordinates_empty &lt;lgl&gt;, .coordinates_outOfRange &lt;lgl&gt;,
## #   .basisOfRecords_notStandard &lt;lgl&gt;, .coordinates_country_inconsistent &lt;lgl&gt;,
## #   .occurrenceAbsent &lt;lgl&gt;, .unLicensed &lt;lgl&gt;, .GBIFflags &lt;lgl&gt;,
## #   .summary &lt;lgl&gt;, .invalidName &lt;lgl&gt;, .rou &lt;lgl&gt;, .val &lt;lgl&gt;, .equ &lt;lgl&gt;,
## #   .zer &lt;lgl&gt;, .cap &lt;lgl&gt;, .cen &lt;lgl&gt;, .gbf &lt;lgl&gt;, .inst &lt;lgl&gt;,
## #   .sequential &lt;lgl&gt;, .lonFlag &lt;lgl&gt;, .latFlag &lt;lgl&gt;, .gridSummary &lt;lgl&gt;, …</code></pre>
</div>
</div>
<div id="de-duplication" class="section level1">
<h1>7.0 De-duplication</h1>
<div id="deduplicate" class="section level2">
<h2>7.1 deDuplicate</h2>
<p>We will FLAG duplicates here. Duplicate flagging can be based on
“ID”: - user-specified — idColumns — This is done ITERATIVELY across
these columns Or based on collectInfoColumns. - This will use a set of
core columns - user-defined — collectionCols AND then with the above
columns, extra columns individually - user-defined — collectInfoColumns
— This is done ITERATIVELY across these columns</p>
<p>Find the most-recent deDuplication file</p>
<div class="sourceCode" id="cb414"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="#cb414-1" tabindex="-1"></a><span class="co"># prevRun &lt;- file_finder(path = DataPath, fileName = &quot;duplicateRun_ID_&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb415"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="#cb415-1" tabindex="-1"></a><span class="co">#These could be hacked to de-duplicate as you wish.</span></span>
<span id="cb415-2"><a href="#cb415-2" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;jbd_dupeSummary.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb415-3"><a href="#cb415-3" tabindex="-1"></a>check_time <span class="ot">&lt;-</span> <span class="fu">jbd_dupeSummary</span>(</span>
<span id="cb415-4"><a href="#cb415-4" tabindex="-1"></a>  <span class="at">data =</span> check_time,</span>
<span id="cb415-5"><a href="#cb415-5" tabindex="-1"></a>  <span class="at">path =</span> DataPath,</span>
<span id="cb415-6"><a href="#cb415-6" tabindex="-1"></a>  <span class="co"># If you want to append a prior run&#39;s file, insert the file path here.</span></span>
<span id="cb415-7"><a href="#cb415-7" tabindex="-1"></a>  <span class="at">appendPath =</span> <span class="cn">NULL</span>,</span>
<span id="cb415-8"><a href="#cb415-8" tabindex="-1"></a>  <span class="co"># Years ABOVE this will not be used to sort dateIdentified</span></span>
<span id="cb415-9"><a href="#cb415-9" tabindex="-1"></a>  <span class="at">maxYear =</span> lubridate<span class="sc">::</span><span class="fu">year</span>(<span class="fu">Sys.Date</span>()),</span>
<span id="cb415-10"><a href="#cb415-10" tabindex="-1"></a>  <span class="co"># Years BELOW this will not be used to sort dateIdentified</span></span>
<span id="cb415-11"><a href="#cb415-11" tabindex="-1"></a>  <span class="at">minYear =</span> <span class="dv">1700</span>,</span>
<span id="cb415-12"><a href="#cb415-12" tabindex="-1"></a>  <span class="co"># options are &quot;ID&quot;,&quot;collectionInfo&quot;, or &quot;both&quot;</span></span>
<span id="cb415-13"><a href="#cb415-13" tabindex="-1"></a>  <span class="at">duplicatedBy =</span> <span class="st">&quot;both&quot;</span>, <span class="co"># I&#39;m only running ID for the first lot because we might recover other info later</span></span>
<span id="cb415-14"><a href="#cb415-14" tabindex="-1"></a>  <span class="co"># The columns to generate completeness info from</span></span>
<span id="cb415-15"><a href="#cb415-15" tabindex="-1"></a>  <span class="at">completeness_cols =</span> <span class="fu">c</span>(<span class="st">&quot;decimalLatitude&quot;</span>,  <span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb415-16"><a href="#cb415-16" tabindex="-1"></a>                        <span class="st">&quot;scientificName&quot;</span>, <span class="st">&quot;eventDate&quot;</span>),</span>
<span id="cb415-17"><a href="#cb415-17" tabindex="-1"></a>  <span class="at">idColumns =</span> <span class="fu">c</span>(<span class="st">&quot;gbifID&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>, <span class="st">&quot;recordId&quot;</span>,<span class="st">&quot;id&quot;</span>),</span>
<span id="cb415-18"><a href="#cb415-18" tabindex="-1"></a>  <span class="co"># The columns to ADDITIONALLY consider when finding duplicates in collectionInfo</span></span>
<span id="cb415-19"><a href="#cb415-19" tabindex="-1"></a>  <span class="at">collectionCols =</span> <span class="fu">c</span>(<span class="st">&quot;decimalLatitude&quot;</span>, <span class="st">&quot;decimalLongitude&quot;</span>, <span class="st">&quot;scientificName&quot;</span>, <span class="st">&quot;eventDate&quot;</span>, </span>
<span id="cb415-20"><a href="#cb415-20" tabindex="-1"></a>                     <span class="st">&quot;recordedBy&quot;</span>),</span>
<span id="cb415-21"><a href="#cb415-21" tabindex="-1"></a>  <span class="co"># The columns to combine, one-by-one with the collectionCols</span></span>
<span id="cb415-22"><a href="#cb415-22" tabindex="-1"></a>  <span class="at">collectInfoColumns =</span> <span class="fu">c</span>(<span class="st">&quot;recordNumber&quot;</span>, <span class="st">&quot;eventID&quot;</span>, <span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;otherCatalogNumbers&quot;</span>,</span>
<span id="cb415-23"><a href="#cb415-23" tabindex="-1"></a>                         <span class="st">&quot;collectionID&quot;</span>),</span>
<span id="cb415-24"><a href="#cb415-24" tabindex="-1"></a>  <span class="co"># Custom comparisons — as a list of </span></span>
<span id="cb415-25"><a href="#cb415-25" tabindex="-1"></a>  <span class="at">CustomComparisons =</span> tibble<span class="sc">::</span><span class="fu">lst</span>(<span class="fu">c</span>(<span class="st">&quot;catalogNumber&quot;</span>, <span class="st">&quot;institutionCode&quot;</span>)),</span>
<span id="cb415-26"><a href="#cb415-26" tabindex="-1"></a>  <span class="co"># The order in which you want to KEEP duplicated based on data source</span></span>
<span id="cb415-27"><a href="#cb415-27" tabindex="-1"></a>  <span class="co"># try unique(check_time$dataSource)</span></span>
<span id="cb415-28"><a href="#cb415-28" tabindex="-1"></a>  <span class="at">sourceOrder =</span> <span class="fu">c</span>(<span class="st">&quot;CAES&quot;</span>, <span class="st">&quot;Gai&quot;</span>, <span class="st">&quot;Ecd&quot;</span>,<span class="st">&quot;BMont&quot;</span>, <span class="st">&quot;BMin&quot;</span>, <span class="st">&quot;SFU&quot;</span>, <span class="st">&quot;ASP&quot;</span>, <span class="st">&quot;KP&quot;</span>, <span class="st">&quot;EcoS&quot;</span>, <span class="st">&quot;EaCo&quot;</span>,</span>
<span id="cb415-29"><a href="#cb415-29" tabindex="-1"></a>                  <span class="st">&quot;USGS&quot;</span>, <span class="st">&quot;ALA&quot;</span>, <span class="st">&quot;INHS&quot;</span>, <span class="st">&quot;GBIF&quot;</span>,<span class="st">&quot;SCAN&quot;</span>,<span class="st">&quot;iDigBio&quot;</span>),</span>
<span id="cb415-30"><a href="#cb415-30" tabindex="-1"></a>  <span class="co"># !!!!!! BELS &gt; GeoLocate</span></span>
<span id="cb415-31"><a href="#cb415-31" tabindex="-1"></a>  <span class="co"># To preference Paige&#39;s data over the RAW GBIF data, use the below</span></span>
<span id="cb415-32"><a href="#cb415-32" tabindex="-1"></a>  <span class="at">PaigeSort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb415-33"><a href="#cb415-33" tabindex="-1"></a>  <span class="co"># To remove duplicates from CONSIDERATION once they are identified = TRUE, </span></span>
<span id="cb415-34"><a href="#cb415-34" tabindex="-1"></a>  <span class="co"># or not = FALSE (by iteration)</span></span>
<span id="cb415-35"><a href="#cb415-35" tabindex="-1"></a>  <span class="at">consecutiveRemoval =</span> <span class="cn">FALSE</span></span>
<span id="cb415-36"><a href="#cb415-36" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="co"># </span><span class="re">END</span><span class="co"> jbd_dupeSummary</span></span>
<span id="cb415-37"><a href="#cb415-37" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>(<span class="at">col_types =</span> <span class="fu">ColTypeR</span>())</span></code></pre></div>
<pre><code>##  — Generating a basic completeness summary from the decimalLatitude, decimalLongitude, scientificName, eventDate columns.
## This summary is simply the sum of complete.cases in each column. It ranges from zero to the N of columns. This will be used to sort duplicate rows and select the most-complete rows.
##  — Formatting the dateIdentified column to date format...
##  — Ordering Paige&#39;s updates...
##  — Ordering data by 1. dataSource, 2. yearIdentified, and 3. completeness...</code></pre>
<pre><code>##  — Working on ID duplicates...</code></pre>
<pre><code>##  — Identified  0  duplicate records across  0  unique records. 
## This was done using the column:  gbifID</code></pre>
<pre><code>## 
## Completed iteration 1 of 4</code></pre>
<pre><code>##  — Identified  654  duplicate records across  326  unique records. 
## This was done using the column:  occurrenceID</code></pre>
<pre><code>## 
## Completed iteration 2 of 4</code></pre>
<pre><code>##  — Identified  84  duplicate records across  42  unique records. 
## This was done using the column:  recordId</code></pre>
<pre><code>## 
## Completed iteration 3 of 4</code></pre>
<pre><code>##  — Identified  0  duplicate records across  0  unique records. 
## This was done using the column:  id</code></pre>
<pre><code>## 
## Completed iteration 4 of 4</code></pre>
<pre><code>##  — Working on collectionInfo duplicates...</code></pre>
<pre><code>##  — Identified 22 duplicate records across 8 unique records. 
## This was done using the columns decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and recordNumber</code></pre>
<pre><code>## 
## Completed iteration 1 of 5</code></pre>
<pre><code>##  — Identified 2,180 duplicate records across 749 unique records. 
## This was done using the columns decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and eventID</code></pre>
<pre><code>## 
## Completed iteration 2 of 5</code></pre>
<pre><code>##  — Identified 331 duplicate records across 165 unique records. 
## This was done using the columns decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and catalogNumber</code></pre>
<pre><code>## 
## Completed iteration 3 of 5</code></pre>
<pre><code>##  — Identified 708 duplicate records across 313 unique records. 
## This was done using the columns decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and otherCatalogNumbers</code></pre>
<pre><code>## 
## Completed iteration 4 of 5</code></pre>
<pre><code>##  — Identified 7,304 duplicate records across 2,896 unique records. 
## This was done using the columns decimalLatitude, decimalLongitude, scientificName, eventDate, recordedBy, and collectionID</code></pre>
<pre><code>## 
## Completed iteration 5 of 5</code></pre>
<pre><code>##  — Working on CustomComparisons duplicates...</code></pre>
<pre><code>##  — Identified 1,249 duplicate records across 573 unique records. 
## This was done using the column(s): catalogNumber, institutionCode</code></pre>
<pre><code>## 
## Completed iteration 1 of 1</code></pre>
<pre><code>##  — Keeping only unique occurrence pairs...
##  — Find and FIRST duplicate to keep and assign other associated duplicates to that one (i.e., across multiple tests a &#39;kept duplicate&#39;, could otherwise be removed)...
##  — Duplicates have been saved in the file and location: E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/duplicateRun_both_2023-08-05.csv
##  — Across the entire dataset, there are now 5,920 duplicates from a total of 153,136 occurrences.</code></pre>
<pre><code>##  — Completed in 0.43 minutes.</code></pre>
</div>
</div>
<div id="data-filtering" class="section level1">
<h1>8.0 Data filtering</h1>
<div id="rm-outliers" class="section level2">
<h2>8.1 rm Outliers</h2>
<p>Remove outliers based on manual identification of these records by
experts</p>
<div class="sourceCode" id="cb442"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb442-1"><a href="#cb442-1" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">&quot;D:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Paige_data/removedBecauseDeterminedOutlier.csv&quot;</span>,</span>
<span id="cb442-2"><a href="#cb442-2" tabindex="-1"></a>  <span class="at">col_types =</span> <span class="fu">ColTypeR</span>())</span></code></pre></div>
<pre><code>## Error: &#39;D:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Paige_data/removedBecauseDeterminedOutlier.csv&#39; does not exist.</code></pre>
<p>Find outliers by occurrenceID</p>
<div class="sourceCode" id="cb444"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb444-1"><a href="#cb444-1" tabindex="-1"></a>Outl_occID <span class="ot">&lt;-</span> check_time <span class="sc">%&gt;%</span></span>
<span id="cb444-2"><a href="#cb444-2" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">drop_na</span>(occurrenceID) <span class="sc">%&gt;%</span></span>
<span id="cb444-3"><a href="#cb444-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(occurrenceID <span class="sc">%in%</span> outliers<span class="sc">$</span>occurrenceID) </span></code></pre></div>
<pre><code>## Error in `dplyr::filter()`:
## ℹ In argument: `occurrenceID %in% outliers$occurrenceID`.
## Caused by error:
## ! object &#39;outliers&#39; not found</code></pre>
<p>Find outliers by occurrenceID</p>
<div class="sourceCode" id="cb446"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb446-1"><a href="#cb446-1" tabindex="-1"></a>Outliers_matched <span class="ot">&lt;-</span> check_time <span class="sc">%&gt;%</span></span>
<span id="cb446-2"><a href="#cb446-2" tabindex="-1"></a>  <span class="co"># Remove matched IDs</span></span>
<span id="cb446-3"><a href="#cb446-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span>occurrenceID <span class="sc">%in%</span> Outl_occID<span class="sc">$</span>occurrenceID) <span class="sc">%&gt;%</span></span>
<span id="cb446-4"><a href="#cb446-4" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">drop_na</span>(catalogNumber) <span class="sc">%&gt;%</span></span>
<span id="cb446-5"><a href="#cb446-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(catalogNumber <span class="sc">%in%</span> outliers<span class="sc">$</span>catalogNumber) <span class="sc">%&gt;%</span></span>
<span id="cb446-6"><a href="#cb446-6" tabindex="-1"></a>  <span class="co"># Re-bind the outlier matches </span></span>
<span id="cb446-7"><a href="#cb446-7" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(Outl_occID)</span></code></pre></div>
<pre><code>## Error in `dplyr::filter()`:
## ℹ In argument: `!occurrenceID %in% Outl_occID$occurrenceID`.
## Caused by error:
## ! object &#39;Outl_occID&#39; not found</code></pre>
<p>Remove the outliers from the dataset</p>
<div class="sourceCode" id="cb448"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="#cb448-1" tabindex="-1"></a>check_time <span class="ot">&lt;-</span> check_time <span class="sc">%&gt;%</span></span>
<span id="cb448-2"><a href="#cb448-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span>database_id <span class="sc">%in%</span> Outliers_matched<span class="sc">$</span>database_id)</span></code></pre></div>
<pre><code>## Error in `dplyr::filter()`:
## ℹ In argument: `!database_id %in% Outliers_matched$database_id`.
## Caused by error:
## ! object &#39;Outliers_matched&#39; not found</code></pre>
</div>
<div id="save-uncleaned" class="section level2">
<h2>8.2 Save uncleaned</h2>
<p>Save the uncleaned dataset</p>
<div class="sourceCode" id="cb450"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb450-1"><a href="#cb450-1" tabindex="-1"></a>check_time <span class="sc">%&gt;%</span> readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb450-2"><a href="#cb450-2" tabindex="-1"></a>                                <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;05_unCleaned_database.csv&quot;</span>,</span>
<span id="cb450-3"><a href="#cb450-3" tabindex="-1"></a>                                      <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Dataset can be re-read here</p>
<div class="sourceCode" id="cb451"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb451-1"><a href="#cb451-1" tabindex="-1"></a>    <span class="co"># check_time &lt;-</span></span>
<span id="cb451-2"><a href="#cb451-2" tabindex="-1"></a>    <span class="co">#   readr::read_csv(paste(DataPath, &quot;Output&quot;, &quot;Intermediate&quot;, &quot;05_unCleaned_database.csv&quot;,</span></span>
<span id="cb451-3"><a href="#cb451-3" tabindex="-1"></a>    <span class="co">#                          sep = &quot;/&quot;))</span></span></code></pre></div>
</div>
<div id="filter" class="section level2">
<h2>8.3 Filter</h2>
<p>Now clean the dataset of extra columns and failed rows….</p>
<div class="sourceCode" id="cb452"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb452-1"><a href="#cb452-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;summaryFun.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb452-2"><a href="#cb452-2" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">summaryFun</span>(</span>
<span id="cb452-3"><a href="#cb452-3" tabindex="-1"></a>  <span class="at">data =</span> check_time,</span>
<span id="cb452-4"><a href="#cb452-4" tabindex="-1"></a>  <span class="co"># Don&#39;t filter these columns (or NULL)</span></span>
<span id="cb452-5"><a href="#cb452-5" tabindex="-1"></a>  <span class="at">dontFilterThese =</span> <span class="fu">c</span>(<span class="st">&quot;.gridSummary&quot;</span>, <span class="st">&quot;.lonFlag&quot;</span>, <span class="st">&quot;.latFlag&quot;</span>, <span class="st">&quot;.uncer_terms&quot;</span>),</span>
<span id="cb452-6"><a href="#cb452-6" tabindex="-1"></a>  <span class="co"># Remove the filtering columns?</span></span>
<span id="cb452-7"><a href="#cb452-7" tabindex="-1"></a>  <span class="at">removeFilterColumns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb452-8"><a href="#cb452-8" tabindex="-1"></a>  <span class="co"># Filter to ONLY cleaned data?</span></span>
<span id="cb452-9"><a href="#cb452-9" tabindex="-1"></a>  <span class="at">filterClean =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>##  — We will NOT flag the following columns. However, they will remain in the data file.
## .gridSummary, .lonFlag, .latFlag, .uncer_terms</code></pre>
<pre><code>##  — summaryColumnR:
## Flagged 33,459 
##   The .summary column was added to the database.</code></pre>
<pre><code>##  — REMOVED all occurrences that were FALSE for the &#39;summary&#39; column.</code></pre>
<div class="sourceCode" id="cb456"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb456-1"><a href="#cb456-1" tabindex="-1"></a><span class="co">#rm(check_time)</span></span></code></pre></div>
<p>Save this CLEANED dataset</p>
<div class="sourceCode" id="cb457"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb457-1"><a href="#cb457-1" tabindex="-1"></a>output <span class="sc">%&gt;%</span></span>
<span id="cb457-2"><a href="#cb457-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">write_csv</span>(.,</span>
<span id="cb457-3"><a href="#cb457-3" tabindex="-1"></a>                   <span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;05_cleaned_database.csv&quot;</span>,</span>
<span id="cb457-4"><a href="#cb457-4" tabindex="-1"></a>                         <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb457-5"><a href="#cb457-5" tabindex="-1"></a><span class="co">#rm(output)</span></span></code></pre></div>
</div>
</div>
<div id="summary-figures" class="section level1">
<h1>9.0 Summary figures</h1>
<div id="duplicate-chorddiagrams" class="section level2">
<h2>9.1 Duplicate chordDiagrams</h2>
<div class="sourceCode" id="cb458"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb458-1"><a href="#cb458-1" tabindex="-1"></a><span class="co">#install.packages(&quot;circlize&quot;)</span></span>
<span id="cb458-2"><a href="#cb458-2" tabindex="-1"></a><span class="co">#if(!require(&quot;BiocManager&quot;, quietly = TRUE))</span></span>
<span id="cb458-3"><a href="#cb458-3" tabindex="-1"></a><span class="co">#  install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb458-4"><a href="#cb458-4" tabindex="-1"></a><span class="co">#BiocManager::install(&quot;ComplexHeatmap&quot;, force = TRUE)</span></span>
<span id="cb458-5"><a href="#cb458-5" tabindex="-1"></a><span class="co">#install.packages(&quot;paletteer&quot;)</span></span>
<span id="cb458-6"><a href="#cb458-6" tabindex="-1"></a><span class="fu">library</span>(paletteer)</span>
<span id="cb458-7"><a href="#cb458-7" tabindex="-1"></a><span class="fu">library</span>(circlize)</span></code></pre></div>
<pre><code>## ========================================
## circlize version 0.4.15
## CRAN page: https://cran.r-project.org/package=circlize
## Github page: https://github.com/jokergoo/circlize
## Documentation: https://jokergoo.github.io/circlize_book/book/
## 
## If you use it in published research, please cite:
## Gu, Z. circlize implements and enhances circular visualization
##   in R. Bioinformatics 2014.
## 
## This message can be suppressed by:
##   suppressPackageStartupMessages(library(circlize))
## ========================================</code></pre>
<div class="sourceCode" id="cb460"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb460-1"><a href="#cb460-1" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)<span class="co"># Find palettes here</span></span></code></pre></div>
<pre><code>## Loading required package: grid</code></pre>
<pre><code>## 
## Attaching package: &#39;grid&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:terra&#39;:
## 
##     depth</code></pre>
<pre><code>## ========================================
## ComplexHeatmap version 2.12.1
## Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/
## Github page: https://github.com/jokergoo/ComplexHeatmap
## Documentation: http://jokergoo.github.io/ComplexHeatmap-reference
## 
## If you use it in published research, please cite either one:
## - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional 
##     genomic data. Bioinformatics 2016.
## - Gu, Z. Complex Heatmap Visualization. iMeta 2022.
## 
## 
## The new InteractiveComplexHeatmap package can directly export static 
## complex heatmaps into an interactive Shiny app with zero effort. Have a try!
## 
## This message can be suppressed by:
##   suppressPackageStartupMessages(library(ComplexHeatmap))
## ========================================</code></pre>
<pre><code>## 
## Attaching package: &#39;ComplexHeatmap&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:terra&#39;:
## 
##     draw</code></pre>
<pre><code>## The following object is masked from &#39;package:R.utils&#39;:
## 
##     draw</code></pre>
<div class="sourceCode" id="cb468"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb468-1"><a href="#cb468-1" tabindex="-1"></a><span class="co">#renv::snapshot()</span></span></code></pre></div>
<p>Read in the most-RECENT file</p>
<div class="sourceCode" id="cb469"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb469-1"><a href="#cb469-1" tabindex="-1"></a>duplicates <span class="ot">&lt;-</span> <span class="fu">file_finder</span>(<span class="at">path =</span> DataPath,</span>
<span id="cb469-2"><a href="#cb469-2" tabindex="-1"></a>                          <span class="at">fileName =</span> <span class="st">&quot;duplicateRun_&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb469-3"><a href="#cb469-3" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_csv</span>()</span></code></pre></div>
<pre><code>##  — Dates found in file name(s). Finding most-recent file from file name...
##  — Found the following file(s): 
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/duplicateRun_both_2023-08-05.csv</code></pre>
<pre><code>## Rows: 6636 Columns: 20
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr  (13): database_id, database_id_match, catalogNumber, institutionCode, d...
## dbl   (2): decimalLatitude, decimalLongitude
## lgl   (4): recordNumber, id, recordId, gbifID
## dttm  (1): eventDate
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Choose the global figure parameters</p>
<div class="sourceCode" id="cb472"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb472-1"><a href="#cb472-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>)<span class="sc">/</span><span class="dv">2</span>, <span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p>Create the chorDiagram. You can leave many of the below values out
but we show here the defaults</p>
<div class="sourceCode" id="cb473"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb473-1"><a href="#cb473-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;chordDiagramR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb473-2"><a href="#cb473-2" tabindex="-1"></a><span class="fu">chordDiagramR</span>(</span>
<span id="cb473-3"><a href="#cb473-3" tabindex="-1"></a>  <span class="co"># The duplicate data from the dupeSummary function output  </span></span>
<span id="cb473-4"><a href="#cb473-4" tabindex="-1"></a>  <span class="at">dupeData =</span> duplicates,</span>
<span id="cb473-5"><a href="#cb473-5" tabindex="-1"></a>  <span class="at">savePath =</span> <span class="fu">paste0</span>(DataPath, <span class="st">&quot;/Output/Figures/&quot;</span>, <span class="st">&quot;ChordDiagram.pdf&quot;</span>),</span>
<span id="cb473-6"><a href="#cb473-6" tabindex="-1"></a>  <span class="co"># These can be modified to help fit the final pdf that&#39;s exported.</span></span>
<span id="cb473-7"><a href="#cb473-7" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">9</span>,</span>
<span id="cb473-8"><a href="#cb473-8" tabindex="-1"></a>  <span class="at">height =</span> <span class="fl">7.5</span>,</span>
<span id="cb473-9"><a href="#cb473-9" tabindex="-1"></a>  <span class="at">bg =</span> <span class="st">&quot;white&quot;</span>,</span>
<span id="cb473-10"><a href="#cb473-10" tabindex="-1"></a>  <span class="co"># How few distinct dataSources should a group have to be listed as &quot;other&quot;</span></span>
<span id="cb473-11"><a href="#cb473-11" tabindex="-1"></a>  <span class="at">smallGrpThreshold =</span> <span class="dv">3</span>,</span>
<span id="cb473-12"><a href="#cb473-12" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;Duplicated record sources&quot;</span>,</span>
<span id="cb473-13"><a href="#cb473-13" tabindex="-1"></a>  <span class="co"># The default list of colour palettes to choose from usign the paleteer package</span></span>
<span id="cb473-14"><a href="#cb473-14" tabindex="-1"></a>  <span class="at">palettes =</span> <span class="fu">c</span>(<span class="st">&quot;cartography::blue.pal&quot;</span>, <span class="st">&quot;cartography::green.pal&quot;</span>, </span>
<span id="cb473-15"><a href="#cb473-15" tabindex="-1"></a>               <span class="st">&quot;cartography::sand.pal&quot;</span>, <span class="st">&quot;cartography::orange.pal&quot;</span>, <span class="st">&quot;cartography::red.pal&quot;</span>,</span>
<span id="cb473-16"><a href="#cb473-16" tabindex="-1"></a>               <span class="st">&quot;cartography::purple.pal&quot;</span>, <span class="st">&quot;cartography::brown.pal&quot;</span>),</span>
<span id="cb473-17"><a href="#cb473-17" tabindex="-1"></a>  <span class="at">canvas.ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.0</span>,<span class="fl">1.0</span>), </span>
<span id="cb473-18"><a href="#cb473-18" tabindex="-1"></a>  <span class="at">canvas.xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.6</span>, <span class="fl">0.25</span>),</span>
<span id="cb473-19"><a href="#cb473-19" tabindex="-1"></a>  <span class="at">text.col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb473-20"><a href="#cb473-20" tabindex="-1"></a>  <span class="at">legendX =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">6</span>, <span class="st">&quot;mm&quot;</span>),</span>
<span id="cb473-21"><a href="#cb473-21" tabindex="-1"></a>  <span class="at">legendY =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="dv">18</span>, <span class="st">&quot;mm&quot;</span>),</span>
<span id="cb473-22"><a href="#cb473-22" tabindex="-1"></a>  <span class="at">legendJustify =</span> <span class="fu">c</span>(<span class="st">&quot;left&quot;</span>, <span class="st">&quot;bottom&quot;</span>),</span>
<span id="cb473-23"><a href="#cb473-23" tabindex="-1"></a>  <span class="at">niceFacing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="A-globally-synthesized-and-flagged-bee-occurrence-dataset-and-cleaning-workflow-5.8.23_files/figure-html/unnamed-chunk-114-1.png" width="768" /></p>
<pre><code>## png 
##   2</code></pre>
<div class="sourceCode" id="cb475"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb475-1"><a href="#cb475-1" tabindex="-1"></a><span class="co"># Save 7*6</span></span></code></pre></div>
</div>
<div id="duplicate-histogram" class="section level2">
<h2>9.2 Duplicate histogram</h2>
<p>Find the existing flag file and read it in</p>
<div class="sourceCode" id="cb476"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb476-1"><a href="#cb476-1" tabindex="-1"></a>flagColumns <span class="ot">&lt;-</span> <span class="fu">file_finder</span>(<span class="at">path =</span> DataPath, <span class="at">fileName =</span> <span class="st">&quot;flagsRecorded_&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb476-2"><a href="#cb476-2" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_csv</span>()</span></code></pre></div>
<pre><code>##  — Dates found in file name(s). Finding most-recent file from file name...
##  — Found the following file(s): 
##  E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Report/flagsRecorded_2023-08-05.csv</code></pre>
<pre><code>## Rows: 153136 Columns: 31
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr  (5): database_id, id, catalogNumber, occurrenceID, dataSource
## lgl (26): .scientificName_empty, .coordinates_empty, .coordinates_outOfRange...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Read in the function and create a figure shoring the total number of
duplicates, kept duplicates, and unique records for each datasource
(simplified to the text before the first underscore). As well as the
proportion of the above for each data source</p>
<div class="sourceCode" id="cb479"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb479-1"><a href="#cb479-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;dupePlotR.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb479-2"><a href="#cb479-2" tabindex="-1"></a><span class="fu">dupePlotR</span>(</span>
<span id="cb479-3"><a href="#cb479-3" tabindex="-1"></a>  <span class="at">flagColumns =</span> flagColumns,</span>
<span id="cb479-4"><a href="#cb479-4" tabindex="-1"></a>  <span class="co"># The outpath to save the plot as</span></span>
<span id="cb479-5"><a href="#cb479-5" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste0</span>(DataPath, <span class="st">&quot;/Output&quot;</span>, <span class="st">&quot;/Figures&quot;</span>, <span class="st">&quot;/duplicatePlot_2.pdf&quot;</span>),</span>
<span id="cb479-6"><a href="#cb479-6" tabindex="-1"></a>  <span class="co"># Colours in order: duplicate, kept duplicate, unique</span></span>
<span id="cb479-7"><a href="#cb479-7" tabindex="-1"></a>  <span class="at">dupeColours =</span> <span class="fu">c</span>(<span class="st">&quot;#F2D2A2&quot;</span>,<span class="st">&quot;#B9D6BC&quot;</span>, <span class="st">&quot;#349B90&quot;</span>),</span>
<span id="cb479-8"><a href="#cb479-8" tabindex="-1"></a>  <span class="co"># Plot size and height</span></span>
<span id="cb479-9"><a href="#cb479-9" tabindex="-1"></a>  <span class="at">base_height =</span> <span class="dv">7</span>, <span class="at">base_width =</span> <span class="dv">7</span>,</span>
<span id="cb479-10"><a href="#cb479-10" tabindex="-1"></a>  <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.85</span>, <span class="fl">0.8</span>),</span>
<span id="cb479-11"><a href="#cb479-11" tabindex="-1"></a>  <span class="co"># Extra variables can be fed into forcats::fct_recode() to change names on plot</span></span>
<span id="cb479-12"><a href="#cb479-12" tabindex="-1"></a>  <span class="at">GBIF =</span> <span class="st">&quot;GBIF&quot;</span>, <span class="at">SCAN =</span> <span class="st">&quot;SCAN&quot;</span>, <span class="at">iDigBio =</span> <span class="st">&quot;iDigBio&quot;</span>, <span class="at">USGS =</span> <span class="st">&quot;USGS&quot;</span>, <span class="at">ALA =</span> <span class="st">&quot;ALA&quot;</span>, </span>
<span id="cb479-13"><a href="#cb479-13" tabindex="-1"></a>  <span class="at">ASP =</span> <span class="st">&quot;ASP&quot;</span>, <span class="at">CAES =</span> <span class="st">&quot;CAES&quot;</span>, <span class="st">&#39;B. Mont.&#39;</span> <span class="ot">=</span> <span class="st">&quot;BMont&quot;</span>, <span class="st">&#39;B. Minkley&#39;</span> <span class="ot">=</span> <span class="st">&quot;BMin&quot;</span>, <span class="at">Ecd =</span> <span class="st">&quot;Ecd&quot;</span>,</span>
<span id="cb479-14"><a href="#cb479-14" tabindex="-1"></a>  <span class="at">Gaiarsa =</span> <span class="st">&quot;Gai&quot;</span>, <span class="at">SFU =</span> <span class="st">&quot;SFU&quot;</span></span>
<span id="cb479-15"><a href="#cb479-15" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="flags-by-source" class="section level2">
<h2>9.3 Flags by source</h2>
<p>Use the uncleaned dataset</p>
<div class="sourceCode" id="cb480"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb480-1"><a href="#cb480-1" tabindex="-1"></a>beeData <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;05_unCleaned_database.csv&quot;</span>,</span>
<span id="cb480-2"><a href="#cb480-2" tabindex="-1"></a>                                 <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>),</span>
<span id="cb480-3"><a href="#cb480-3" tabindex="-1"></a>                           <span class="at">col_types =</span> <span class="fu">ColTypeR</span>())</span></code></pre></div>
<p>Read in the function amd Visualise all flags for each dataSource
(simplified to the text before the first underscore)</p>
<div class="sourceCode" id="cb481"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb481-1"><a href="#cb481-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;plotFlagSummary.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb481-2"><a href="#cb481-2" tabindex="-1"></a></span>
<span id="cb481-3"><a href="#cb481-3" tabindex="-1"></a><span class="fu">plotFlagSummary</span>(</span>
<span id="cb481-4"><a href="#cb481-4" tabindex="-1"></a>  <span class="at">plotData =</span> beeData,</span>
<span id="cb481-5"><a href="#cb481-5" tabindex="-1"></a>  <span class="co"># Colours in order of pass (TRUE), fail (FALSE), and NA</span></span>
<span id="cb481-6"><a href="#cb481-6" tabindex="-1"></a>  <span class="at">flagColours =</span> <span class="fu">c</span>(<span class="st">&quot;#127852&quot;</span>, <span class="st">&quot;#A7002D&quot;</span>, <span class="st">&quot;#BDBABB&quot;</span>),</span>
<span id="cb481-7"><a href="#cb481-7" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(<span class="st">&quot;FlagsPlot_&quot;</span>, <span class="fu">Sys.Date</span>(),<span class="st">&quot;.pdf&quot;</span>),</span>
<span id="cb481-8"><a href="#cb481-8" tabindex="-1"></a>  <span class="at">outpath =</span> <span class="fu">paste0</span>(DataPath, <span class="st">&quot;/Output/Figures&quot;</span>),</span>
<span id="cb481-9"><a href="#cb481-9" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">15</span>, <span class="at">height =</span> <span class="dv">9</span>,</span>
<span id="cb481-10"><a href="#cb481-10" tabindex="-1"></a>    <span class="co"># OPTIONAL:</span></span>
<span id="cb481-11"><a href="#cb481-11" tabindex="-1"></a>      <span class="co">#       #  # Filter to species</span></span>
<span id="cb481-12"><a href="#cb481-12" tabindex="-1"></a>      <span class="co">#         speciesName = &quot;Holcopasites heliopsis&quot;,</span></span>
<span id="cb481-13"><a href="#cb481-13" tabindex="-1"></a>      <span class="co">#           # column to look in</span></span>
<span id="cb481-14"><a href="#cb481-14" tabindex="-1"></a>      <span class="co">#         nameColumn = &quot;species&quot;,</span></span>
<span id="cb481-15"><a href="#cb481-15" tabindex="-1"></a>      <span class="co">#         # Save the filtered data</span></span>
<span id="cb481-16"><a href="#cb481-16" tabindex="-1"></a>      <span class="co">#         saveFiltered = TRUE,</span></span>
<span id="cb481-17"><a href="#cb481-17" tabindex="-1"></a>      <span class="co">#   # Filter column to display on map</span></span>
<span id="cb481-18"><a href="#cb481-18" tabindex="-1"></a>      <span class="co">#         filterColumn = &quot;.summary&quot;,</span></span>
<span id="cb481-19"><a href="#cb481-19" tabindex="-1"></a>      <span class="co">#         plotMap = TRUE,</span></span>
<span id="cb481-20"><a href="#cb481-20" tabindex="-1"></a>      <span class="co">#     # amount to jitter points if desired, e.g. 0.25 or NULL</span></span>
<span id="cb481-21"><a href="#cb481-21" tabindex="-1"></a>      <span class="co">#   jitterValue = NULL,</span></span>
<span id="cb481-22"><a href="#cb481-22" tabindex="-1"></a>      <span class="co">#     # Map opacity value for points between 0 and 1</span></span>
<span id="cb481-23"><a href="#cb481-23" tabindex="-1"></a>      <span class="co">#   mapAlpha = 1,</span></span>
<span id="cb481-24"><a href="#cb481-24" tabindex="-1"></a>  <span class="co"># Extra variables can be fed into forcats::fct_recode() to change names on plot</span></span>
<span id="cb481-25"><a href="#cb481-25" tabindex="-1"></a>  <span class="at">GBIF =</span> <span class="st">&quot;GBIF&quot;</span>, <span class="at">SCAN =</span> <span class="st">&quot;SCAN&quot;</span>, <span class="at">iDigBio =</span> <span class="st">&quot;iDigBio&quot;</span>, <span class="at">USGS =</span> <span class="st">&quot;USGS&quot;</span>, <span class="at">ALA =</span> <span class="st">&quot;ALA&quot;</span>, </span>
<span id="cb481-26"><a href="#cb481-26" tabindex="-1"></a>  <span class="at">ASP =</span> <span class="st">&quot;ASP&quot;</span>, <span class="at">CAES =</span> <span class="st">&quot;CAES&quot;</span>, <span class="st">&#39;B. Mont.&#39;</span> <span class="ot">=</span> <span class="st">&quot;BMont&quot;</span>, <span class="st">&#39;B. Minkley&#39;</span> <span class="ot">=</span> <span class="st">&quot;BMin&quot;</span>, <span class="at">Ecd =</span> <span class="st">&quot;Ecd&quot;</span>,</span>
<span id="cb481-27"><a href="#cb481-27" tabindex="-1"></a>  <span class="at">Gaiarsa =</span> <span class="st">&quot;Gai&quot;</span>, <span class="at">SFU =</span> <span class="st">&quot;SFU&quot;</span></span>
<span id="cb481-28"><a href="#cb481-28" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>##  — Preparing data to plot...</code></pre>
<pre><code>## `summarise()` has grouped output by &#39;database&#39;, &#39;flags&#39;. You can override using
## the `.groups` argument.</code></pre>
<pre><code>##  — Building plot...</code></pre>
</div>
<div id="maps" class="section level2">
<h2>9.4 Maps</h2>
<p>Import CLEANED dataset (you can change this option)</p>
<div class="sourceCode" id="cb485"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb485-1"><a href="#cb485-1" tabindex="-1"></a>mapData <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="fu">paste</span>(DataPath, <span class="st">&quot;Output&quot;</span>, <span class="st">&quot;Intermediate&quot;</span>, <span class="st">&quot;05_cleaned_database.csv&quot;</span>,</span>
<span id="cb485-2"><a href="#cb485-2" tabindex="-1"></a>                                 <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<pre><code>## Rows: 119677 Columns: 101
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr  (75): database_id, scientificName, family, subfamily, genus, subgenus, ...
## dbl  (11): decimalLatitude, decimalLongitude, day, month, year, individualCo...
## lgl  (13): subspecies, island, spatiallyValid, coreid, associatedOrganisms, ...
## dttm  (2): eventDate, occurrenceYear
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Read in the function</p>
<div class="sourceCode" id="cb487"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb487-1"><a href="#cb487-1" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste</span>(ScriptPath, <span class="st">&quot;summaryMaps.R&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>))</span>
<span id="cb487-2"><a href="#cb487-2" tabindex="-1"></a>SummaryMaps <span class="ot">&lt;-</span> <span class="fu">summaryMaps</span>(</span>
<span id="cb487-3"><a href="#cb487-3" tabindex="-1"></a>  <span class="at">mapData =</span> mapData,</span>
<span id="cb487-4"><a href="#cb487-4" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">10</span>,</span>
<span id="cb487-5"><a href="#cb487-5" tabindex="-1"></a>  <span class="at">class_n =</span> <span class="dv">15</span>,</span>
<span id="cb487-6"><a href="#cb487-6" tabindex="-1"></a>  <span class="at">class_Style =</span> <span class="st">&quot;jenks&quot;</span>,</span>
<span id="cb487-7"><a href="#cb487-7" tabindex="-1"></a>  <span class="at">filename =</span> <span class="fu">paste0</span>(DataPath, <span class="st">&quot;/Output/Figures/&quot;</span>, <span class="st">&quot;CountryMaps_jenks.pdf&quot;</span>)</span>
<span id="cb487-8"><a href="#cb487-8" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Loading required package: classInt</code></pre>
<pre><code>##  — Extracting country data from points...</code></pre>
<pre><code>## although coordinates are longitude/latitude, st_intersection assumes that they
## are planar</code></pre>
<pre><code>## Extraction complete.</code></pre>
<div class="sourceCode" id="cb492"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb492-1"><a href="#cb492-1" tabindex="-1"></a>SummaryMaps</span></code></pre></div>
<pre><code>## [1] &quot;E:/BeeDataCleaning2022/BeeDataCleaning/BeeDataCleaning/BeeData/Output/Figures/CountryMaps_jenks.pdf&quot;</code></pre>
</div>
</div>
<div id="fin" class="section level1">
<h1>FIN</h1>
</div>
</div>

   
   
            
      

  <script>
    $(document).ready(function () {

			
 	 	$('#content img').addClass("image-thumb");
		
		$('#content img:not(.no-lightbox)').addClass("image-lb");
	$('#content').magnificPopup({
	    type:'image',
	    closeOnContentClick: false,
	    closeBtnInside: false,
	    delegate: '.image-lb',
	    gallery: {enabled: true },
	    image: {
	        verticalFit: true,
		titleSrc: 'alt'
	    }
 	});
 	    });
  </script>



    <!-- dynamically load mathjax for compatibility with self-contained -->
  <script>
    (function () {
	var script = document.createElement("script");
	script.type = "text/javascript";
	script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
	document.getElementsByTagName("head")[0].appendChild(script);
    })();
  </script>
  
</body>
</html>
